// rclnodejs/generated/test_msgs/test_msgs__msg__Builtins.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const DurationWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Duration.js');
const TimeWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Time.js');
const BuiltinsRefStruct = StructType({
  duration_value: DurationWrapper.refObjectType,
  time_value: TimeWrapper.refObjectType,
});
const BuiltinsRefArray = ArrayType(BuiltinsRefStruct);
const BuiltinsRefStructArray = StructType({
  data: BuiltinsRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class BuiltinsWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new BuiltinsRefStruct(other._refObject.toObject());
      this._wrapperFields.duration_value =  new DurationWrapper(other._wrapperFields.duration_value);
      this._wrapperFields.time_value =  new TimeWrapper(other._wrapperFields.time_value);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new BuiltinsRefStruct();
    this._wrapperFields.duration_value = new DurationWrapper();
    this._wrapperFields.time_value = new TimeWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new BuiltinsWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new BuiltinsArrayWrapper;
  }
  static get ArrayType() {
    return BuiltinsArrayWrapper;
  }
  static get refObjectArrayType() {
    return BuiltinsRefStructArray
  }
  static get refObjectType() {
    return BuiltinsRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.duration_value.freeze(own, checkConsistency);
    this._refObject.duration_value = this._wrapperFields.duration_value.refObject;
    this._wrapperFields.time_value.freeze(own, checkConsistency);
    this._refObject.time_value = this._wrapperFields.time_value.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.duration_value.copyRefObject(refObject.duration_value);
    this._wrapperFields.time_value.copyRefObject(refObject.time_value);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    DurationWrapper.freeStruct(refObject.duration_value);
    TimeWrapper.freeStruct(refObject.time_value);
  }
  static destoryRawROS(msg) {
    BuiltinsWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'test_msgs', subFolder: 'msg', interfaceName: 'Builtins'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get duration_value() {
    return this._wrapperFields.duration_value;
  }
  set duration_value(value) {
    if (value instanceof DurationWrapper) {
      this._wrapperFields.duration_value.copy(value);
    } else {
      this._wrapperFields.duration_value.copy(new DurationWrapper(value));
    }
  }
  get time_value() {
    return this._wrapperFields.time_value;
  }
  set time_value(value) {
    if (value instanceof TimeWrapper) {
      this._wrapperFields.time_value.copy(value);
    } else {
      this._wrapperFields.time_value.copy(new TimeWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new BuiltinsRefStruct(refObject.toObject());
    this._wrapperFields.duration_value.copyRefObject(this._refObject.duration_value);
    this._wrapperFields.time_value.copyRefObject(this._refObject.time_value);
  }
  copy(other) {
    this._refObject = new BuiltinsRefStruct(other._refObject.toObject());
    this._wrapperFields.duration_value.copy(other._wrapperFields.duration_value);
    this._wrapperFields.time_value.copy(other._wrapperFields.time_value);
  }
  static get classType() {
    return BuiltinsWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"duration_value","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Duration","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"time_value","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"test_msgs","type":"Builtins","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Builtins"};
  }
  hasMember(name) {
    let memberNames = ["duration_value","time_value"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class BuiltinsArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof BuiltinsWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new BuiltinsWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to BuiltinsArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to BuiltinsArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new BuiltinsRefArray(size);
    this._refObject = new BuiltinsRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new BuiltinsWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof BuiltinsArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "BuiltinsArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      BuiltinsWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return BuiltinsWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return BuiltinsArrayWrapper;
  }
}
module.exports = BuiltinsWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "duration_value",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Duration",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "time_value",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Time",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "test_msgs",
    "type": "Builtins",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Builtins"
}
*/
