// rclnodejs/generated/test_msgs/test_msgs__msg__Strings.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const StringsRefStruct = StructType({
  string_value: primitiveTypes.string,
  string_value_default1: primitiveTypes.string,
  string_value_default2: primitiveTypes.string,
  string_value_default3: primitiveTypes.string,
  string_value_default4: primitiveTypes.string,
  string_value_default5: primitiveTypes.string,
  bounded_string_value: primitiveTypes.string,
  bounded_string_value_default1: primitiveTypes.string,
  bounded_string_value_default2: primitiveTypes.string,
  bounded_string_value_default3: primitiveTypes.string,
  bounded_string_value_default4: primitiveTypes.string,
  bounded_string_value_default5: primitiveTypes.string,
});
const StringsRefArray = ArrayType(StringsRefStruct);
const StringsRefStructArray = StructType({
  data: StringsRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class StringsWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new StringsRefStruct(other._refObject.toObject());
      this._string_valueIntialized = true;
      this._wrapperFields.string_value =  new StringWrapper(other._wrapperFields.string_value);
      this._string_value_default1Intialized = true;
      this._wrapperFields.string_value_default1 =  new StringWrapper(other._wrapperFields.string_value_default1);
      this._string_value_default2Intialized = true;
      this._wrapperFields.string_value_default2 =  new StringWrapper(other._wrapperFields.string_value_default2);
      this._string_value_default3Intialized = true;
      this._wrapperFields.string_value_default3 =  new StringWrapper(other._wrapperFields.string_value_default3);
      this._string_value_default4Intialized = true;
      this._wrapperFields.string_value_default4 =  new StringWrapper(other._wrapperFields.string_value_default4);
      this._string_value_default5Intialized = true;
      this._wrapperFields.string_value_default5 =  new StringWrapper(other._wrapperFields.string_value_default5);
      this._bounded_string_valueIntialized = true;
      this._wrapperFields.bounded_string_value =  new StringWrapper(other._wrapperFields.bounded_string_value);
      this._bounded_string_value_default1Intialized = true;
      this._wrapperFields.bounded_string_value_default1 =  new StringWrapper(other._wrapperFields.bounded_string_value_default1);
      this._bounded_string_value_default2Intialized = true;
      this._wrapperFields.bounded_string_value_default2 =  new StringWrapper(other._wrapperFields.bounded_string_value_default2);
      this._bounded_string_value_default3Intialized = true;
      this._wrapperFields.bounded_string_value_default3 =  new StringWrapper(other._wrapperFields.bounded_string_value_default3);
      this._bounded_string_value_default4Intialized = true;
      this._wrapperFields.bounded_string_value_default4 =  new StringWrapper(other._wrapperFields.bounded_string_value_default4);
      this._bounded_string_value_default5Intialized = true;
      this._wrapperFields.bounded_string_value_default5 =  new StringWrapper(other._wrapperFields.bounded_string_value_default5);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new StringsRefStruct();
    this._string_valueIntialized = false;
    this._wrapperFields.string_value = new StringWrapper();
    this._refObject.string_value_default1 = "Hello world!";
    this._string_value_default1Intialized = true;
    this._wrapperFields.string_value_default1 = new StringWrapper();
    this._refObject.string_value_default2 = "Hello'world!";
    this._string_value_default2Intialized = true;
    this._wrapperFields.string_value_default2 = new StringWrapper();
    this._refObject.string_value_default3 = "Hello\"world!";
    this._string_value_default3Intialized = true;
    this._wrapperFields.string_value_default3 = new StringWrapper();
    this._refObject.string_value_default4 = "Hello'world!";
    this._string_value_default4Intialized = true;
    this._wrapperFields.string_value_default4 = new StringWrapper();
    this._refObject.string_value_default5 = "Hello\"world!";
    this._string_value_default5Intialized = true;
    this._wrapperFields.string_value_default5 = new StringWrapper();
    this._bounded_string_valueIntialized = false;
    this._wrapperFields.bounded_string_value = new StringWrapper();
    this._refObject.bounded_string_value_default1 = "Hello world!";
    this._bounded_string_value_default1Intialized = true;
    this._wrapperFields.bounded_string_value_default1 = new StringWrapper();
    this._refObject.bounded_string_value_default2 = "Hello'world!";
    this._bounded_string_value_default2Intialized = true;
    this._wrapperFields.bounded_string_value_default2 = new StringWrapper();
    this._refObject.bounded_string_value_default3 = "Hello\"world!";
    this._bounded_string_value_default3Intialized = true;
    this._wrapperFields.bounded_string_value_default3 = new StringWrapper();
    this._refObject.bounded_string_value_default4 = "Hello'world!";
    this._bounded_string_value_default4Intialized = true;
    this._wrapperFields.bounded_string_value_default4 = new StringWrapper();
    this._refObject.bounded_string_value_default5 = "Hello\"world!";
    this._bounded_string_value_default5Intialized = true;
    this._wrapperFields.bounded_string_value_default5 = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new StringsWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new StringsArrayWrapper;
  }
  static get ArrayType() {
    return StringsArrayWrapper;
  }
  static get refObjectArrayType() {
    return StringsRefStructArray
  }
  static get refObjectType() {
    return StringsRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._string_valueIntialized) {
        throw new TypeError('Invalid argument: string_value in Strings');
      }
      if (!this._string_value_default1Intialized) {
        throw new TypeError('Invalid argument: string_value_default1 in Strings');
      }
      if (!this._string_value_default2Intialized) {
        throw new TypeError('Invalid argument: string_value_default2 in Strings');
      }
      if (!this._string_value_default3Intialized) {
        throw new TypeError('Invalid argument: string_value_default3 in Strings');
      }
      if (!this._string_value_default4Intialized) {
        throw new TypeError('Invalid argument: string_value_default4 in Strings');
      }
      if (!this._string_value_default5Intialized) {
        throw new TypeError('Invalid argument: string_value_default5 in Strings');
      }
      if (!this._bounded_string_valueIntialized) {
        throw new TypeError('Invalid argument: bounded_string_value in Strings');
      }
      if (!this._bounded_string_value_default1Intialized) {
        throw new TypeError('Invalid argument: bounded_string_value_default1 in Strings');
      }
      if (!this._bounded_string_value_default2Intialized) {
        throw new TypeError('Invalid argument: bounded_string_value_default2 in Strings');
      }
      if (!this._bounded_string_value_default3Intialized) {
        throw new TypeError('Invalid argument: bounded_string_value_default3 in Strings');
      }
      if (!this._bounded_string_value_default4Intialized) {
        throw new TypeError('Invalid argument: bounded_string_value_default4 in Strings');
      }
      if (!this._bounded_string_value_default5Intialized) {
        throw new TypeError('Invalid argument: bounded_string_value_default5 in Strings');
      }
    }
    if (own) {
      this._wrapperFields.string_value.freeze(own, checkConsistency);
    }
    this._refObject.string_value = this._wrapperFields.string_value.refObject;
    if (own) {
      this._wrapperFields.string_value_default1.freeze(own, checkConsistency);
    }
    this._refObject.string_value_default1 = this._wrapperFields.string_value_default1.refObject;
    if (own) {
      this._wrapperFields.string_value_default2.freeze(own, checkConsistency);
    }
    this._refObject.string_value_default2 = this._wrapperFields.string_value_default2.refObject;
    if (own) {
      this._wrapperFields.string_value_default3.freeze(own, checkConsistency);
    }
    this._refObject.string_value_default3 = this._wrapperFields.string_value_default3.refObject;
    if (own) {
      this._wrapperFields.string_value_default4.freeze(own, checkConsistency);
    }
    this._refObject.string_value_default4 = this._wrapperFields.string_value_default4.refObject;
    if (own) {
      this._wrapperFields.string_value_default5.freeze(own, checkConsistency);
    }
    this._refObject.string_value_default5 = this._wrapperFields.string_value_default5.refObject;
    if (own) {
      this._wrapperFields.bounded_string_value.freeze(own, checkConsistency);
    }
    this._refObject.bounded_string_value = this._wrapperFields.bounded_string_value.refObject;
    if (own) {
      this._wrapperFields.bounded_string_value_default1.freeze(own, checkConsistency);
    }
    this._refObject.bounded_string_value_default1 = this._wrapperFields.bounded_string_value_default1.refObject;
    if (own) {
      this._wrapperFields.bounded_string_value_default2.freeze(own, checkConsistency);
    }
    this._refObject.bounded_string_value_default2 = this._wrapperFields.bounded_string_value_default2.refObject;
    if (own) {
      this._wrapperFields.bounded_string_value_default3.freeze(own, checkConsistency);
    }
    this._refObject.bounded_string_value_default3 = this._wrapperFields.bounded_string_value_default3.refObject;
    if (own) {
      this._wrapperFields.bounded_string_value_default4.freeze(own, checkConsistency);
    }
    this._refObject.bounded_string_value_default4 = this._wrapperFields.bounded_string_value_default4.refObject;
    if (own) {
      this._wrapperFields.bounded_string_value_default5.freeze(own, checkConsistency);
    }
    this._refObject.bounded_string_value_default5 = this._wrapperFields.bounded_string_value_default5.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._string_valueIntialized = true;
    this._wrapperFields.string_value.data = refObject.string_value.data;
    this._string_value_default1Intialized = true;
    this._wrapperFields.string_value_default1.data = refObject.string_value_default1.data;
    this._string_value_default2Intialized = true;
    this._wrapperFields.string_value_default2.data = refObject.string_value_default2.data;
    this._string_value_default3Intialized = true;
    this._wrapperFields.string_value_default3.data = refObject.string_value_default3.data;
    this._string_value_default4Intialized = true;
    this._wrapperFields.string_value_default4.data = refObject.string_value_default4.data;
    this._string_value_default5Intialized = true;
    this._wrapperFields.string_value_default5.data = refObject.string_value_default5.data;
    this._bounded_string_valueIntialized = true;
    this._wrapperFields.bounded_string_value.data = refObject.bounded_string_value.data;
    this._bounded_string_value_default1Intialized = true;
    this._wrapperFields.bounded_string_value_default1.data = refObject.bounded_string_value_default1.data;
    this._bounded_string_value_default2Intialized = true;
    this._wrapperFields.bounded_string_value_default2.data = refObject.bounded_string_value_default2.data;
    this._bounded_string_value_default3Intialized = true;
    this._wrapperFields.bounded_string_value_default3.data = refObject.bounded_string_value_default3.data;
    this._bounded_string_value_default4Intialized = true;
    this._wrapperFields.bounded_string_value_default4.data = refObject.bounded_string_value_default4.data;
    this._bounded_string_value_default5Intialized = true;
    this._wrapperFields.bounded_string_value_default5.data = refObject.bounded_string_value_default5.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.string_value);
    StringWrapper.freeStruct(refObject.string_value_default1);
    StringWrapper.freeStruct(refObject.string_value_default2);
    StringWrapper.freeStruct(refObject.string_value_default3);
    StringWrapper.freeStruct(refObject.string_value_default4);
    StringWrapper.freeStruct(refObject.string_value_default5);
    StringWrapper.freeStruct(refObject.bounded_string_value);
    StringWrapper.freeStruct(refObject.bounded_string_value_default1);
    StringWrapper.freeStruct(refObject.bounded_string_value_default2);
    StringWrapper.freeStruct(refObject.bounded_string_value_default3);
    StringWrapper.freeStruct(refObject.bounded_string_value_default4);
    StringWrapper.freeStruct(refObject.bounded_string_value_default5);
  }
  static destoryRawROS(msg) {
    StringsWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'test_msgs', subFolder: 'msg', interfaceName: 'Strings'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get string_value() {
    if (!this._string_valueIntialized) {
      return undefined;
    }
    return this._wrapperFields.string_value.data;
  }
  set string_value(value) {
    this._string_valueIntialized = true;
    this._wrapperFields.string_value.data = value;
  }
  get string_value_default1() {
    if (!this._string_value_default1Intialized) {
      return undefined;
    }
    return this._wrapperFields.string_value_default1.data;
  }
  set string_value_default1(value) {
    this._string_value_default1Intialized = true;
    this._wrapperFields.string_value_default1.data = value;
  }
  get string_value_default2() {
    if (!this._string_value_default2Intialized) {
      return undefined;
    }
    return this._wrapperFields.string_value_default2.data;
  }
  set string_value_default2(value) {
    this._string_value_default2Intialized = true;
    this._wrapperFields.string_value_default2.data = value;
  }
  get string_value_default3() {
    if (!this._string_value_default3Intialized) {
      return undefined;
    }
    return this._wrapperFields.string_value_default3.data;
  }
  set string_value_default3(value) {
    this._string_value_default3Intialized = true;
    this._wrapperFields.string_value_default3.data = value;
  }
  get string_value_default4() {
    if (!this._string_value_default4Intialized) {
      return undefined;
    }
    return this._wrapperFields.string_value_default4.data;
  }
  set string_value_default4(value) {
    this._string_value_default4Intialized = true;
    this._wrapperFields.string_value_default4.data = value;
  }
  get string_value_default5() {
    if (!this._string_value_default5Intialized) {
      return undefined;
    }
    return this._wrapperFields.string_value_default5.data;
  }
  set string_value_default5(value) {
    this._string_value_default5Intialized = true;
    this._wrapperFields.string_value_default5.data = value;
  }
  get bounded_string_value() {
    if (!this._bounded_string_valueIntialized) {
      return undefined;
    }
    return this._wrapperFields.bounded_string_value.data;
  }
  set bounded_string_value(value) {
    this._bounded_string_valueIntialized = true;
    this._wrapperFields.bounded_string_value.data = value;
  }
  get bounded_string_value_default1() {
    if (!this._bounded_string_value_default1Intialized) {
      return undefined;
    }
    return this._wrapperFields.bounded_string_value_default1.data;
  }
  set bounded_string_value_default1(value) {
    this._bounded_string_value_default1Intialized = true;
    this._wrapperFields.bounded_string_value_default1.data = value;
  }
  get bounded_string_value_default2() {
    if (!this._bounded_string_value_default2Intialized) {
      return undefined;
    }
    return this._wrapperFields.bounded_string_value_default2.data;
  }
  set bounded_string_value_default2(value) {
    this._bounded_string_value_default2Intialized = true;
    this._wrapperFields.bounded_string_value_default2.data = value;
  }
  get bounded_string_value_default3() {
    if (!this._bounded_string_value_default3Intialized) {
      return undefined;
    }
    return this._wrapperFields.bounded_string_value_default3.data;
  }
  set bounded_string_value_default3(value) {
    this._bounded_string_value_default3Intialized = true;
    this._wrapperFields.bounded_string_value_default3.data = value;
  }
  get bounded_string_value_default4() {
    if (!this._bounded_string_value_default4Intialized) {
      return undefined;
    }
    return this._wrapperFields.bounded_string_value_default4.data;
  }
  set bounded_string_value_default4(value) {
    this._bounded_string_value_default4Intialized = true;
    this._wrapperFields.bounded_string_value_default4.data = value;
  }
  get bounded_string_value_default5() {
    if (!this._bounded_string_value_default5Intialized) {
      return undefined;
    }
    return this._wrapperFields.bounded_string_value_default5.data;
  }
  set bounded_string_value_default5(value) {
    this._bounded_string_value_default5Intialized = true;
    this._wrapperFields.bounded_string_value_default5.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new StringsRefStruct(refObject.toObject());
    this._string_valueIntialized = true;
    this._wrapperFields.string_value.copyRefObject(this._refObject.string_value);
    this._string_value_default1Intialized = true;
    this._wrapperFields.string_value_default1.copyRefObject(this._refObject.string_value_default1);
    this._string_value_default2Intialized = true;
    this._wrapperFields.string_value_default2.copyRefObject(this._refObject.string_value_default2);
    this._string_value_default3Intialized = true;
    this._wrapperFields.string_value_default3.copyRefObject(this._refObject.string_value_default3);
    this._string_value_default4Intialized = true;
    this._wrapperFields.string_value_default4.copyRefObject(this._refObject.string_value_default4);
    this._string_value_default5Intialized = true;
    this._wrapperFields.string_value_default5.copyRefObject(this._refObject.string_value_default5);
    this._bounded_string_valueIntialized = true;
    this._wrapperFields.bounded_string_value.copyRefObject(this._refObject.bounded_string_value);
    this._bounded_string_value_default1Intialized = true;
    this._wrapperFields.bounded_string_value_default1.copyRefObject(this._refObject.bounded_string_value_default1);
    this._bounded_string_value_default2Intialized = true;
    this._wrapperFields.bounded_string_value_default2.copyRefObject(this._refObject.bounded_string_value_default2);
    this._bounded_string_value_default3Intialized = true;
    this._wrapperFields.bounded_string_value_default3.copyRefObject(this._refObject.bounded_string_value_default3);
    this._bounded_string_value_default4Intialized = true;
    this._wrapperFields.bounded_string_value_default4.copyRefObject(this._refObject.bounded_string_value_default4);
    this._bounded_string_value_default5Intialized = true;
    this._wrapperFields.bounded_string_value_default5.copyRefObject(this._refObject.bounded_string_value_default5);
  }
  copy(other) {
    this._refObject = new StringsRefStruct(other._refObject.toObject());
    this._string_valueIntialized = true;
    this._wrapperFields.string_value.copy(other._wrapperFields.string_value);
    this._string_value_default1Intialized = true;
    this._wrapperFields.string_value_default1.copy(other._wrapperFields.string_value_default1);
    this._string_value_default2Intialized = true;
    this._wrapperFields.string_value_default2.copy(other._wrapperFields.string_value_default2);
    this._string_value_default3Intialized = true;
    this._wrapperFields.string_value_default3.copy(other._wrapperFields.string_value_default3);
    this._string_value_default4Intialized = true;
    this._wrapperFields.string_value_default4.copy(other._wrapperFields.string_value_default4);
    this._string_value_default5Intialized = true;
    this._wrapperFields.string_value_default5.copy(other._wrapperFields.string_value_default5);
    this._bounded_string_valueIntialized = true;
    this._wrapperFields.bounded_string_value.copy(other._wrapperFields.bounded_string_value);
    this._bounded_string_value_default1Intialized = true;
    this._wrapperFields.bounded_string_value_default1.copy(other._wrapperFields.bounded_string_value_default1);
    this._bounded_string_value_default2Intialized = true;
    this._wrapperFields.bounded_string_value_default2.copy(other._wrapperFields.bounded_string_value_default2);
    this._bounded_string_value_default3Intialized = true;
    this._wrapperFields.bounded_string_value_default3.copy(other._wrapperFields.bounded_string_value_default3);
    this._bounded_string_value_default4Intialized = true;
    this._wrapperFields.bounded_string_value_default4.copy(other._wrapperFields.bounded_string_value_default4);
    this._bounded_string_value_default5Intialized = true;
    this._wrapperFields.bounded_string_value_default5.copy(other._wrapperFields.bounded_string_value_default5);
  }
  static get classType() {
    return StringsWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"string","name":"STRING_CONST","value":"Hello world!"}],"fields":[{"name":"string_value","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"string_value_default1","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":"Hello world!"},{"name":"string_value_default2","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":"Hello'world!"},{"name":"string_value_default3","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":"Hello\"world!"},{"name":"string_value_default4","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":"Hello'world!"},{"name":"string_value_default5","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":"Hello\"world!"},{"name":"bounded_string_value","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":22,"isPrimitiveType":true},"default_value":null},{"name":"bounded_string_value_default1","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":22,"isPrimitiveType":true},"default_value":"Hello world!"},{"name":"bounded_string_value_default2","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":22,"isPrimitiveType":true},"default_value":"Hello'world!"},{"name":"bounded_string_value_default3","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":22,"isPrimitiveType":true},"default_value":"Hello\"world!"},{"name":"bounded_string_value_default4","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":22,"isPrimitiveType":true},"default_value":"Hello'world!"},{"name":"bounded_string_value_default5","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":22,"isPrimitiveType":true},"default_value":"Hello\"world!"}],"baseType":{"pkgName":"test_msgs","type":"Strings","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Strings"};
  }
  hasMember(name) {
    let memberNames = ["string_value","string_value_default1","string_value_default2","string_value_default3","string_value_default4","string_value_default5","bounded_string_value","bounded_string_value_default1","bounded_string_value_default2","bounded_string_value_default3","bounded_string_value_default4","bounded_string_value_default5"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class StringsArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof StringsWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new StringsWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to StringsArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to StringsArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new StringsRefArray(size);
    this._refObject = new StringsRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new StringsWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof StringsArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "StringsArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      StringsWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return StringsWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return StringsArrayWrapper;
  }
}
// Define constants (1 in total)
Object.defineProperty(StringsWrapper, "STRING_CONST", {value: "Hello world!", writable: false, enumerable: true, configurable: true});
module.exports = StringsWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "string",
      "name": "STRING_CONST",
      "value": "Hello world!"
    }
  ],
  "fields": [
    {
      "name": "string_value",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "string_value_default1",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": "Hello world!"
    },
    {
      "name": "string_value_default2",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": "Hello'world!"
    },
    {
      "name": "string_value_default3",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": "Hello\"world!"
    },
    {
      "name": "string_value_default4",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": "Hello'world!"
    },
    {
      "name": "string_value_default5",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": "Hello\"world!"
    },
    {
      "name": "bounded_string_value",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": 22,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "bounded_string_value_default1",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": 22,
        "isPrimitiveType": true
      },
      "default_value": "Hello world!"
    },
    {
      "name": "bounded_string_value_default2",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": 22,
        "isPrimitiveType": true
      },
      "default_value": "Hello'world!"
    },
    {
      "name": "bounded_string_value_default3",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": 22,
        "isPrimitiveType": true
      },
      "default_value": "Hello\"world!"
    },
    {
      "name": "bounded_string_value_default4",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": 22,
        "isPrimitiveType": true
      },
      "default_value": "Hello'world!"
    },
    {
      "name": "bounded_string_value_default5",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": 22,
        "isPrimitiveType": true
      },
      "default_value": "Hello\"world!"
    }
  ],
  "baseType": {
    "pkgName": "test_msgs",
    "type": "Strings",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Strings"
}
*/
