// rclnodejs/generated/test_msgs/test_msgs__action__NestedMessage_Result.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const BuiltinsWrapper = require('../../generated/test_msgs/test_msgs__msg__Builtins.js');
const BasicTypesWrapper = require('../../generated/test_msgs/test_msgs__msg__BasicTypes.js');
const TimeWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Time.js');
const NestedMessage_ResultRefStruct = StructType({
  nested_field_no_pkg: BuiltinsWrapper.refObjectType,
  nested_field: BasicTypesWrapper.refObjectType,
  nested_different_pkg: TimeWrapper.refObjectType,
});
const NestedMessage_ResultRefArray = ArrayType(NestedMessage_ResultRefStruct);
const NestedMessage_ResultRefStructArray = StructType({
  data: NestedMessage_ResultRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class NestedMessage_ResultWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new NestedMessage_ResultRefStruct(other._refObject.toObject());
      this._wrapperFields.nested_field_no_pkg =  new BuiltinsWrapper(other._wrapperFields.nested_field_no_pkg);
      this._wrapperFields.nested_field =  new BasicTypesWrapper(other._wrapperFields.nested_field);
      this._wrapperFields.nested_different_pkg =  new TimeWrapper(other._wrapperFields.nested_different_pkg);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new NestedMessage_ResultRefStruct();
    this._wrapperFields.nested_field_no_pkg = new BuiltinsWrapper();
    this._wrapperFields.nested_field = new BasicTypesWrapper();
    this._wrapperFields.nested_different_pkg = new TimeWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new NestedMessage_ResultWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new NestedMessage_ResultArrayWrapper;
  }
  static get ArrayType() {
    return NestedMessage_ResultArrayWrapper;
  }
  static get refObjectArrayType() {
    return NestedMessage_ResultRefStructArray
  }
  static get refObjectType() {
    return NestedMessage_ResultRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.nested_field_no_pkg.freeze(own, checkConsistency);
    this._refObject.nested_field_no_pkg = this._wrapperFields.nested_field_no_pkg.refObject;
    this._wrapperFields.nested_field.freeze(own, checkConsistency);
    this._refObject.nested_field = this._wrapperFields.nested_field.refObject;
    this._wrapperFields.nested_different_pkg.freeze(own, checkConsistency);
    this._refObject.nested_different_pkg = this._wrapperFields.nested_different_pkg.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.nested_field_no_pkg.copyRefObject(refObject.nested_field_no_pkg);
    this._wrapperFields.nested_field.copyRefObject(refObject.nested_field);
    this._wrapperFields.nested_different_pkg.copyRefObject(refObject.nested_different_pkg);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    BuiltinsWrapper.freeStruct(refObject.nested_field_no_pkg);
    BasicTypesWrapper.freeStruct(refObject.nested_field);
    TimeWrapper.freeStruct(refObject.nested_different_pkg);
  }
  static destoryRawROS(msg) {
    NestedMessage_ResultWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'test_msgs', subFolder: 'action', interfaceName: 'NestedMessage_Result'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get nested_field_no_pkg() {
    return this._wrapperFields.nested_field_no_pkg;
  }
  set nested_field_no_pkg(value) {
    if (value instanceof BuiltinsWrapper) {
      this._wrapperFields.nested_field_no_pkg.copy(value);
    } else {
      this._wrapperFields.nested_field_no_pkg.copy(new BuiltinsWrapper(value));
    }
  }
  get nested_field() {
    return this._wrapperFields.nested_field;
  }
  set nested_field(value) {
    if (value instanceof BasicTypesWrapper) {
      this._wrapperFields.nested_field.copy(value);
    } else {
      this._wrapperFields.nested_field.copy(new BasicTypesWrapper(value));
    }
  }
  get nested_different_pkg() {
    return this._wrapperFields.nested_different_pkg;
  }
  set nested_different_pkg(value) {
    if (value instanceof TimeWrapper) {
      this._wrapperFields.nested_different_pkg.copy(value);
    } else {
      this._wrapperFields.nested_different_pkg.copy(new TimeWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new NestedMessage_ResultRefStruct(refObject.toObject());
    this._wrapperFields.nested_field_no_pkg.copyRefObject(this._refObject.nested_field_no_pkg);
    this._wrapperFields.nested_field.copyRefObject(this._refObject.nested_field);
    this._wrapperFields.nested_different_pkg.copyRefObject(this._refObject.nested_different_pkg);
  }
  copy(other) {
    this._refObject = new NestedMessage_ResultRefStruct(other._refObject.toObject());
    this._wrapperFields.nested_field_no_pkg.copy(other._wrapperFields.nested_field_no_pkg);
    this._wrapperFields.nested_field.copy(other._wrapperFields.nested_field);
    this._wrapperFields.nested_different_pkg.copy(other._wrapperFields.nested_different_pkg);
  }
  static get classType() {
    return NestedMessage_ResultWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"nested_field_no_pkg","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"test_msgs","type":"Builtins","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"nested_field","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"test_msgs","type":"BasicTypes","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"nested_different_pkg","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"test_msgs","type":"NestedMessage_Result","stringUpperBound":null,"isPrimitiveType":false},"msgName":"NestedMessage_Result"};
  }
  hasMember(name) {
    let memberNames = ["nested_field_no_pkg","nested_field","nested_different_pkg"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class NestedMessage_ResultArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof NestedMessage_ResultWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new NestedMessage_ResultWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to NestedMessage_ResultArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to NestedMessage_ResultArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new NestedMessage_ResultRefArray(size);
    this._refObject = new NestedMessage_ResultRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new NestedMessage_ResultWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof NestedMessage_ResultArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "NestedMessage_ResultArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      NestedMessage_ResultWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return NestedMessage_ResultWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return NestedMessage_ResultArrayWrapper;
  }
}
module.exports = NestedMessage_ResultWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "nested_field_no_pkg",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "test_msgs",
        "type": "Builtins",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "nested_field",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "test_msgs",
        "type": "BasicTypes",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "nested_different_pkg",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Time",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "test_msgs",
    "type": "NestedMessage_Result",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "NestedMessage_Result"
}
*/
