// rclnodejs/generated/test_msgs/test_msgs__action__NestedMessage.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
class test_msgs__action__NestedMessage {
  static get Goal() {
    return require('./test_msgs__action__NestedMessage_Goal.js');
  }
  static get Result() {
    return require('./test_msgs__action__NestedMessage_Result.js');
  }
  static get Feedback() {
    return require('./test_msgs__action__NestedMessage_Feedback.js');
  }
  static get impl() {
    return {
      get SendGoalService() {
        return require('./test_msgs__action__NestedMessage_SendGoal.js');
      },
      get GetResultService() {
        return require('./test_msgs__action__NestedMessage_GetResult.js');
      },
      get FeedbackMessage() {
        return require('./test_msgs__action__NestedMessage_FeedbackMessage.js');
      },
      get CancelGoal() {
        return require('../action_msgs/action_msgs__srv__CancelGoal.js');
      },
      get GoalStatusArray() {
        return require('../action_msgs/action_msgs__msg__GoalStatusArray.js');
      }
    };
  }
  static type() {
    return {pkgName: 'test_msgs', subFolder: 'action', interfaceName: 'NestedMessage'};
  }
}
module.exports = test_msgs__action__NestedMessage;
