// rclnodejs/generated/test_msgs/test_msgs__msg__Defaults.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const DefaultsRefStruct = StructType({
  bool_value: primitiveTypes.bool,
  byte_value: primitiveTypes.byte,
  char_value: primitiveTypes.char,
  float32_value: primitiveTypes.float32,
  float64_value: primitiveTypes.float64,
  int8_value: primitiveTypes.int8,
  uint8_value: primitiveTypes.uint8,
  int16_value: primitiveTypes.int16,
  uint16_value: primitiveTypes.uint16,
  int32_value: primitiveTypes.int32,
  uint32_value: primitiveTypes.uint32,
  int64_value: primitiveTypes.int64,
  uint64_value: primitiveTypes.uint64,
});
const DefaultsRefArray = ArrayType(DefaultsRefStruct);
const DefaultsRefStructArray = StructType({
  data: DefaultsRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class DefaultsWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new DefaultsRefStruct(other._refObject.toObject());
      this._bool_valueIntialized = true;
      this._byte_valueIntialized = true;
      this._char_valueIntialized = true;
      this._float32_valueIntialized = true;
      this._float64_valueIntialized = true;
      this._int8_valueIntialized = true;
      this._uint8_valueIntialized = true;
      this._int16_valueIntialized = true;
      this._uint16_valueIntialized = true;
      this._int32_valueIntialized = true;
      this._uint32_valueIntialized = true;
      this._int64_valueIntialized = true;
      this._uint64_valueIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new DefaultsRefStruct();
    this._refObject.bool_value = true;
    this._bool_valueIntialized = true;
    this._refObject.byte_value = 50;
    this._byte_valueIntialized = true;
    this._refObject.char_value = 100;
    this._char_valueIntialized = true;
    this._refObject.float32_value = 1.125;
    this._float32_valueIntialized = true;
    this._refObject.float64_value = 1.125;
    this._float64_valueIntialized = true;
    this._refObject.int8_value = -50;
    this._int8_valueIntialized = true;
    this._refObject.uint8_value = 200;
    this._uint8_valueIntialized = true;
    this._refObject.int16_value = -1000;
    this._int16_valueIntialized = true;
    this._refObject.uint16_value = 2000;
    this._uint16_valueIntialized = true;
    this._refObject.int32_value = -30000;
    this._int32_valueIntialized = true;
    this._refObject.uint32_value = 60000;
    this._uint32_valueIntialized = true;
    this._refObject.int64_value = -40000000;
    this._int64_valueIntialized = true;
    this._refObject.uint64_value = 50000000;
    this._uint64_valueIntialized = true;
  }
  static createFromRefObject(refObject) {
    let self = new DefaultsWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new DefaultsArrayWrapper;
  }
  static get ArrayType() {
    return DefaultsArrayWrapper;
  }
  static get refObjectArrayType() {
    return DefaultsRefStructArray
  }
  static get refObjectType() {
    return DefaultsRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._bool_valueIntialized) {
        throw new TypeError('Invalid argument: bool_value in Defaults');
      }
      if (!this._byte_valueIntialized) {
        throw new TypeError('Invalid argument: byte_value in Defaults');
      }
      if (!this._char_valueIntialized) {
        throw new TypeError('Invalid argument: char_value in Defaults');
      }
      if (!this._float32_valueIntialized) {
        throw new TypeError('Invalid argument: float32_value in Defaults');
      }
      if (!this._float64_valueIntialized) {
        throw new TypeError('Invalid argument: float64_value in Defaults');
      }
      if (!this._int8_valueIntialized) {
        throw new TypeError('Invalid argument: int8_value in Defaults');
      }
      if (!this._uint8_valueIntialized) {
        throw new TypeError('Invalid argument: uint8_value in Defaults');
      }
      if (!this._int16_valueIntialized) {
        throw new TypeError('Invalid argument: int16_value in Defaults');
      }
      if (!this._uint16_valueIntialized) {
        throw new TypeError('Invalid argument: uint16_value in Defaults');
      }
      if (!this._int32_valueIntialized) {
        throw new TypeError('Invalid argument: int32_value in Defaults');
      }
      if (!this._uint32_valueIntialized) {
        throw new TypeError('Invalid argument: uint32_value in Defaults');
      }
      if (!this._int64_valueIntialized) {
        throw new TypeError('Invalid argument: int64_value in Defaults');
      }
      if (!this._uint64_valueIntialized) {
        throw new TypeError('Invalid argument: uint64_value in Defaults');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._bool_valueIntialized = true;
    this._refObject.bool_value = refObject.bool_value;
    this._byte_valueIntialized = true;
    this._refObject.byte_value = refObject.byte_value;
    this._char_valueIntialized = true;
    this._refObject.char_value = refObject.char_value;
    this._float32_valueIntialized = true;
    this._refObject.float32_value = refObject.float32_value;
    this._float64_valueIntialized = true;
    this._refObject.float64_value = refObject.float64_value;
    this._int8_valueIntialized = true;
    this._refObject.int8_value = refObject.int8_value;
    this._uint8_valueIntialized = true;
    this._refObject.uint8_value = refObject.uint8_value;
    this._int16_valueIntialized = true;
    this._refObject.int16_value = refObject.int16_value;
    this._uint16_valueIntialized = true;
    this._refObject.uint16_value = refObject.uint16_value;
    this._int32_valueIntialized = true;
    this._refObject.int32_value = refObject.int32_value;
    this._uint32_valueIntialized = true;
    this._refObject.uint32_value = refObject.uint32_value;
    this._int64_valueIntialized = true;
    this._refObject.int64_value = refObject.int64_value;
    this._uint64_valueIntialized = true;
    this._refObject.uint64_value = refObject.uint64_value;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    DefaultsWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'test_msgs', subFolder: 'msg', interfaceName: 'Defaults'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get bool_value() {
    if (!this._bool_valueIntialized) {
      return undefined;
    }
    return this._refObject.bool_value;
  }
  set bool_value(value) {
    this._bool_valueIntialized = true;
    this._refObject.bool_value = value;
  }
  get byte_value() {
    if (!this._byte_valueIntialized) {
      return undefined;
    }
    return this._refObject.byte_value;
  }
  set byte_value(value) {
    this._byte_valueIntialized = true;
    this._refObject.byte_value = value;
  }
  get char_value() {
    if (!this._char_valueIntialized) {
      return undefined;
    }
    return this._refObject.char_value;
  }
  set char_value(value) {
    this._char_valueIntialized = true;
    this._refObject.char_value = value;
  }
  get float32_value() {
    if (!this._float32_valueIntialized) {
      return undefined;
    }
    return this._refObject.float32_value;
  }
  set float32_value(value) {
    this._float32_valueIntialized = true;
    this._refObject.float32_value = value;
  }
  get float64_value() {
    if (!this._float64_valueIntialized) {
      return undefined;
    }
    return this._refObject.float64_value;
  }
  set float64_value(value) {
    this._float64_valueIntialized = true;
    this._refObject.float64_value = value;
  }
  get int8_value() {
    if (!this._int8_valueIntialized) {
      return undefined;
    }
    return this._refObject.int8_value;
  }
  set int8_value(value) {
    this._int8_valueIntialized = true;
    this._refObject.int8_value = value;
  }
  get uint8_value() {
    if (!this._uint8_valueIntialized) {
      return undefined;
    }
    return this._refObject.uint8_value;
  }
  set uint8_value(value) {
    this._uint8_valueIntialized = true;
    this._refObject.uint8_value = value;
  }
  get int16_value() {
    if (!this._int16_valueIntialized) {
      return undefined;
    }
    return this._refObject.int16_value;
  }
  set int16_value(value) {
    this._int16_valueIntialized = true;
    this._refObject.int16_value = value;
  }
  get uint16_value() {
    if (!this._uint16_valueIntialized) {
      return undefined;
    }
    return this._refObject.uint16_value;
  }
  set uint16_value(value) {
    this._uint16_valueIntialized = true;
    this._refObject.uint16_value = value;
  }
  get int32_value() {
    if (!this._int32_valueIntialized) {
      return undefined;
    }
    return this._refObject.int32_value;
  }
  set int32_value(value) {
    this._int32_valueIntialized = true;
    this._refObject.int32_value = value;
  }
  get uint32_value() {
    if (!this._uint32_valueIntialized) {
      return undefined;
    }
    return this._refObject.uint32_value;
  }
  set uint32_value(value) {
    this._uint32_valueIntialized = true;
    this._refObject.uint32_value = value;
  }
  get int64_value() {
    if (!this._int64_valueIntialized) {
      return undefined;
    }
    return this._refObject.int64_value;
  }
  set int64_value(value) {
    this._int64_valueIntialized = true;
    this._refObject.int64_value = value;
  }
  get uint64_value() {
    if (!this._uint64_valueIntialized) {
      return undefined;
    }
    return this._refObject.uint64_value;
  }
  set uint64_value(value) {
    this._uint64_valueIntialized = true;
    this._refObject.uint64_value = value;
  }
  copyRefObject(refObject) {
    this._refObject = new DefaultsRefStruct(refObject.toObject());
    this._bool_valueIntialized = true;
    this._byte_valueIntialized = true;
    this._char_valueIntialized = true;
    this._float32_valueIntialized = true;
    this._float64_valueIntialized = true;
    this._int8_valueIntialized = true;
    this._uint8_valueIntialized = true;
    this._int16_valueIntialized = true;
    this._uint16_valueIntialized = true;
    this._int32_valueIntialized = true;
    this._uint32_valueIntialized = true;
    this._int64_valueIntialized = true;
    this._uint64_valueIntialized = true;
  }
  copy(other) {
    this._refObject = new DefaultsRefStruct(other._refObject.toObject());
    this._bool_valueIntialized = true;
    this._byte_valueIntialized = true;
    this._char_valueIntialized = true;
    this._float32_valueIntialized = true;
    this._float64_valueIntialized = true;
    this._int8_valueIntialized = true;
    this._uint8_valueIntialized = true;
    this._int16_valueIntialized = true;
    this._uint16_valueIntialized = true;
    this._int32_valueIntialized = true;
    this._uint32_valueIntialized = true;
    this._int64_valueIntialized = true;
    this._uint64_valueIntialized = true;
  }
  static get classType() {
    return DefaultsWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"bool_value","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":true},{"name":"byte_value","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"byte","stringUpperBound":null,"isPrimitiveType":true},"default_value":50},{"name":"char_value","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"char","stringUpperBound":null,"isPrimitiveType":true},"default_value":100},{"name":"float32_value","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":1.125},{"name":"float64_value","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":1.125},{"name":"int8_value","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int8","stringUpperBound":null,"isPrimitiveType":true},"default_value":-50},{"name":"uint8_value","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":200},{"name":"int16_value","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int16","stringUpperBound":null,"isPrimitiveType":true},"default_value":-1000},{"name":"uint16_value","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":2000},{"name":"int32_value","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":-30000},{"name":"uint32_value","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":60000},{"name":"int64_value","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int64","stringUpperBound":null,"isPrimitiveType":true},"default_value":-40000000},{"name":"uint64_value","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":50000000}],"baseType":{"pkgName":"test_msgs","type":"Defaults","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Defaults"};
  }
  hasMember(name) {
    let memberNames = ["bool_value","byte_value","char_value","float32_value","float64_value","int8_value","uint8_value","int16_value","uint16_value","int32_value","uint32_value","int64_value","uint64_value"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class DefaultsArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof DefaultsWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new DefaultsWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DefaultsArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DefaultsArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new DefaultsRefArray(size);
    this._refObject = new DefaultsRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new DefaultsWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof DefaultsArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "DefaultsArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      DefaultsWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return DefaultsWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return DefaultsArrayWrapper;
  }
}
module.exports = DefaultsWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "bool_value",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": true
    },
    {
      "name": "byte_value",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "byte",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": 50
    },
    {
      "name": "char_value",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "char",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": 100
    },
    {
      "name": "float32_value",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": 1.125
    },
    {
      "name": "float64_value",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": 1.125
    },
    {
      "name": "int8_value",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": -50
    },
    {
      "name": "uint8_value",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": 200
    },
    {
      "name": "int16_value",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": -1000
    },
    {
      "name": "uint16_value",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": 2000
    },
    {
      "name": "int32_value",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": -30000
    },
    {
      "name": "uint32_value",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": 60000
    },
    {
      "name": "int64_value",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": -40000000
    },
    {
      "name": "uint64_value",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": 50000000
    }
  ],
  "baseType": {
    "pkgName": "test_msgs",
    "type": "Defaults",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Defaults"
}
*/
