// rclnodejs/generated/dwb_msgs/dwb_msgs__srv__GenerateTrajectory_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Pose2DWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Pose2D.js');
const Twist2DWrapper = require('../../generated/nav_2d_msgs/nav_2d_msgs__msg__Twist2D.js');
const GenerateTrajectory_RequestRefStruct = StructType({
  start_pose: Pose2DWrapper.refObjectType,
  start_vel: Twist2DWrapper.refObjectType,
  cmd_vel: Twist2DWrapper.refObjectType,
});
const GenerateTrajectory_RequestRefArray = ArrayType(GenerateTrajectory_RequestRefStruct);
const GenerateTrajectory_RequestRefStructArray = StructType({
  data: GenerateTrajectory_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GenerateTrajectory_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GenerateTrajectory_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.start_pose =  new Pose2DWrapper(other._wrapperFields.start_pose);
      this._wrapperFields.start_vel =  new Twist2DWrapper(other._wrapperFields.start_vel);
      this._wrapperFields.cmd_vel =  new Twist2DWrapper(other._wrapperFields.cmd_vel);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GenerateTrajectory_RequestRefStruct();
    this._wrapperFields.start_pose = new Pose2DWrapper();
    this._wrapperFields.start_vel = new Twist2DWrapper();
    this._wrapperFields.cmd_vel = new Twist2DWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GenerateTrajectory_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GenerateTrajectory_RequestArrayWrapper;
  }
  static get ArrayType() {
    return GenerateTrajectory_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return GenerateTrajectory_RequestRefStructArray
  }
  static get refObjectType() {
    return GenerateTrajectory_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.start_pose.freeze(own, checkConsistency);
    this._refObject.start_pose = this._wrapperFields.start_pose.refObject;
    this._wrapperFields.start_vel.freeze(own, checkConsistency);
    this._refObject.start_vel = this._wrapperFields.start_vel.refObject;
    this._wrapperFields.cmd_vel.freeze(own, checkConsistency);
    this._refObject.cmd_vel = this._wrapperFields.cmd_vel.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.start_pose.copyRefObject(refObject.start_pose);
    this._wrapperFields.start_vel.copyRefObject(refObject.start_vel);
    this._wrapperFields.cmd_vel.copyRefObject(refObject.cmd_vel);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    Pose2DWrapper.freeStruct(refObject.start_pose);
    Twist2DWrapper.freeStruct(refObject.start_vel);
    Twist2DWrapper.freeStruct(refObject.cmd_vel);
  }
  static destoryRawROS(msg) {
    GenerateTrajectory_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'dwb_msgs', subFolder: 'srv', interfaceName: 'GenerateTrajectory_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get start_pose() {
    return this._wrapperFields.start_pose;
  }
  set start_pose(value) {
    if (value instanceof Pose2DWrapper) {
      this._wrapperFields.start_pose.copy(value);
    } else {
      this._wrapperFields.start_pose.copy(new Pose2DWrapper(value));
    }
  }
  get start_vel() {
    return this._wrapperFields.start_vel;
  }
  set start_vel(value) {
    if (value instanceof Twist2DWrapper) {
      this._wrapperFields.start_vel.copy(value);
    } else {
      this._wrapperFields.start_vel.copy(new Twist2DWrapper(value));
    }
  }
  get cmd_vel() {
    return this._wrapperFields.cmd_vel;
  }
  set cmd_vel(value) {
    if (value instanceof Twist2DWrapper) {
      this._wrapperFields.cmd_vel.copy(value);
    } else {
      this._wrapperFields.cmd_vel.copy(new Twist2DWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new GenerateTrajectory_RequestRefStruct(refObject.toObject());
    this._wrapperFields.start_pose.copyRefObject(this._refObject.start_pose);
    this._wrapperFields.start_vel.copyRefObject(this._refObject.start_vel);
    this._wrapperFields.cmd_vel.copyRefObject(this._refObject.cmd_vel);
  }
  copy(other) {
    this._refObject = new GenerateTrajectory_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.start_pose.copy(other._wrapperFields.start_pose);
    this._wrapperFields.start_vel.copy(other._wrapperFields.start_vel);
    this._wrapperFields.cmd_vel.copy(other._wrapperFields.cmd_vel);
  }
  static get classType() {
    return GenerateTrajectory_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"start_pose","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Pose2D","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"start_vel","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"nav_2d_msgs","type":"Twist2D","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"cmd_vel","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"nav_2d_msgs","type":"Twist2D","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"dwb_msgs","type":"GenerateTrajectory_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GenerateTrajectory_Request"};
  }
  hasMember(name) {
    let memberNames = ["start_pose","start_vel","cmd_vel"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GenerateTrajectory_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GenerateTrajectory_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GenerateTrajectory_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GenerateTrajectory_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GenerateTrajectory_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GenerateTrajectory_RequestRefArray(size);
    this._refObject = new GenerateTrajectory_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GenerateTrajectory_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GenerateTrajectory_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GenerateTrajectory_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GenerateTrajectory_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GenerateTrajectory_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GenerateTrajectory_RequestArrayWrapper;
  }
}
module.exports = GenerateTrajectory_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "start_pose",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Pose2D",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "start_vel",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "nav_2d_msgs",
        "type": "Twist2D",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "cmd_vel",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "nav_2d_msgs",
        "type": "Twist2D",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "dwb_msgs",
    "type": "GenerateTrajectory_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GenerateTrajectory_Request"
}
*/
