// rclnodejs/generated/dwb_msgs/dwb_msgs__srv__GenerateTwists_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Twist2DWrapper = require('../../generated/nav_2d_msgs/nav_2d_msgs__msg__Twist2D.js');
const GenerateTwists_RequestRefStruct = StructType({
  current_vel: Twist2DWrapper.refObjectType,
});
const GenerateTwists_RequestRefArray = ArrayType(GenerateTwists_RequestRefStruct);
const GenerateTwists_RequestRefStructArray = StructType({
  data: GenerateTwists_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GenerateTwists_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GenerateTwists_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.current_vel =  new Twist2DWrapper(other._wrapperFields.current_vel);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GenerateTwists_RequestRefStruct();
    this._wrapperFields.current_vel = new Twist2DWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GenerateTwists_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GenerateTwists_RequestArrayWrapper;
  }
  static get ArrayType() {
    return GenerateTwists_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return GenerateTwists_RequestRefStructArray
  }
  static get refObjectType() {
    return GenerateTwists_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.current_vel.freeze(own, checkConsistency);
    this._refObject.current_vel = this._wrapperFields.current_vel.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.current_vel.copyRefObject(refObject.current_vel);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    Twist2DWrapper.freeStruct(refObject.current_vel);
  }
  static destoryRawROS(msg) {
    GenerateTwists_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'dwb_msgs', subFolder: 'srv', interfaceName: 'GenerateTwists_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get current_vel() {
    return this._wrapperFields.current_vel;
  }
  set current_vel(value) {
    if (value instanceof Twist2DWrapper) {
      this._wrapperFields.current_vel.copy(value);
    } else {
      this._wrapperFields.current_vel.copy(new Twist2DWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new GenerateTwists_RequestRefStruct(refObject.toObject());
    this._wrapperFields.current_vel.copyRefObject(this._refObject.current_vel);
  }
  copy(other) {
    this._refObject = new GenerateTwists_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.current_vel.copy(other._wrapperFields.current_vel);
  }
  static get classType() {
    return GenerateTwists_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"current_vel","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"nav_2d_msgs","type":"Twist2D","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"dwb_msgs","type":"GenerateTwists_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GenerateTwists_Request"};
  }
  hasMember(name) {
    let memberNames = ["current_vel"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GenerateTwists_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GenerateTwists_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GenerateTwists_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GenerateTwists_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GenerateTwists_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GenerateTwists_RequestRefArray(size);
    this._refObject = new GenerateTwists_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GenerateTwists_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GenerateTwists_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GenerateTwists_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GenerateTwists_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GenerateTwists_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GenerateTwists_RequestArrayWrapper;
  }
}
module.exports = GenerateTwists_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "current_vel",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "nav_2d_msgs",
        "type": "Twist2D",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "dwb_msgs",
    "type": "GenerateTwists_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GenerateTwists_Request"
}
*/
