// rclnodejs/generated/dwb_msgs/dwb_msgs__srv__ScoreTrajectory_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Pose2DStampedWrapper = require('../../generated/nav_2d_msgs/nav_2d_msgs__msg__Pose2DStamped.js');
const Twist2DWrapper = require('../../generated/nav_2d_msgs/nav_2d_msgs__msg__Twist2D.js');
const Path2DWrapper = require('../../generated/nav_2d_msgs/nav_2d_msgs__msg__Path2D.js');
const Trajectory2DWrapper = require('../../generated/dwb_msgs/dwb_msgs__msg__Trajectory2D.js');
const ScoreTrajectory_RequestRefStruct = StructType({
  pose: Pose2DStampedWrapper.refObjectType,
  velocity: Twist2DWrapper.refObjectType,
  global_plan: Path2DWrapper.refObjectType,
  traj: Trajectory2DWrapper.refObjectType,
});
const ScoreTrajectory_RequestRefArray = ArrayType(ScoreTrajectory_RequestRefStruct);
const ScoreTrajectory_RequestRefStructArray = StructType({
  data: ScoreTrajectory_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ScoreTrajectory_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ScoreTrajectory_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.pose =  new Pose2DStampedWrapper(other._wrapperFields.pose);
      this._wrapperFields.velocity =  new Twist2DWrapper(other._wrapperFields.velocity);
      this._wrapperFields.global_plan =  new Path2DWrapper(other._wrapperFields.global_plan);
      this._wrapperFields.traj =  new Trajectory2DWrapper(other._wrapperFields.traj);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ScoreTrajectory_RequestRefStruct();
    this._wrapperFields.pose = new Pose2DStampedWrapper();
    this._wrapperFields.velocity = new Twist2DWrapper();
    this._wrapperFields.global_plan = new Path2DWrapper();
    this._wrapperFields.traj = new Trajectory2DWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ScoreTrajectory_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ScoreTrajectory_RequestArrayWrapper;
  }
  static get ArrayType() {
    return ScoreTrajectory_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return ScoreTrajectory_RequestRefStructArray
  }
  static get refObjectType() {
    return ScoreTrajectory_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.pose.freeze(own, checkConsistency);
    this._refObject.pose = this._wrapperFields.pose.refObject;
    this._wrapperFields.velocity.freeze(own, checkConsistency);
    this._refObject.velocity = this._wrapperFields.velocity.refObject;
    this._wrapperFields.global_plan.freeze(own, checkConsistency);
    this._refObject.global_plan = this._wrapperFields.global_plan.refObject;
    this._wrapperFields.traj.freeze(own, checkConsistency);
    this._refObject.traj = this._wrapperFields.traj.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.pose.copyRefObject(refObject.pose);
    this._wrapperFields.velocity.copyRefObject(refObject.velocity);
    this._wrapperFields.global_plan.copyRefObject(refObject.global_plan);
    this._wrapperFields.traj.copyRefObject(refObject.traj);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    Pose2DStampedWrapper.freeStruct(refObject.pose);
    Twist2DWrapper.freeStruct(refObject.velocity);
    Path2DWrapper.freeStruct(refObject.global_plan);
    Trajectory2DWrapper.freeStruct(refObject.traj);
  }
  static destoryRawROS(msg) {
    ScoreTrajectory_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'dwb_msgs', subFolder: 'srv', interfaceName: 'ScoreTrajectory_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get pose() {
    return this._wrapperFields.pose;
  }
  set pose(value) {
    if (value instanceof Pose2DStampedWrapper) {
      this._wrapperFields.pose.copy(value);
    } else {
      this._wrapperFields.pose.copy(new Pose2DStampedWrapper(value));
    }
  }
  get velocity() {
    return this._wrapperFields.velocity;
  }
  set velocity(value) {
    if (value instanceof Twist2DWrapper) {
      this._wrapperFields.velocity.copy(value);
    } else {
      this._wrapperFields.velocity.copy(new Twist2DWrapper(value));
    }
  }
  get global_plan() {
    return this._wrapperFields.global_plan;
  }
  set global_plan(value) {
    if (value instanceof Path2DWrapper) {
      this._wrapperFields.global_plan.copy(value);
    } else {
      this._wrapperFields.global_plan.copy(new Path2DWrapper(value));
    }
  }
  get traj() {
    return this._wrapperFields.traj;
  }
  set traj(value) {
    if (value instanceof Trajectory2DWrapper) {
      this._wrapperFields.traj.copy(value);
    } else {
      this._wrapperFields.traj.copy(new Trajectory2DWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new ScoreTrajectory_RequestRefStruct(refObject.toObject());
    this._wrapperFields.pose.copyRefObject(this._refObject.pose);
    this._wrapperFields.velocity.copyRefObject(this._refObject.velocity);
    this._wrapperFields.global_plan.copyRefObject(this._refObject.global_plan);
    this._wrapperFields.traj.copyRefObject(this._refObject.traj);
  }
  copy(other) {
    this._refObject = new ScoreTrajectory_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.pose.copy(other._wrapperFields.pose);
    this._wrapperFields.velocity.copy(other._wrapperFields.velocity);
    this._wrapperFields.global_plan.copy(other._wrapperFields.global_plan);
    this._wrapperFields.traj.copy(other._wrapperFields.traj);
  }
  static get classType() {
    return ScoreTrajectory_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"pose","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"nav_2d_msgs","type":"Pose2DStamped","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"velocity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"nav_2d_msgs","type":"Twist2D","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"global_plan","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"nav_2d_msgs","type":"Path2D","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"traj","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"dwb_msgs","type":"Trajectory2D","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"dwb_msgs","type":"ScoreTrajectory_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ScoreTrajectory_Request"};
  }
  hasMember(name) {
    let memberNames = ["pose","velocity","global_plan","traj"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ScoreTrajectory_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ScoreTrajectory_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ScoreTrajectory_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ScoreTrajectory_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ScoreTrajectory_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ScoreTrajectory_RequestRefArray(size);
    this._refObject = new ScoreTrajectory_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ScoreTrajectory_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ScoreTrajectory_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ScoreTrajectory_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ScoreTrajectory_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ScoreTrajectory_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ScoreTrajectory_RequestArrayWrapper;
  }
}
module.exports = ScoreTrajectory_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "pose",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "nav_2d_msgs",
        "type": "Pose2DStamped",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "velocity",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "nav_2d_msgs",
        "type": "Twist2D",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "global_plan",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "nav_2d_msgs",
        "type": "Path2D",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "traj",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "dwb_msgs",
        "type": "Trajectory2D",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "dwb_msgs",
    "type": "ScoreTrajectory_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ScoreTrajectory_Request"
}
*/
