// rclnodejs/generated/dwb_msgs/dwb_msgs__msg__TrajectoryScore.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Trajectory2DWrapper = require('../../generated/dwb_msgs/dwb_msgs__msg__Trajectory2D.js');
const CriticScoreWrapper = require('../../generated/dwb_msgs/dwb_msgs__msg__CriticScore.js');
const TrajectoryScoreRefStruct = StructType({
  traj: Trajectory2DWrapper.refObjectType,
  scores: CriticScoreWrapper.refObjectArrayType,
  total: primitiveTypes.float32,
});
const TrajectoryScoreRefArray = ArrayType(TrajectoryScoreRefStruct);
const TrajectoryScoreRefStructArray = StructType({
  data: TrajectoryScoreRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class TrajectoryScoreWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new TrajectoryScoreRefStruct(other._refObject.toObject());
      this._wrapperFields.traj =  new Trajectory2DWrapper(other._wrapperFields.traj);
      this._wrapperFields.scores = CriticScoreWrapper.createArray();
      this._wrapperFields.scores.copy(other._wrapperFields.scores);
      this._totalIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new TrajectoryScoreRefStruct();
    this._wrapperFields.traj = new Trajectory2DWrapper();
    this._wrapperFields.scores = CriticScoreWrapper.createArray();
    this._totalIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new TrajectoryScoreWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new TrajectoryScoreArrayWrapper;
  }
  static get ArrayType() {
    return TrajectoryScoreArrayWrapper;
  }
  static get refObjectArrayType() {
    return TrajectoryScoreRefStructArray
  }
  static get refObjectType() {
    return TrajectoryScoreRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._totalIntialized) {
        throw new TypeError('Invalid argument: total in TrajectoryScore');
      }
    }
    this._wrapperFields.traj.freeze(own, checkConsistency);
    this._refObject.traj = this._wrapperFields.traj.refObject;
    this._wrapperFields.scores.freeze(own, checkConsistency);
    this._refObject.scores = this._wrapperFields.scores.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.traj.copyRefObject(refObject.traj);
    this._wrapperFields.scores.copyRefObject(refObject.scores);
    this._totalIntialized = true;
    this._refObject.total = refObject.total;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    Trajectory2DWrapper.freeStruct(refObject.traj);
    if (refObject.scores.size != 0) {
      CriticScoreWrapper.ArrayType.freeArray(refObject.scores);
      if (CriticScoreWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.scores, CriticScoreWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    TrajectoryScoreWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'dwb_msgs', subFolder: 'msg', interfaceName: 'TrajectoryScore'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get traj() {
    return this._wrapperFields.traj;
  }
  set traj(value) {
    if (value instanceof Trajectory2DWrapper) {
      this._wrapperFields.traj.copy(value);
    } else {
      this._wrapperFields.traj.copy(new Trajectory2DWrapper(value));
    }
  }
  get scores() {
    return this._wrapperFields.scores;
  }
  set scores(value) {
    this._wrapperFields.scores.fill(value);
  }
  get total() {
    if (!this._totalIntialized) {
      return undefined;
    }
    return this._refObject.total;
  }
  set total(value) {
    this._totalIntialized = true;
    this._refObject.total = value;
  }
  copyRefObject(refObject) {
    this._refObject = new TrajectoryScoreRefStruct(refObject.toObject());
    this._wrapperFields.traj.copyRefObject(this._refObject.traj);
    this._wrapperFields.scores.copyRefObject(this._refObject.scores);
    this._totalIntialized = true;
  }
  copy(other) {
    this._refObject = new TrajectoryScoreRefStruct(other._refObject.toObject());
    this._wrapperFields.traj.copy(other._wrapperFields.traj);
    this._wrapperFields.scores.copy(other._wrapperFields.scores);
    this._totalIntialized = true;
  }
  static get classType() {
    return TrajectoryScoreWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"traj","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"dwb_msgs","type":"Trajectory2D","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"scores","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"dwb_msgs","type":"CriticScore","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"total","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"dwb_msgs","type":"TrajectoryScore","stringUpperBound":null,"isPrimitiveType":false},"msgName":"TrajectoryScore"};
  }
  hasMember(name) {
    let memberNames = ["traj","scores","total"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class TrajectoryScoreArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof TrajectoryScoreWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new TrajectoryScoreWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TrajectoryScoreArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TrajectoryScoreArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new TrajectoryScoreRefArray(size);
    this._refObject = new TrajectoryScoreRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new TrajectoryScoreWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof TrajectoryScoreArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "TrajectoryScoreArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      TrajectoryScoreWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return TrajectoryScoreWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return TrajectoryScoreArrayWrapper;
  }
}
module.exports = TrajectoryScoreWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "traj",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "dwb_msgs",
        "type": "Trajectory2D",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "scores",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "dwb_msgs",
        "type": "CriticScore",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "total",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "dwb_msgs",
    "type": "TrajectoryScore",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "TrajectoryScore"
}
*/
