// rclnodejs/generated/dwb_msgs/dwb_msgs__msg__Trajectory2D.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Twist2DWrapper = require('../../generated/nav_2d_msgs/nav_2d_msgs__msg__Twist2D.js');
const DurationWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Duration.js');
const Pose2DWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Pose2D.js');
const Trajectory2DRefStruct = StructType({
  velocity: Twist2DWrapper.refObjectType,
  time_offsets: DurationWrapper.refObjectArrayType,
  poses: Pose2DWrapper.refObjectArrayType,
});
const Trajectory2DRefArray = ArrayType(Trajectory2DRefStruct);
const Trajectory2DRefStructArray = StructType({
  data: Trajectory2DRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class Trajectory2DWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new Trajectory2DRefStruct(other._refObject.toObject());
      this._wrapperFields.velocity =  new Twist2DWrapper(other._wrapperFields.velocity);
      this._wrapperFields.time_offsets = DurationWrapper.createArray();
      this._wrapperFields.time_offsets.copy(other._wrapperFields.time_offsets);
      this._wrapperFields.poses = Pose2DWrapper.createArray();
      this._wrapperFields.poses.copy(other._wrapperFields.poses);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new Trajectory2DRefStruct();
    this._wrapperFields.velocity = new Twist2DWrapper();
    this._wrapperFields.time_offsets = DurationWrapper.createArray();
    this._wrapperFields.poses = Pose2DWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new Trajectory2DWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new Trajectory2DArrayWrapper;
  }
  static get ArrayType() {
    return Trajectory2DArrayWrapper;
  }
  static get refObjectArrayType() {
    return Trajectory2DRefStructArray
  }
  static get refObjectType() {
    return Trajectory2DRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.velocity.freeze(own, checkConsistency);
    this._refObject.velocity = this._wrapperFields.velocity.refObject;
    this._wrapperFields.time_offsets.freeze(own, checkConsistency);
    this._refObject.time_offsets = this._wrapperFields.time_offsets.refObject;
    this._wrapperFields.poses.freeze(own, checkConsistency);
    this._refObject.poses = this._wrapperFields.poses.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.velocity.copyRefObject(refObject.velocity);
    this._wrapperFields.time_offsets.copyRefObject(refObject.time_offsets);
    this._wrapperFields.poses.copyRefObject(refObject.poses);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    Twist2DWrapper.freeStruct(refObject.velocity);
    if (refObject.time_offsets.size != 0) {
      DurationWrapper.ArrayType.freeArray(refObject.time_offsets);
      if (DurationWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.time_offsets, DurationWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.poses.size != 0) {
      Pose2DWrapper.ArrayType.freeArray(refObject.poses);
      if (Pose2DWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.poses, Pose2DWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    Trajectory2DWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'dwb_msgs', subFolder: 'msg', interfaceName: 'Trajectory2D'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get velocity() {
    return this._wrapperFields.velocity;
  }
  set velocity(value) {
    if (value instanceof Twist2DWrapper) {
      this._wrapperFields.velocity.copy(value);
    } else {
      this._wrapperFields.velocity.copy(new Twist2DWrapper(value));
    }
  }
  get time_offsets() {
    return this._wrapperFields.time_offsets;
  }
  set time_offsets(value) {
    this._wrapperFields.time_offsets.fill(value);
  }
  get poses() {
    return this._wrapperFields.poses;
  }
  set poses(value) {
    this._wrapperFields.poses.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new Trajectory2DRefStruct(refObject.toObject());
    this._wrapperFields.velocity.copyRefObject(this._refObject.velocity);
    this._wrapperFields.time_offsets.copyRefObject(this._refObject.time_offsets);
    this._wrapperFields.poses.copyRefObject(this._refObject.poses);
  }
  copy(other) {
    this._refObject = new Trajectory2DRefStruct(other._refObject.toObject());
    this._wrapperFields.velocity.copy(other._wrapperFields.velocity);
    this._wrapperFields.time_offsets.copy(other._wrapperFields.time_offsets);
    this._wrapperFields.poses.copy(other._wrapperFields.poses);
  }
  static get classType() {
    return Trajectory2DWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"velocity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"nav_2d_msgs","type":"Twist2D","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"time_offsets","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"builtin_interfaces","type":"Duration","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"poses","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"geometry_msgs","type":"Pose2D","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"dwb_msgs","type":"Trajectory2D","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Trajectory2D"};
  }
  hasMember(name) {
    let memberNames = ["velocity","time_offsets","poses"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class Trajectory2DArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof Trajectory2DWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new Trajectory2DWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Trajectory2DArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Trajectory2DArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new Trajectory2DRefArray(size);
    this._refObject = new Trajectory2DRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new Trajectory2DWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof Trajectory2DArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "Trajectory2DArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      Trajectory2DWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return Trajectory2DWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return Trajectory2DArrayWrapper;
  }
}
module.exports = Trajectory2DWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "velocity",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "nav_2d_msgs",
        "type": "Twist2D",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "time_offsets",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "builtin_interfaces",
        "type": "Duration",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "poses",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "geometry_msgs",
        "type": "Pose2D",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "dwb_msgs",
    "type": "Trajectory2D",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Trajectory2D"
}
*/
