// rclnodejs/generated/dwb_msgs/dwb_msgs__srv__ScoreTrajectory_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TrajectoryScoreWrapper = require('../../generated/dwb_msgs/dwb_msgs__msg__TrajectoryScore.js');
const ScoreTrajectory_ResponseRefStruct = StructType({
  score: TrajectoryScoreWrapper.refObjectType,
});
const ScoreTrajectory_ResponseRefArray = ArrayType(ScoreTrajectory_ResponseRefStruct);
const ScoreTrajectory_ResponseRefStructArray = StructType({
  data: ScoreTrajectory_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ScoreTrajectory_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ScoreTrajectory_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.score =  new TrajectoryScoreWrapper(other._wrapperFields.score);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ScoreTrajectory_ResponseRefStruct();
    this._wrapperFields.score = new TrajectoryScoreWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ScoreTrajectory_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ScoreTrajectory_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return ScoreTrajectory_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return ScoreTrajectory_ResponseRefStructArray
  }
  static get refObjectType() {
    return ScoreTrajectory_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.score.freeze(own, checkConsistency);
    this._refObject.score = this._wrapperFields.score.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.score.copyRefObject(refObject.score);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    TrajectoryScoreWrapper.freeStruct(refObject.score);
  }
  static destoryRawROS(msg) {
    ScoreTrajectory_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'dwb_msgs', subFolder: 'srv', interfaceName: 'ScoreTrajectory_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get score() {
    return this._wrapperFields.score;
  }
  set score(value) {
    if (value instanceof TrajectoryScoreWrapper) {
      this._wrapperFields.score.copy(value);
    } else {
      this._wrapperFields.score.copy(new TrajectoryScoreWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new ScoreTrajectory_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.score.copyRefObject(this._refObject.score);
  }
  copy(other) {
    this._refObject = new ScoreTrajectory_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.score.copy(other._wrapperFields.score);
  }
  static get classType() {
    return ScoreTrajectory_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"score","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"dwb_msgs","type":"TrajectoryScore","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"dwb_msgs","type":"ScoreTrajectory_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ScoreTrajectory_Response"};
  }
  hasMember(name) {
    let memberNames = ["score"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ScoreTrajectory_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ScoreTrajectory_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ScoreTrajectory_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ScoreTrajectory_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ScoreTrajectory_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ScoreTrajectory_ResponseRefArray(size);
    this._refObject = new ScoreTrajectory_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ScoreTrajectory_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ScoreTrajectory_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ScoreTrajectory_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ScoreTrajectory_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ScoreTrajectory_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ScoreTrajectory_ResponseArrayWrapper;
  }
}
module.exports = ScoreTrajectory_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "score",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "dwb_msgs",
        "type": "TrajectoryScore",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "dwb_msgs",
    "type": "ScoreTrajectory_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ScoreTrajectory_Response"
}
*/
