// rclnodejs/generated/dwb_msgs/dwb_msgs__msg__LocalPlanEvaluation.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const TrajectoryScoreWrapper = require('../../generated/dwb_msgs/dwb_msgs__msg__TrajectoryScore.js');
const LocalPlanEvaluationRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  twists: TrajectoryScoreWrapper.refObjectArrayType,
  best_index: primitiveTypes.uint16,
  worst_index: primitiveTypes.uint16,
});
const LocalPlanEvaluationRefArray = ArrayType(LocalPlanEvaluationRefStruct);
const LocalPlanEvaluationRefStructArray = StructType({
  data: LocalPlanEvaluationRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class LocalPlanEvaluationWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new LocalPlanEvaluationRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.twists = TrajectoryScoreWrapper.createArray();
      this._wrapperFields.twists.copy(other._wrapperFields.twists);
      this._best_indexIntialized = true;
      this._worst_indexIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new LocalPlanEvaluationRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.twists = TrajectoryScoreWrapper.createArray();
    this._best_indexIntialized = false;
    this._worst_indexIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new LocalPlanEvaluationWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LocalPlanEvaluationArrayWrapper;
  }
  static get ArrayType() {
    return LocalPlanEvaluationArrayWrapper;
  }
  static get refObjectArrayType() {
    return LocalPlanEvaluationRefStructArray
  }
  static get refObjectType() {
    return LocalPlanEvaluationRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._best_indexIntialized) {
        throw new TypeError('Invalid argument: best_index in LocalPlanEvaluation');
      }
      if (!this._worst_indexIntialized) {
        throw new TypeError('Invalid argument: worst_index in LocalPlanEvaluation');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.twists.freeze(own, checkConsistency);
    this._refObject.twists = this._wrapperFields.twists.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.twists.copyRefObject(refObject.twists);
    this._best_indexIntialized = true;
    this._refObject.best_index = refObject.best_index;
    this._worst_indexIntialized = true;
    this._refObject.worst_index = refObject.worst_index;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.twists.size != 0) {
      TrajectoryScoreWrapper.ArrayType.freeArray(refObject.twists);
      if (TrajectoryScoreWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.twists, TrajectoryScoreWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    LocalPlanEvaluationWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'dwb_msgs', subFolder: 'msg', interfaceName: 'LocalPlanEvaluation'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get twists() {
    return this._wrapperFields.twists;
  }
  set twists(value) {
    this._wrapperFields.twists.fill(value);
  }
  get best_index() {
    if (!this._best_indexIntialized) {
      return undefined;
    }
    return this._refObject.best_index;
  }
  set best_index(value) {
    this._best_indexIntialized = true;
    this._refObject.best_index = value;
  }
  get worst_index() {
    if (!this._worst_indexIntialized) {
      return undefined;
    }
    return this._refObject.worst_index;
  }
  set worst_index(value) {
    this._worst_indexIntialized = true;
    this._refObject.worst_index = value;
  }
  copyRefObject(refObject) {
    this._refObject = new LocalPlanEvaluationRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.twists.copyRefObject(this._refObject.twists);
    this._best_indexIntialized = true;
    this._worst_indexIntialized = true;
  }
  copy(other) {
    this._refObject = new LocalPlanEvaluationRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.twists.copy(other._wrapperFields.twists);
    this._best_indexIntialized = true;
    this._worst_indexIntialized = true;
  }
  static get classType() {
    return LocalPlanEvaluationWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"twists","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"dwb_msgs","type":"TrajectoryScore","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"best_index","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"worst_index","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"dwb_msgs","type":"LocalPlanEvaluation","stringUpperBound":null,"isPrimitiveType":false},"msgName":"LocalPlanEvaluation"};
  }
  hasMember(name) {
    let memberNames = ["header","twists","best_index","worst_index"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class LocalPlanEvaluationArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LocalPlanEvaluationWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LocalPlanEvaluationWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LocalPlanEvaluationArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LocalPlanEvaluationArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new LocalPlanEvaluationRefArray(size);
    this._refObject = new LocalPlanEvaluationRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LocalPlanEvaluationWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LocalPlanEvaluationArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LocalPlanEvaluationArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LocalPlanEvaluationWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LocalPlanEvaluationWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LocalPlanEvaluationArrayWrapper;
  }
}
module.exports = LocalPlanEvaluationWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "twists",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "dwb_msgs",
        "type": "TrajectoryScore",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "best_index",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "worst_index",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "dwb_msgs",
    "type": "LocalPlanEvaluation",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "LocalPlanEvaluation"
}
*/
