// rclnodejs/generated/turtlebot3_msgs/turtlebot3_msgs__msg__VersionInfo.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const VersionInfoRefStruct = StructType({
  hardware: primitiveTypes.string,
  firmware: primitiveTypes.string,
  software: primitiveTypes.string,
});
const VersionInfoRefArray = ArrayType(VersionInfoRefStruct);
const VersionInfoRefStructArray = StructType({
  data: VersionInfoRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class VersionInfoWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new VersionInfoRefStruct(other._refObject.toObject());
      this._hardwareIntialized = true;
      this._wrapperFields.hardware =  new StringWrapper(other._wrapperFields.hardware);
      this._firmwareIntialized = true;
      this._wrapperFields.firmware =  new StringWrapper(other._wrapperFields.firmware);
      this._softwareIntialized = true;
      this._wrapperFields.software =  new StringWrapper(other._wrapperFields.software);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new VersionInfoRefStruct();
    this._hardwareIntialized = false;
    this._wrapperFields.hardware = new StringWrapper();
    this._firmwareIntialized = false;
    this._wrapperFields.firmware = new StringWrapper();
    this._softwareIntialized = false;
    this._wrapperFields.software = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new VersionInfoWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new VersionInfoArrayWrapper;
  }
  static get ArrayType() {
    return VersionInfoArrayWrapper;
  }
  static get refObjectArrayType() {
    return VersionInfoRefStructArray
  }
  static get refObjectType() {
    return VersionInfoRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._hardwareIntialized) {
        throw new TypeError('Invalid argument: hardware in VersionInfo');
      }
      if (!this._firmwareIntialized) {
        throw new TypeError('Invalid argument: firmware in VersionInfo');
      }
      if (!this._softwareIntialized) {
        throw new TypeError('Invalid argument: software in VersionInfo');
      }
    }
    if (own) {
      this._wrapperFields.hardware.freeze(own, checkConsistency);
    }
    this._refObject.hardware = this._wrapperFields.hardware.refObject;
    if (own) {
      this._wrapperFields.firmware.freeze(own, checkConsistency);
    }
    this._refObject.firmware = this._wrapperFields.firmware.refObject;
    if (own) {
      this._wrapperFields.software.freeze(own, checkConsistency);
    }
    this._refObject.software = this._wrapperFields.software.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._hardwareIntialized = true;
    this._wrapperFields.hardware.data = refObject.hardware.data;
    this._firmwareIntialized = true;
    this._wrapperFields.firmware.data = refObject.firmware.data;
    this._softwareIntialized = true;
    this._wrapperFields.software.data = refObject.software.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.hardware);
    StringWrapper.freeStruct(refObject.firmware);
    StringWrapper.freeStruct(refObject.software);
  }
  static destoryRawROS(msg) {
    VersionInfoWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'turtlebot3_msgs', subFolder: 'msg', interfaceName: 'VersionInfo'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get hardware() {
    if (!this._hardwareIntialized) {
      return undefined;
    }
    return this._wrapperFields.hardware.data;
  }
  set hardware(value) {
    this._hardwareIntialized = true;
    this._wrapperFields.hardware.data = value;
  }
  get firmware() {
    if (!this._firmwareIntialized) {
      return undefined;
    }
    return this._wrapperFields.firmware.data;
  }
  set firmware(value) {
    this._firmwareIntialized = true;
    this._wrapperFields.firmware.data = value;
  }
  get software() {
    if (!this._softwareIntialized) {
      return undefined;
    }
    return this._wrapperFields.software.data;
  }
  set software(value) {
    this._softwareIntialized = true;
    this._wrapperFields.software.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new VersionInfoRefStruct(refObject.toObject());
    this._hardwareIntialized = true;
    this._wrapperFields.hardware.copyRefObject(this._refObject.hardware);
    this._firmwareIntialized = true;
    this._wrapperFields.firmware.copyRefObject(this._refObject.firmware);
    this._softwareIntialized = true;
    this._wrapperFields.software.copyRefObject(this._refObject.software);
  }
  copy(other) {
    this._refObject = new VersionInfoRefStruct(other._refObject.toObject());
    this._hardwareIntialized = true;
    this._wrapperFields.hardware.copy(other._wrapperFields.hardware);
    this._firmwareIntialized = true;
    this._wrapperFields.firmware.copy(other._wrapperFields.firmware);
    this._softwareIntialized = true;
    this._wrapperFields.software.copy(other._wrapperFields.software);
  }
  static get classType() {
    return VersionInfoWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"hardware","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"firmware","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"software","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"turtlebot3_msgs","type":"VersionInfo","stringUpperBound":null,"isPrimitiveType":false},"msgName":"VersionInfo"};
  }
  hasMember(name) {
    let memberNames = ["hardware","firmware","software"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class VersionInfoArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof VersionInfoWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new VersionInfoWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VersionInfoArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VersionInfoArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new VersionInfoRefArray(size);
    this._refObject = new VersionInfoRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new VersionInfoWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof VersionInfoArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "VersionInfoArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      VersionInfoWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return VersionInfoWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return VersionInfoArrayWrapper;
  }
}
module.exports = VersionInfoWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "hardware",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "firmware",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "software",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "turtlebot3_msgs",
    "type": "VersionInfo",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "VersionInfo"
}
*/
