// rclnodejs/generated/turtlebot3_msgs/turtlebot3_msgs__action__Patrol.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
class turtlebot3_msgs__action__Patrol {
  static get Goal() {
    return require('./turtlebot3_msgs__action__Patrol_Goal.js');
  }
  static get Result() {
    return require('./turtlebot3_msgs__action__Patrol_Result.js');
  }
  static get Feedback() {
    return require('./turtlebot3_msgs__action__Patrol_Feedback.js');
  }
  static get impl() {
    return {
      get SendGoalService() {
        return require('./turtlebot3_msgs__action__Patrol_SendGoal.js');
      },
      get GetResultService() {
        return require('./turtlebot3_msgs__action__Patrol_GetResult.js');
      },
      get FeedbackMessage() {
        return require('./turtlebot3_msgs__action__Patrol_FeedbackMessage.js');
      },
      get CancelGoal() {
        return require('../action_msgs/action_msgs__srv__CancelGoal.js');
      },
      get GoalStatusArray() {
        return require('../action_msgs/action_msgs__msg__GoalStatusArray.js');
      }
    };
  }
  static type() {
    return {pkgName: 'turtlebot3_msgs', subFolder: 'action', interfaceName: 'Patrol'};
  }
}
module.exports = turtlebot3_msgs__action__Patrol;
