// rclnodejs/generated/turtlebot3_msgs/turtlebot3_msgs__srv__Dqn_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Dqn_RequestRefStruct = StructType({
  action: primitiveTypes.uint8,
  init: primitiveTypes.bool,
});
const Dqn_RequestRefArray = ArrayType(Dqn_RequestRefStruct);
const Dqn_RequestRefStructArray = StructType({
  data: Dqn_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class Dqn_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new Dqn_RequestRefStruct(other._refObject.toObject());
      this._actionIntialized = true;
      this._initIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new Dqn_RequestRefStruct();
    this._actionIntialized = false;
    this._initIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new Dqn_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new Dqn_RequestArrayWrapper;
  }
  static get ArrayType() {
    return Dqn_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return Dqn_RequestRefStructArray
  }
  static get refObjectType() {
    return Dqn_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._actionIntialized) {
        throw new TypeError('Invalid argument: action in Dqn_Request');
      }
      if (!this._initIntialized) {
        throw new TypeError('Invalid argument: init in Dqn_Request');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._actionIntialized = true;
    this._refObject.action = refObject.action;
    this._initIntialized = true;
    this._refObject.init = refObject.init;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    Dqn_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'turtlebot3_msgs', subFolder: 'srv', interfaceName: 'Dqn_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get action() {
    if (!this._actionIntialized) {
      return undefined;
    }
    return this._refObject.action;
  }
  set action(value) {
    this._actionIntialized = true;
    this._refObject.action = value;
  }
  get init() {
    if (!this._initIntialized) {
      return undefined;
    }
    return this._refObject.init;
  }
  set init(value) {
    this._initIntialized = true;
    this._refObject.init = value;
  }
  copyRefObject(refObject) {
    this._refObject = new Dqn_RequestRefStruct(refObject.toObject());
    this._actionIntialized = true;
    this._initIntialized = true;
  }
  copy(other) {
    this._refObject = new Dqn_RequestRefStruct(other._refObject.toObject());
    this._actionIntialized = true;
    this._initIntialized = true;
  }
  static get classType() {
    return Dqn_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"action","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"init","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"turtlebot3_msgs","type":"Dqn_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Dqn_Request"};
  }
  hasMember(name) {
    let memberNames = ["action","init"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class Dqn_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof Dqn_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new Dqn_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Dqn_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Dqn_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new Dqn_RequestRefArray(size);
    this._refObject = new Dqn_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new Dqn_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof Dqn_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "Dqn_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      Dqn_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return Dqn_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return Dqn_RequestArrayWrapper;
  }
}
module.exports = Dqn_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "action",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "init",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "turtlebot3_msgs",
    "type": "Dqn_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Dqn_Request"
}
*/
