// rclnodejs/generated/turtlebot3_msgs/turtlebot3_msgs__srv__Dqn_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const Dqn_ResponseRefStruct = StructType({
  state: Float32Wrapper.refObjectArrayType,
  reward: primitiveTypes.float32,
  done: primitiveTypes.bool,
});
const Dqn_ResponseRefArray = ArrayType(Dqn_ResponseRefStruct);
const Dqn_ResponseRefStructArray = StructType({
  data: Dqn_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class Dqn_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new Dqn_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.state = Float32Wrapper.createArray();
      this._wrapperFields.state.copy(other._wrapperFields.state);
      this._rewardIntialized = true;
      this._doneIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new Dqn_ResponseRefStruct();
    this._wrapperFields.state = Float32Wrapper.createArray();
    this._rewardIntialized = false;
    this._doneIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new Dqn_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new Dqn_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return Dqn_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return Dqn_ResponseRefStructArray
  }
  static get refObjectType() {
    return Dqn_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._rewardIntialized) {
        throw new TypeError('Invalid argument: reward in Dqn_Response');
      }
      if (!this._doneIntialized) {
        throw new TypeError('Invalid argument: done in Dqn_Response');
      }
    }
    this._wrapperFields.state.freeze(own, checkConsistency);
    this._refObject.state = this._wrapperFields.state.refObject;
    if (own) {
      this._wrapperFields.state.fill(Float32Array.from([]));
      this._wrapperFields.state.freeze(own, checkConsistency);
      this._refObject.state = this._wrapperFields.state.refObject;
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.state.copyRefObject(refObject.state);
    this._rewardIntialized = true;
    this._refObject.reward = refObject.reward;
    this._doneIntialized = true;
    this._refObject.done = refObject.done;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.state.size != 0) {
      Float32Wrapper.ArrayType.freeArray(refObject.state);
      if (Float32Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.state, Float32Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    Dqn_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'turtlebot3_msgs', subFolder: 'srv', interfaceName: 'Dqn_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get state() {
    return this._wrapperFields['state'].data;
  }
  set state(value) {
    this._wrapperFields['state'].fill(value);
  }
  get reward() {
    if (!this._rewardIntialized) {
      return undefined;
    }
    return this._refObject.reward;
  }
  set reward(value) {
    this._rewardIntialized = true;
    this._refObject.reward = value;
  }
  get done() {
    if (!this._doneIntialized) {
      return undefined;
    }
    return this._refObject.done;
  }
  set done(value) {
    this._doneIntialized = true;
    this._refObject.done = value;
  }
  copyRefObject(refObject) {
    this._refObject = new Dqn_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.state.copyRefObject(this._refObject.state);
    this._rewardIntialized = true;
    this._doneIntialized = true;
  }
  copy(other) {
    this._refObject = new Dqn_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.state.copy(other._wrapperFields.state);
    this._rewardIntialized = true;
    this._doneIntialized = true;
  }
  static get classType() {
    return Dqn_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"state","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reward","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"done","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"turtlebot3_msgs","type":"Dqn_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Dqn_Response"};
  }
  hasMember(name) {
    let memberNames = ["state","reward","done"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class Dqn_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof Dqn_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new Dqn_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Dqn_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Dqn_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new Dqn_ResponseRefArray(size);
    this._refObject = new Dqn_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new Dqn_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof Dqn_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "Dqn_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      Dqn_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return Dqn_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return Dqn_ResponseArrayWrapper;
  }
}
module.exports = Dqn_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "state",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reward",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "done",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "turtlebot3_msgs",
    "type": "Dqn_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Dqn_Response"
}
*/
