// rclnodejs/generated/cartographer_ros_msgs/cartographer_ros_msgs__srv__StartTrajectory_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TrajectoryOptionsWrapper = require('../../generated/cartographer_ros_msgs/cartographer_ros_msgs__msg__TrajectoryOptions.js');
const SensorTopicsWrapper = require('../../generated/cartographer_ros_msgs/cartographer_ros_msgs__msg__SensorTopics.js');
const StartTrajectory_RequestRefStruct = StructType({
  options: TrajectoryOptionsWrapper.refObjectType,
  topics: SensorTopicsWrapper.refObjectType,
});
const StartTrajectory_RequestRefArray = ArrayType(StartTrajectory_RequestRefStruct);
const StartTrajectory_RequestRefStructArray = StructType({
  data: StartTrajectory_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class StartTrajectory_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new StartTrajectory_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.options =  new TrajectoryOptionsWrapper(other._wrapperFields.options);
      this._wrapperFields.topics =  new SensorTopicsWrapper(other._wrapperFields.topics);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new StartTrajectory_RequestRefStruct();
    this._wrapperFields.options = new TrajectoryOptionsWrapper();
    this._wrapperFields.topics = new SensorTopicsWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new StartTrajectory_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new StartTrajectory_RequestArrayWrapper;
  }
  static get ArrayType() {
    return StartTrajectory_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return StartTrajectory_RequestRefStructArray
  }
  static get refObjectType() {
    return StartTrajectory_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.options.freeze(own, checkConsistency);
    this._refObject.options = this._wrapperFields.options.refObject;
    this._wrapperFields.topics.freeze(own, checkConsistency);
    this._refObject.topics = this._wrapperFields.topics.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.options.copyRefObject(refObject.options);
    this._wrapperFields.topics.copyRefObject(refObject.topics);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    TrajectoryOptionsWrapper.freeStruct(refObject.options);
    SensorTopicsWrapper.freeStruct(refObject.topics);
  }
  static destoryRawROS(msg) {
    StartTrajectory_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'cartographer_ros_msgs', subFolder: 'srv', interfaceName: 'StartTrajectory_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get options() {
    return this._wrapperFields.options;
  }
  set options(value) {
    if (value instanceof TrajectoryOptionsWrapper) {
      this._wrapperFields.options.copy(value);
    } else {
      this._wrapperFields.options.copy(new TrajectoryOptionsWrapper(value));
    }
  }
  get topics() {
    return this._wrapperFields.topics;
  }
  set topics(value) {
    if (value instanceof SensorTopicsWrapper) {
      this._wrapperFields.topics.copy(value);
    } else {
      this._wrapperFields.topics.copy(new SensorTopicsWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new StartTrajectory_RequestRefStruct(refObject.toObject());
    this._wrapperFields.options.copyRefObject(this._refObject.options);
    this._wrapperFields.topics.copyRefObject(this._refObject.topics);
  }
  copy(other) {
    this._refObject = new StartTrajectory_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.options.copy(other._wrapperFields.options);
    this._wrapperFields.topics.copy(other._wrapperFields.topics);
  }
  static get classType() {
    return StartTrajectory_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"options","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"cartographer_ros_msgs","type":"TrajectoryOptions","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"topics","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"cartographer_ros_msgs","type":"SensorTopics","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"cartographer_ros_msgs","type":"StartTrajectory_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"StartTrajectory_Request"};
  }
  hasMember(name) {
    let memberNames = ["options","topics"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class StartTrajectory_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof StartTrajectory_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new StartTrajectory_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to StartTrajectory_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to StartTrajectory_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new StartTrajectory_RequestRefArray(size);
    this._refObject = new StartTrajectory_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new StartTrajectory_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof StartTrajectory_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "StartTrajectory_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      StartTrajectory_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return StartTrajectory_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return StartTrajectory_RequestArrayWrapper;
  }
}
module.exports = StartTrajectory_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "options",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "cartographer_ros_msgs",
        "type": "TrajectoryOptions",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "topics",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "cartographer_ros_msgs",
        "type": "SensorTopics",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "cartographer_ros_msgs",
    "type": "StartTrajectory_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "StartTrajectory_Request"
}
*/
