// rclnodejs/generated/cartographer_ros_msgs/cartographer_ros_msgs__msg__SubmapTexture.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const UInt8Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt8.js');
const PoseWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Pose.js');
const SubmapTextureRefStruct = StructType({
  cells: UInt8Wrapper.refObjectArrayType,
  width: primitiveTypes.int32,
  height: primitiveTypes.int32,
  resolution: primitiveTypes.float64,
  slice_pose: PoseWrapper.refObjectType,
});
const SubmapTextureRefArray = ArrayType(SubmapTextureRefStruct);
const SubmapTextureRefStructArray = StructType({
  data: SubmapTextureRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SubmapTextureWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SubmapTextureRefStruct(other._refObject.toObject());
      this._wrapperFields.cells = UInt8Wrapper.createArray();
      this._wrapperFields.cells.copy(other._wrapperFields.cells);
      this._widthIntialized = true;
      this._heightIntialized = true;
      this._resolutionIntialized = true;
      this._wrapperFields.slice_pose =  new PoseWrapper(other._wrapperFields.slice_pose);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SubmapTextureRefStruct();
    this._wrapperFields.cells = UInt8Wrapper.createArray();
    this._widthIntialized = false;
    this._heightIntialized = false;
    this._resolutionIntialized = false;
    this._wrapperFields.slice_pose = new PoseWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new SubmapTextureWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SubmapTextureArrayWrapper;
  }
  static get ArrayType() {
    return SubmapTextureArrayWrapper;
  }
  static get refObjectArrayType() {
    return SubmapTextureRefStructArray
  }
  static get refObjectType() {
    return SubmapTextureRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._widthIntialized) {
        throw new TypeError('Invalid argument: width in SubmapTexture');
      }
      if (!this._heightIntialized) {
        throw new TypeError('Invalid argument: height in SubmapTexture');
      }
      if (!this._resolutionIntialized) {
        throw new TypeError('Invalid argument: resolution in SubmapTexture');
      }
    }
    this._wrapperFields.cells.freeze(own, checkConsistency);
    this._refObject.cells = this._wrapperFields.cells.refObject;
    if (own) {
      this._wrapperFields.cells.fill(Uint8Array.from([]));
      this._wrapperFields.cells.freeze(own, checkConsistency);
      this._refObject.cells = this._wrapperFields.cells.refObject;
    }
    this._wrapperFields.slice_pose.freeze(own, checkConsistency);
    this._refObject.slice_pose = this._wrapperFields.slice_pose.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.cells.copyRefObject(refObject.cells);
    this._widthIntialized = true;
    this._refObject.width = refObject.width;
    this._heightIntialized = true;
    this._refObject.height = refObject.height;
    this._resolutionIntialized = true;
    this._refObject.resolution = refObject.resolution;
    this._wrapperFields.slice_pose.copyRefObject(refObject.slice_pose);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.cells.size != 0) {
      UInt8Wrapper.ArrayType.freeArray(refObject.cells);
      if (UInt8Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.cells, UInt8Wrapper.refObjectArrayType, 'data');
      }
    }
    PoseWrapper.freeStruct(refObject.slice_pose);
  }
  static destoryRawROS(msg) {
    SubmapTextureWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'cartographer_ros_msgs', subFolder: 'msg', interfaceName: 'SubmapTexture'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get cells() {
    return this._wrapperFields['cells'].data;
  }
  set cells(value) {
    this._wrapperFields['cells'].fill(value);
  }
  get width() {
    if (!this._widthIntialized) {
      return undefined;
    }
    return this._refObject.width;
  }
  set width(value) {
    this._widthIntialized = true;
    this._refObject.width = value;
  }
  get height() {
    if (!this._heightIntialized) {
      return undefined;
    }
    return this._refObject.height;
  }
  set height(value) {
    this._heightIntialized = true;
    this._refObject.height = value;
  }
  get resolution() {
    if (!this._resolutionIntialized) {
      return undefined;
    }
    return this._refObject.resolution;
  }
  set resolution(value) {
    this._resolutionIntialized = true;
    this._refObject.resolution = value;
  }
  get slice_pose() {
    return this._wrapperFields.slice_pose;
  }
  set slice_pose(value) {
    if (value instanceof PoseWrapper) {
      this._wrapperFields.slice_pose.copy(value);
    } else {
      this._wrapperFields.slice_pose.copy(new PoseWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new SubmapTextureRefStruct(refObject.toObject());
    this._wrapperFields.cells.copyRefObject(this._refObject.cells);
    this._widthIntialized = true;
    this._heightIntialized = true;
    this._resolutionIntialized = true;
    this._wrapperFields.slice_pose.copyRefObject(this._refObject.slice_pose);
  }
  copy(other) {
    this._refObject = new SubmapTextureRefStruct(other._refObject.toObject());
    this._wrapperFields.cells.copy(other._wrapperFields.cells);
    this._widthIntialized = true;
    this._heightIntialized = true;
    this._resolutionIntialized = true;
    this._wrapperFields.slice_pose.copy(other._wrapperFields.slice_pose);
  }
  static get classType() {
    return SubmapTextureWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"cells","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"width","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"height","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"resolution","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"slice_pose","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Pose","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"cartographer_ros_msgs","type":"SubmapTexture","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SubmapTexture"};
  }
  hasMember(name) {
    let memberNames = ["cells","width","height","resolution","slice_pose"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SubmapTextureArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SubmapTextureWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SubmapTextureWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SubmapTextureArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SubmapTextureArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SubmapTextureRefArray(size);
    this._refObject = new SubmapTextureRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SubmapTextureWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SubmapTextureArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SubmapTextureArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SubmapTextureWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SubmapTextureWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SubmapTextureArrayWrapper;
  }
}
module.exports = SubmapTextureWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "cells",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "width",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "height",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "resolution",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "slice_pose",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Pose",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "cartographer_ros_msgs",
    "type": "SubmapTexture",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SubmapTexture"
}
*/
