// rclnodejs/generated/cartographer_ros_msgs/cartographer_ros_msgs__msg__StatusCode.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StatusCodeRefStruct = StructType({
  _dummy: primitiveTypes.bool,
});
const StatusCodeRefArray = ArrayType(StatusCodeRefStruct);
const StatusCodeRefStructArray = StructType({
  data: StatusCodeRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class StatusCodeWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new StatusCodeRefStruct(other._refObject.toObject());
      this.__dummyIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new StatusCodeRefStruct();
    this.__dummyIntialized = true;
  }
  static createFromRefObject(refObject) {
    let self = new StatusCodeWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new StatusCodeArrayWrapper;
  }
  static get ArrayType() {
    return StatusCodeArrayWrapper;
  }
  static get refObjectArrayType() {
    return StatusCodeRefStructArray
  }
  static get refObjectType() {
    return StatusCodeRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this.__dummyIntialized) {
        throw new TypeError('Invalid argument: _dummy in StatusCode');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this.__dummyIntialized = true;
    this._refObject._dummy = refObject._dummy;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    StatusCodeWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'cartographer_ros_msgs', subFolder: 'msg', interfaceName: 'StatusCode'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get _dummy() {
    if (!this.__dummyIntialized) {
      return undefined;
    }
    return this._refObject._dummy;
  }
  set _dummy(value) {
    this.__dummyIntialized = true;
    this._refObject._dummy = value;
  }
  copyRefObject(refObject) {
    this._refObject = new StatusCodeRefStruct(refObject.toObject());
    this.__dummyIntialized = true;
  }
  copy(other) {
    this._refObject = new StatusCodeRefStruct(other._refObject.toObject());
    this.__dummyIntialized = true;
  }
  static get classType() {
    return StatusCodeWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"OK","value":0},{"type":"uint8","name":"CANCELLED","value":1},{"type":"uint8","name":"UNKNOWN","value":2},{"type":"uint8","name":"INVALID_ARGUMENT","value":3},{"type":"uint8","name":"DEADLINE_EXCEEDED","value":4},{"type":"uint8","name":"NOT_FOUND","value":5},{"type":"uint8","name":"ALREADY_EXISTS","value":6},{"type":"uint8","name":"PERMISSION_DENIED","value":7},{"type":"uint8","name":"RESOURCE_EXHAUSTED","value":8},{"type":"uint8","name":"FAILED_PRECONDITION","value":9},{"type":"uint8","name":"ABORTED","value":10},{"type":"uint8","name":"OUT_OF_RANGE","value":11},{"type":"uint8","name":"UNIMPLEMENTED","value":12},{"type":"uint8","name":"INTERNAL","value":13},{"type":"uint8","name":"UNAVAILABLE","value":14},{"type":"uint8","name":"DATA_LOSS","value":15}],"fields":[],"baseType":{"pkgName":"cartographer_ros_msgs","type":"StatusCode","stringUpperBound":null,"isPrimitiveType":false},"msgName":"StatusCode"};
  }
  hasMember(name) {
    let memberNames = ["_dummy"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class StatusCodeArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof StatusCodeWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new StatusCodeWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to StatusCodeArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to StatusCodeArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new StatusCodeRefArray(size);
    this._refObject = new StatusCodeRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new StatusCodeWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof StatusCodeArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "StatusCodeArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      StatusCodeWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return StatusCodeWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return StatusCodeArrayWrapper;
  }
}
// Define constants (16 in total)
Object.defineProperty(StatusCodeWrapper, "OK", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StatusCodeWrapper, "CANCELLED", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StatusCodeWrapper, "UNKNOWN", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StatusCodeWrapper, "INVALID_ARGUMENT", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StatusCodeWrapper, "DEADLINE_EXCEEDED", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StatusCodeWrapper, "NOT_FOUND", {value: 5, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StatusCodeWrapper, "ALREADY_EXISTS", {value: 6, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StatusCodeWrapper, "PERMISSION_DENIED", {value: 7, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StatusCodeWrapper, "RESOURCE_EXHAUSTED", {value: 8, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StatusCodeWrapper, "FAILED_PRECONDITION", {value: 9, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StatusCodeWrapper, "ABORTED", {value: 10, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StatusCodeWrapper, "OUT_OF_RANGE", {value: 11, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StatusCodeWrapper, "UNIMPLEMENTED", {value: 12, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StatusCodeWrapper, "INTERNAL", {value: 13, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StatusCodeWrapper, "UNAVAILABLE", {value: 14, writable: false, enumerable: true, configurable: true});
Object.defineProperty(StatusCodeWrapper, "DATA_LOSS", {value: 15, writable: false, enumerable: true, configurable: true});
module.exports = StatusCodeWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "OK",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "CANCELLED",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "UNKNOWN",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "INVALID_ARGUMENT",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "DEADLINE_EXCEEDED",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "NOT_FOUND",
      "value": 5
    },
    {
      "type": "uint8",
      "name": "ALREADY_EXISTS",
      "value": 6
    },
    {
      "type": "uint8",
      "name": "PERMISSION_DENIED",
      "value": 7
    },
    {
      "type": "uint8",
      "name": "RESOURCE_EXHAUSTED",
      "value": 8
    },
    {
      "type": "uint8",
      "name": "FAILED_PRECONDITION",
      "value": 9
    },
    {
      "type": "uint8",
      "name": "ABORTED",
      "value": 10
    },
    {
      "type": "uint8",
      "name": "OUT_OF_RANGE",
      "value": 11
    },
    {
      "type": "uint8",
      "name": "UNIMPLEMENTED",
      "value": 12
    },
    {
      "type": "uint8",
      "name": "INTERNAL",
      "value": 13
    },
    {
      "type": "uint8",
      "name": "UNAVAILABLE",
      "value": 14
    },
    {
      "type": "uint8",
      "name": "DATA_LOSS",
      "value": 15
    }
  ],
  "fields": [],
  "baseType": {
    "pkgName": "cartographer_ros_msgs",
    "type": "StatusCode",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "StatusCode"
}
*/
