// rclnodejs/generated/cartographer_ros_msgs/cartographer_ros_msgs__srv__SubmapQuery_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StatusResponseWrapper = require('../../generated/cartographer_ros_msgs/cartographer_ros_msgs__msg__StatusResponse.js');
const SubmapTextureWrapper = require('../../generated/cartographer_ros_msgs/cartographer_ros_msgs__msg__SubmapTexture.js');
const SubmapQuery_ResponseRefStruct = StructType({
  status: StatusResponseWrapper.refObjectType,
  submap_version: primitiveTypes.int32,
  textures: SubmapTextureWrapper.refObjectArrayType,
});
const SubmapQuery_ResponseRefArray = ArrayType(SubmapQuery_ResponseRefStruct);
const SubmapQuery_ResponseRefStructArray = StructType({
  data: SubmapQuery_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SubmapQuery_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SubmapQuery_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.status =  new StatusResponseWrapper(other._wrapperFields.status);
      this._submap_versionIntialized = true;
      this._wrapperFields.textures = SubmapTextureWrapper.createArray();
      this._wrapperFields.textures.copy(other._wrapperFields.textures);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SubmapQuery_ResponseRefStruct();
    this._wrapperFields.status = new StatusResponseWrapper();
    this._submap_versionIntialized = false;
    this._wrapperFields.textures = SubmapTextureWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new SubmapQuery_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SubmapQuery_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return SubmapQuery_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return SubmapQuery_ResponseRefStructArray
  }
  static get refObjectType() {
    return SubmapQuery_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._submap_versionIntialized) {
        throw new TypeError('Invalid argument: submap_version in SubmapQuery_Response');
      }
    }
    this._wrapperFields.status.freeze(own, checkConsistency);
    this._refObject.status = this._wrapperFields.status.refObject;
    this._wrapperFields.textures.freeze(own, checkConsistency);
    this._refObject.textures = this._wrapperFields.textures.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.status.copyRefObject(refObject.status);
    this._submap_versionIntialized = true;
    this._refObject.submap_version = refObject.submap_version;
    this._wrapperFields.textures.copyRefObject(refObject.textures);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StatusResponseWrapper.freeStruct(refObject.status);
    if (refObject.textures.size != 0) {
      SubmapTextureWrapper.ArrayType.freeArray(refObject.textures);
      if (SubmapTextureWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.textures, SubmapTextureWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    SubmapQuery_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'cartographer_ros_msgs', subFolder: 'srv', interfaceName: 'SubmapQuery_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get status() {
    return this._wrapperFields.status;
  }
  set status(value) {
    if (value instanceof StatusResponseWrapper) {
      this._wrapperFields.status.copy(value);
    } else {
      this._wrapperFields.status.copy(new StatusResponseWrapper(value));
    }
  }
  get submap_version() {
    if (!this._submap_versionIntialized) {
      return undefined;
    }
    return this._refObject.submap_version;
  }
  set submap_version(value) {
    this._submap_versionIntialized = true;
    this._refObject.submap_version = value;
  }
  get textures() {
    return this._wrapperFields.textures;
  }
  set textures(value) {
    this._wrapperFields.textures.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new SubmapQuery_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.status.copyRefObject(this._refObject.status);
    this._submap_versionIntialized = true;
    this._wrapperFields.textures.copyRefObject(this._refObject.textures);
  }
  copy(other) {
    this._refObject = new SubmapQuery_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.status.copy(other._wrapperFields.status);
    this._submap_versionIntialized = true;
    this._wrapperFields.textures.copy(other._wrapperFields.textures);
  }
  static get classType() {
    return SubmapQuery_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"status","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"cartographer_ros_msgs","type":"StatusResponse","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"submap_version","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"textures","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"cartographer_ros_msgs","type":"SubmapTexture","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"cartographer_ros_msgs","type":"SubmapQuery_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SubmapQuery_Response"};
  }
  hasMember(name) {
    let memberNames = ["status","submap_version","textures"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SubmapQuery_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SubmapQuery_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SubmapQuery_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SubmapQuery_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SubmapQuery_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SubmapQuery_ResponseRefArray(size);
    this._refObject = new SubmapQuery_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SubmapQuery_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SubmapQuery_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SubmapQuery_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SubmapQuery_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SubmapQuery_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SubmapQuery_ResponseArrayWrapper;
  }
}
module.exports = SubmapQuery_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "status",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "cartographer_ros_msgs",
        "type": "StatusResponse",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "submap_version",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "textures",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "cartographer_ros_msgs",
        "type": "SubmapTexture",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "cartographer_ros_msgs",
    "type": "SubmapQuery_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SubmapQuery_Response"
}
*/
