// rclnodejs/generated/cartographer_ros_msgs/cartographer_ros_msgs__srv__SubmapQuery_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const SubmapQuery_RequestRefStruct = StructType({
  trajectory_id: primitiveTypes.int32,
  submap_index: primitiveTypes.int32,
});
const SubmapQuery_RequestRefArray = ArrayType(SubmapQuery_RequestRefStruct);
const SubmapQuery_RequestRefStructArray = StructType({
  data: SubmapQuery_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SubmapQuery_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SubmapQuery_RequestRefStruct(other._refObject.toObject());
      this._trajectory_idIntialized = true;
      this._submap_indexIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SubmapQuery_RequestRefStruct();
    this._trajectory_idIntialized = false;
    this._submap_indexIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new SubmapQuery_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SubmapQuery_RequestArrayWrapper;
  }
  static get ArrayType() {
    return SubmapQuery_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return SubmapQuery_RequestRefStructArray
  }
  static get refObjectType() {
    return SubmapQuery_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._trajectory_idIntialized) {
        throw new TypeError('Invalid argument: trajectory_id in SubmapQuery_Request');
      }
      if (!this._submap_indexIntialized) {
        throw new TypeError('Invalid argument: submap_index in SubmapQuery_Request');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._trajectory_idIntialized = true;
    this._refObject.trajectory_id = refObject.trajectory_id;
    this._submap_indexIntialized = true;
    this._refObject.submap_index = refObject.submap_index;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    SubmapQuery_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'cartographer_ros_msgs', subFolder: 'srv', interfaceName: 'SubmapQuery_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get trajectory_id() {
    if (!this._trajectory_idIntialized) {
      return undefined;
    }
    return this._refObject.trajectory_id;
  }
  set trajectory_id(value) {
    this._trajectory_idIntialized = true;
    this._refObject.trajectory_id = value;
  }
  get submap_index() {
    if (!this._submap_indexIntialized) {
      return undefined;
    }
    return this._refObject.submap_index;
  }
  set submap_index(value) {
    this._submap_indexIntialized = true;
    this._refObject.submap_index = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SubmapQuery_RequestRefStruct(refObject.toObject());
    this._trajectory_idIntialized = true;
    this._submap_indexIntialized = true;
  }
  copy(other) {
    this._refObject = new SubmapQuery_RequestRefStruct(other._refObject.toObject());
    this._trajectory_idIntialized = true;
    this._submap_indexIntialized = true;
  }
  static get classType() {
    return SubmapQuery_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"trajectory_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"submap_index","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"cartographer_ros_msgs","type":"SubmapQuery_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SubmapQuery_Request"};
  }
  hasMember(name) {
    let memberNames = ["trajectory_id","submap_index"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SubmapQuery_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SubmapQuery_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SubmapQuery_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SubmapQuery_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SubmapQuery_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SubmapQuery_RequestRefArray(size);
    this._refObject = new SubmapQuery_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SubmapQuery_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SubmapQuery_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SubmapQuery_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SubmapQuery_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SubmapQuery_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SubmapQuery_RequestArrayWrapper;
  }
}
module.exports = SubmapQuery_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "trajectory_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "submap_index",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "cartographer_ros_msgs",
    "type": "SubmapQuery_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SubmapQuery_Request"
}
*/
