// rclnodejs/generated/cartographer_ros_msgs/cartographer_ros_msgs__msg__SubmapEntry.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PoseWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Pose.js');
const SubmapEntryRefStruct = StructType({
  trajectory_id: primitiveTypes.int32,
  submap_index: primitiveTypes.int32,
  submap_version: primitiveTypes.int32,
  pose: PoseWrapper.refObjectType,
});
const SubmapEntryRefArray = ArrayType(SubmapEntryRefStruct);
const SubmapEntryRefStructArray = StructType({
  data: SubmapEntryRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SubmapEntryWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SubmapEntryRefStruct(other._refObject.toObject());
      this._trajectory_idIntialized = true;
      this._submap_indexIntialized = true;
      this._submap_versionIntialized = true;
      this._wrapperFields.pose =  new PoseWrapper(other._wrapperFields.pose);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SubmapEntryRefStruct();
    this._trajectory_idIntialized = false;
    this._submap_indexIntialized = false;
    this._submap_versionIntialized = false;
    this._wrapperFields.pose = new PoseWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new SubmapEntryWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SubmapEntryArrayWrapper;
  }
  static get ArrayType() {
    return SubmapEntryArrayWrapper;
  }
  static get refObjectArrayType() {
    return SubmapEntryRefStructArray
  }
  static get refObjectType() {
    return SubmapEntryRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._trajectory_idIntialized) {
        throw new TypeError('Invalid argument: trajectory_id in SubmapEntry');
      }
      if (!this._submap_indexIntialized) {
        throw new TypeError('Invalid argument: submap_index in SubmapEntry');
      }
      if (!this._submap_versionIntialized) {
        throw new TypeError('Invalid argument: submap_version in SubmapEntry');
      }
    }
    this._wrapperFields.pose.freeze(own, checkConsistency);
    this._refObject.pose = this._wrapperFields.pose.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._trajectory_idIntialized = true;
    this._refObject.trajectory_id = refObject.trajectory_id;
    this._submap_indexIntialized = true;
    this._refObject.submap_index = refObject.submap_index;
    this._submap_versionIntialized = true;
    this._refObject.submap_version = refObject.submap_version;
    this._wrapperFields.pose.copyRefObject(refObject.pose);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    PoseWrapper.freeStruct(refObject.pose);
  }
  static destoryRawROS(msg) {
    SubmapEntryWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'cartographer_ros_msgs', subFolder: 'msg', interfaceName: 'SubmapEntry'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get trajectory_id() {
    if (!this._trajectory_idIntialized) {
      return undefined;
    }
    return this._refObject.trajectory_id;
  }
  set trajectory_id(value) {
    this._trajectory_idIntialized = true;
    this._refObject.trajectory_id = value;
  }
  get submap_index() {
    if (!this._submap_indexIntialized) {
      return undefined;
    }
    return this._refObject.submap_index;
  }
  set submap_index(value) {
    this._submap_indexIntialized = true;
    this._refObject.submap_index = value;
  }
  get submap_version() {
    if (!this._submap_versionIntialized) {
      return undefined;
    }
    return this._refObject.submap_version;
  }
  set submap_version(value) {
    this._submap_versionIntialized = true;
    this._refObject.submap_version = value;
  }
  get pose() {
    return this._wrapperFields.pose;
  }
  set pose(value) {
    if (value instanceof PoseWrapper) {
      this._wrapperFields.pose.copy(value);
    } else {
      this._wrapperFields.pose.copy(new PoseWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new SubmapEntryRefStruct(refObject.toObject());
    this._trajectory_idIntialized = true;
    this._submap_indexIntialized = true;
    this._submap_versionIntialized = true;
    this._wrapperFields.pose.copyRefObject(this._refObject.pose);
  }
  copy(other) {
    this._refObject = new SubmapEntryRefStruct(other._refObject.toObject());
    this._trajectory_idIntialized = true;
    this._submap_indexIntialized = true;
    this._submap_versionIntialized = true;
    this._wrapperFields.pose.copy(other._wrapperFields.pose);
  }
  static get classType() {
    return SubmapEntryWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"trajectory_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"submap_index","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"submap_version","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pose","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Pose","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"cartographer_ros_msgs","type":"SubmapEntry","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SubmapEntry"};
  }
  hasMember(name) {
    let memberNames = ["trajectory_id","submap_index","submap_version","pose"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SubmapEntryArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SubmapEntryWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SubmapEntryWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SubmapEntryArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SubmapEntryArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SubmapEntryRefArray(size);
    this._refObject = new SubmapEntryRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SubmapEntryWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SubmapEntryArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SubmapEntryArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SubmapEntryWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SubmapEntryWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SubmapEntryArrayWrapper;
  }
}
module.exports = SubmapEntryWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "trajectory_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "submap_index",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "submap_version",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pose",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Pose",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "cartographer_ros_msgs",
    "type": "SubmapEntry",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SubmapEntry"
}
*/
