// rclnodejs/generated/cartographer_ros_msgs/cartographer_ros_msgs__msg__LandmarkEntry.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const PoseWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Pose.js');
const LandmarkEntryRefStruct = StructType({
  id: primitiveTypes.string,
  tracking_from_landmark_transform: PoseWrapper.refObjectType,
  translation_weight: primitiveTypes.float64,
  rotation_weight: primitiveTypes.float64,
});
const LandmarkEntryRefArray = ArrayType(LandmarkEntryRefStruct);
const LandmarkEntryRefStructArray = StructType({
  data: LandmarkEntryRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class LandmarkEntryWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new LandmarkEntryRefStruct(other._refObject.toObject());
      this._idIntialized = true;
      this._wrapperFields.id =  new StringWrapper(other._wrapperFields.id);
      this._wrapperFields.tracking_from_landmark_transform =  new PoseWrapper(other._wrapperFields.tracking_from_landmark_transform);
      this._translation_weightIntialized = true;
      this._rotation_weightIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new LandmarkEntryRefStruct();
    this._idIntialized = false;
    this._wrapperFields.id = new StringWrapper();
    this._wrapperFields.tracking_from_landmark_transform = new PoseWrapper();
    this._translation_weightIntialized = false;
    this._rotation_weightIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new LandmarkEntryWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LandmarkEntryArrayWrapper;
  }
  static get ArrayType() {
    return LandmarkEntryArrayWrapper;
  }
  static get refObjectArrayType() {
    return LandmarkEntryRefStructArray
  }
  static get refObjectType() {
    return LandmarkEntryRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._idIntialized) {
        throw new TypeError('Invalid argument: id in LandmarkEntry');
      }
      if (!this._translation_weightIntialized) {
        throw new TypeError('Invalid argument: translation_weight in LandmarkEntry');
      }
      if (!this._rotation_weightIntialized) {
        throw new TypeError('Invalid argument: rotation_weight in LandmarkEntry');
      }
    }
    if (own) {
      this._wrapperFields.id.freeze(own, checkConsistency);
    }
    this._refObject.id = this._wrapperFields.id.refObject;
    this._wrapperFields.tracking_from_landmark_transform.freeze(own, checkConsistency);
    this._refObject.tracking_from_landmark_transform = this._wrapperFields.tracking_from_landmark_transform.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._idIntialized = true;
    this._wrapperFields.id.data = refObject.id.data;
    this._wrapperFields.tracking_from_landmark_transform.copyRefObject(refObject.tracking_from_landmark_transform);
    this._translation_weightIntialized = true;
    this._refObject.translation_weight = refObject.translation_weight;
    this._rotation_weightIntialized = true;
    this._refObject.rotation_weight = refObject.rotation_weight;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.id);
    PoseWrapper.freeStruct(refObject.tracking_from_landmark_transform);
  }
  static destoryRawROS(msg) {
    LandmarkEntryWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'cartographer_ros_msgs', subFolder: 'msg', interfaceName: 'LandmarkEntry'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get id() {
    if (!this._idIntialized) {
      return undefined;
    }
    return this._wrapperFields.id.data;
  }
  set id(value) {
    this._idIntialized = true;
    this._wrapperFields.id.data = value;
  }
  get tracking_from_landmark_transform() {
    return this._wrapperFields.tracking_from_landmark_transform;
  }
  set tracking_from_landmark_transform(value) {
    if (value instanceof PoseWrapper) {
      this._wrapperFields.tracking_from_landmark_transform.copy(value);
    } else {
      this._wrapperFields.tracking_from_landmark_transform.copy(new PoseWrapper(value));
    }
  }
  get translation_weight() {
    if (!this._translation_weightIntialized) {
      return undefined;
    }
    return this._refObject.translation_weight;
  }
  set translation_weight(value) {
    this._translation_weightIntialized = true;
    this._refObject.translation_weight = value;
  }
  get rotation_weight() {
    if (!this._rotation_weightIntialized) {
      return undefined;
    }
    return this._refObject.rotation_weight;
  }
  set rotation_weight(value) {
    this._rotation_weightIntialized = true;
    this._refObject.rotation_weight = value;
  }
  copyRefObject(refObject) {
    this._refObject = new LandmarkEntryRefStruct(refObject.toObject());
    this._idIntialized = true;
    this._wrapperFields.id.copyRefObject(this._refObject.id);
    this._wrapperFields.tracking_from_landmark_transform.copyRefObject(this._refObject.tracking_from_landmark_transform);
    this._translation_weightIntialized = true;
    this._rotation_weightIntialized = true;
  }
  copy(other) {
    this._refObject = new LandmarkEntryRefStruct(other._refObject.toObject());
    this._idIntialized = true;
    this._wrapperFields.id.copy(other._wrapperFields.id);
    this._wrapperFields.tracking_from_landmark_transform.copy(other._wrapperFields.tracking_from_landmark_transform);
    this._translation_weightIntialized = true;
    this._rotation_weightIntialized = true;
  }
  static get classType() {
    return LandmarkEntryWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"tracking_from_landmark_transform","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Pose","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"translation_weight","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"rotation_weight","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"cartographer_ros_msgs","type":"LandmarkEntry","stringUpperBound":null,"isPrimitiveType":false},"msgName":"LandmarkEntry"};
  }
  hasMember(name) {
    let memberNames = ["id","tracking_from_landmark_transform","translation_weight","rotation_weight"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class LandmarkEntryArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LandmarkEntryWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LandmarkEntryWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LandmarkEntryArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LandmarkEntryArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new LandmarkEntryRefArray(size);
    this._refObject = new LandmarkEntryRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LandmarkEntryWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LandmarkEntryArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LandmarkEntryArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LandmarkEntryWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LandmarkEntryWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LandmarkEntryArrayWrapper;
  }
}
module.exports = LandmarkEntryWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "tracking_from_landmark_transform",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Pose",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "translation_weight",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "rotation_weight",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "cartographer_ros_msgs",
    "type": "LandmarkEntry",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "LandmarkEntry"
}
*/
