// rclnodejs/generated/cartographer_ros_msgs/cartographer_ros_msgs__srv__FinishTrajectory_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StatusResponseWrapper = require('../../generated/cartographer_ros_msgs/cartographer_ros_msgs__msg__StatusResponse.js');
const FinishTrajectory_ResponseRefStruct = StructType({
  status: StatusResponseWrapper.refObjectType,
});
const FinishTrajectory_ResponseRefArray = ArrayType(FinishTrajectory_ResponseRefStruct);
const FinishTrajectory_ResponseRefStructArray = StructType({
  data: FinishTrajectory_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class FinishTrajectory_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new FinishTrajectory_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.status =  new StatusResponseWrapper(other._wrapperFields.status);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new FinishTrajectory_ResponseRefStruct();
    this._wrapperFields.status = new StatusResponseWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new FinishTrajectory_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new FinishTrajectory_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return FinishTrajectory_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return FinishTrajectory_ResponseRefStructArray
  }
  static get refObjectType() {
    return FinishTrajectory_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.status.freeze(own, checkConsistency);
    this._refObject.status = this._wrapperFields.status.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.status.copyRefObject(refObject.status);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StatusResponseWrapper.freeStruct(refObject.status);
  }
  static destoryRawROS(msg) {
    FinishTrajectory_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'cartographer_ros_msgs', subFolder: 'srv', interfaceName: 'FinishTrajectory_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get status() {
    return this._wrapperFields.status;
  }
  set status(value) {
    if (value instanceof StatusResponseWrapper) {
      this._wrapperFields.status.copy(value);
    } else {
      this._wrapperFields.status.copy(new StatusResponseWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new FinishTrajectory_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.status.copyRefObject(this._refObject.status);
  }
  copy(other) {
    this._refObject = new FinishTrajectory_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.status.copy(other._wrapperFields.status);
  }
  static get classType() {
    return FinishTrajectory_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"status","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"cartographer_ros_msgs","type":"StatusResponse","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"cartographer_ros_msgs","type":"FinishTrajectory_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"FinishTrajectory_Response"};
  }
  hasMember(name) {
    let memberNames = ["status"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class FinishTrajectory_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof FinishTrajectory_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new FinishTrajectory_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FinishTrajectory_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FinishTrajectory_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new FinishTrajectory_ResponseRefArray(size);
    this._refObject = new FinishTrajectory_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new FinishTrajectory_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof FinishTrajectory_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "FinishTrajectory_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      FinishTrajectory_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return FinishTrajectory_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return FinishTrajectory_ResponseArrayWrapper;
  }
}
module.exports = FinishTrajectory_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "status",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "cartographer_ros_msgs",
        "type": "StatusResponse",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "cartographer_ros_msgs",
    "type": "FinishTrajectory_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "FinishTrajectory_Response"
}
*/
