// rclnodejs/generated/realsense2_camera_msgs/realsense2_camera_msgs__msg__IMUInfo.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const Float64Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float64.js');
const IMUInfoRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  data: ArrayType(primitiveTypes.float64, 12),
  noise_variances: ArrayType(primitiveTypes.float64, 3),
  bias_variances: ArrayType(primitiveTypes.float64, 3),
});
const IMUInfoRefArray = ArrayType(IMUInfoRefStruct);
const IMUInfoRefStructArray = StructType({
  data: IMUInfoRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class IMUInfoWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new IMUInfoRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.data = Float64Wrapper.createArray();
      this._wrapperFields.data.copy(other._wrapperFields.data);
      this._wrapperFields.noise_variances = Float64Wrapper.createArray();
      this._wrapperFields.noise_variances.copy(other._wrapperFields.noise_variances);
      this._wrapperFields.bias_variances = Float64Wrapper.createArray();
      this._wrapperFields.bias_variances.copy(other._wrapperFields.bias_variances);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new IMUInfoRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.data = Float64Wrapper.createArray();
    this._wrapperFields.noise_variances = Float64Wrapper.createArray();
    this._wrapperFields.bias_variances = Float64Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new IMUInfoWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new IMUInfoArrayWrapper;
  }
  static get ArrayType() {
    return IMUInfoArrayWrapper;
  }
  static get refObjectArrayType() {
    return IMUInfoRefStructArray
  }
  static get refObjectType() {
    return IMUInfoRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    for (let i = 0; i < this._wrapperFields.data.data.length; i++) {
      this._refObject.data[i] = this._wrapperFields.data.data[i];
    }
    for (let i = 0; i < this._wrapperFields.noise_variances.data.length; i++) {
      this._refObject.noise_variances[i] = this._wrapperFields.noise_variances.data[i];
    }
    for (let i = 0; i < this._wrapperFields.bias_variances.data.length; i++) {
      this._refObject.bias_variances[i] = this._wrapperFields.bias_variances.data[i];
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._refObject.data = refObject.data;
    this._wrapperFields.data.fill(refObject.data.toArray());
    this._refObject.noise_variances = refObject.noise_variances;
    this._wrapperFields.noise_variances.fill(refObject.noise_variances.toArray());
    this._refObject.bias_variances = refObject.bias_variances;
    this._wrapperFields.bias_variances.fill(refObject.bias_variances.toArray());
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
  }
  static destoryRawROS(msg) {
    IMUInfoWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'realsense2_camera_msgs', subFolder: 'msg', interfaceName: 'IMUInfo'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get data() {
    return this._wrapperFields['data'].data;
  }
  set data(value) {
    if (value.length !== 12) {
      throw new RangeError('The length of the array must be 12.');
    }
    this._wrapperFields['data'].fill(value);
  }
  get noise_variances() {
    return this._wrapperFields['noise_variances'].data;
  }
  set noise_variances(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['noise_variances'].fill(value);
  }
  get bias_variances() {
    return this._wrapperFields['bias_variances'].data;
  }
  set bias_variances(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['bias_variances'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new IMUInfoRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.data.fill(refObject.data.toArray());
    this._wrapperFields.noise_variances.fill(refObject.noise_variances.toArray());
    this._wrapperFields.bias_variances.fill(refObject.bias_variances.toArray());
  }
  copy(other) {
    this._refObject = new IMUInfoRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.data.copy(other._wrapperFields.data);
    this._wrapperFields.noise_variances.copy(other._wrapperFields.noise_variances);
    this._wrapperFields.bias_variances.copy(other._wrapperFields.bias_variances);
  }
  static get classType() {
    return IMUInfoWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"data","type":{"isArray":true,"arraySize":12,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"noise_variances","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"bias_variances","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"realsense2_camera_msgs","type":"IMUInfo","stringUpperBound":null,"isPrimitiveType":false},"msgName":"IMUInfo"};
  }
  hasMember(name) {
    let memberNames = ["header","data","noise_variances","bias_variances"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class IMUInfoArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof IMUInfoWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new IMUInfoWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to IMUInfoArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to IMUInfoArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new IMUInfoRefArray(size);
    this._refObject = new IMUInfoRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new IMUInfoWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof IMUInfoArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "IMUInfoArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      IMUInfoWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return IMUInfoWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return IMUInfoArrayWrapper;
  }
}
module.exports = IMUInfoWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "data",
      "type": {
        "isArray": true,
        "arraySize": 12,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "noise_variances",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "bias_variances",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "realsense2_camera_msgs",
    "type": "IMUInfo",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "IMUInfo"
}
*/
