// rclnodejs/generated/controller_manager_msgs/controller_manager_msgs__srv__ListControllers_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const ControllerStateWrapper = require('../../generated/controller_manager_msgs/controller_manager_msgs__msg__ControllerState.js');
const ListControllers_ResponseRefStruct = StructType({
  controller: ControllerStateWrapper.refObjectArrayType,
});
const ListControllers_ResponseRefArray = ArrayType(ListControllers_ResponseRefStruct);
const ListControllers_ResponseRefStructArray = StructType({
  data: ListControllers_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ListControllers_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ListControllers_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.controller = ControllerStateWrapper.createArray();
      this._wrapperFields.controller.copy(other._wrapperFields.controller);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ListControllers_ResponseRefStruct();
    this._wrapperFields.controller = ControllerStateWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new ListControllers_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ListControllers_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return ListControllers_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return ListControllers_ResponseRefStructArray
  }
  static get refObjectType() {
    return ListControllers_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.controller.freeze(own, checkConsistency);
    this._refObject.controller = this._wrapperFields.controller.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.controller.copyRefObject(refObject.controller);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.controller.size != 0) {
      ControllerStateWrapper.ArrayType.freeArray(refObject.controller);
      if (ControllerStateWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.controller, ControllerStateWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    ListControllers_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'controller_manager_msgs', subFolder: 'srv', interfaceName: 'ListControllers_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get controller() {
    return this._wrapperFields.controller;
  }
  set controller(value) {
    this._wrapperFields.controller.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new ListControllers_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.controller.copyRefObject(this._refObject.controller);
  }
  copy(other) {
    this._refObject = new ListControllers_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.controller.copy(other._wrapperFields.controller);
  }
  static get classType() {
    return ListControllers_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"controller","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"controller_manager_msgs","type":"ControllerState","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"controller_manager_msgs","type":"ListControllers_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ListControllers_Response"};
  }
  hasMember(name) {
    let memberNames = ["controller"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ListControllers_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ListControllers_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ListControllers_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ListControllers_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ListControllers_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ListControllers_ResponseRefArray(size);
    this._refObject = new ListControllers_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ListControllers_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ListControllers_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ListControllers_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ListControllers_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ListControllers_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ListControllers_ResponseArrayWrapper;
  }
}
module.exports = ListControllers_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "controller",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "controller_manager_msgs",
        "type": "ControllerState",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "controller_manager_msgs",
    "type": "ListControllers_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ListControllers_Response"
}
*/
