// rclnodejs/generated/controller_manager_msgs/controller_manager_msgs__srv__ListControllerTypes_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const ListControllerTypes_ResponseRefStruct = StructType({
  types: StringWrapper.refObjectArrayType,
  base_classes: StringWrapper.refObjectArrayType,
});
const ListControllerTypes_ResponseRefArray = ArrayType(ListControllerTypes_ResponseRefStruct);
const ListControllerTypes_ResponseRefStructArray = StructType({
  data: ListControllerTypes_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ListControllerTypes_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    this._typesArray = [];
    this._base_classesArray = [];
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ListControllerTypes_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.types = StringWrapper.createArray();
      this._wrapperFields.types.copy(other._wrapperFields.types);
      this.types = other.types;
      this._wrapperFields.base_classes = StringWrapper.createArray();
      this._wrapperFields.base_classes.copy(other._wrapperFields.base_classes);
      this.base_classes = other.base_classes;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ListControllerTypes_ResponseRefStruct();
    this._wrapperFields.types = StringWrapper.createArray();
    this._wrapperFields.base_classes = StringWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new ListControllerTypes_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ListControllerTypes_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return ListControllerTypes_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return ListControllerTypes_ResponseRefStructArray
  }
  static get refObjectType() {
    return ListControllerTypes_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    if (!own) {
      this._wrapperFields.types.fill(this._typesArray);
      this._wrapperFields.types.freeze(own, checkConsistency);
      this._refObject.types = this._wrapperFields.types.refObject;
    } else {
      this._wrapperFields.types.fill([]);
      this._wrapperFields.types.freeze(own, checkConsistency);
      this._refObject.types = this._wrapperFields.types.refObject;
    }
    if (!own) {
      this._wrapperFields.base_classes.fill(this._base_classesArray);
      this._wrapperFields.base_classes.freeze(own, checkConsistency);
      this._refObject.base_classes = this._wrapperFields.base_classes.refObject;
    } else {
      this._wrapperFields.base_classes.fill([]);
      this._wrapperFields.base_classes.freeze(own, checkConsistency);
      this._refObject.base_classes = this._wrapperFields.base_classes.refObject;
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    refObject.types.data.length = refObject.types.size;
    for (let index = 0; index < refObject.types.size; index++) {
      this._typesArray[index] = refObject.types.data[index].data;
    }
    refObject.base_classes.data.length = refObject.base_classes.size;
    for (let index = 0; index < refObject.base_classes.size; index++) {
      this._base_classesArray[index] = refObject.base_classes.data[index].data;
    }
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.types.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.types);
      if (StringWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.types, StringWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.base_classes.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.base_classes);
      if (StringWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.base_classes, StringWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    ListControllerTypes_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'controller_manager_msgs', subFolder: 'srv', interfaceName: 'ListControllerTypes_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get types() {
    return this._typesArray;
  }
  set types(value) {
    this._typesArray = value;
  }
  get base_classes() {
    return this._base_classesArray;
  }
  set base_classes(value) {
    this._base_classesArray = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ListControllerTypes_ResponseRefStruct(refObject.toObject());
    refObject.types.data.length = refObject.types.size;
    for (let index = 0; index < refObject.types.size; index++) {
      this._typesArray[index] = refObject.types.data[index].data;
    }
    refObject.base_classes.data.length = refObject.base_classes.size;
    for (let index = 0; index < refObject.base_classes.size; index++) {
      this._base_classesArray[index] = refObject.base_classes.data[index].data;
    }
  }
  copy(other) {
    this._refObject = new ListControllerTypes_ResponseRefStruct(other._refObject.toObject());
    this._typesArray = other._typesArray.slice();
    this._base_classesArray = other._base_classesArray.slice();
  }
  static get classType() {
    return ListControllerTypes_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"types","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"base_classes","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"controller_manager_msgs","type":"ListControllerTypes_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ListControllerTypes_Response"};
  }
  hasMember(name) {
    let memberNames = ["types","base_classes"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ListControllerTypes_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ListControllerTypes_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ListControllerTypes_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ListControllerTypes_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ListControllerTypes_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ListControllerTypes_ResponseRefArray(size);
    this._refObject = new ListControllerTypes_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ListControllerTypes_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ListControllerTypes_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ListControllerTypes_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ListControllerTypes_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ListControllerTypes_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ListControllerTypes_ResponseArrayWrapper;
  }
}
module.exports = ListControllerTypes_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "types",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "base_classes",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "controller_manager_msgs",
    "type": "ListControllerTypes_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ListControllerTypes_Response"
}
*/
