// rclnodejs/generated/control_msgs/control_msgs__action__SingleJointPosition_Goal.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const DurationWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Duration.js');
const SingleJointPosition_GoalRefStruct = StructType({
  position: primitiveTypes.float64,
  min_duration: DurationWrapper.refObjectType,
  max_velocity: primitiveTypes.float64,
});
const SingleJointPosition_GoalRefArray = ArrayType(SingleJointPosition_GoalRefStruct);
const SingleJointPosition_GoalRefStructArray = StructType({
  data: SingleJointPosition_GoalRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SingleJointPosition_GoalWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SingleJointPosition_GoalRefStruct(other._refObject.toObject());
      this._positionIntialized = true;
      this._wrapperFields.min_duration =  new DurationWrapper(other._wrapperFields.min_duration);
      this._max_velocityIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SingleJointPosition_GoalRefStruct();
    this._positionIntialized = false;
    this._wrapperFields.min_duration = new DurationWrapper();
    this._max_velocityIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new SingleJointPosition_GoalWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SingleJointPosition_GoalArrayWrapper;
  }
  static get ArrayType() {
    return SingleJointPosition_GoalArrayWrapper;
  }
  static get refObjectArrayType() {
    return SingleJointPosition_GoalRefStructArray
  }
  static get refObjectType() {
    return SingleJointPosition_GoalRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._positionIntialized) {
        throw new TypeError('Invalid argument: position in SingleJointPosition_Goal');
      }
      if (!this._max_velocityIntialized) {
        throw new TypeError('Invalid argument: max_velocity in SingleJointPosition_Goal');
      }
    }
    this._wrapperFields.min_duration.freeze(own, checkConsistency);
    this._refObject.min_duration = this._wrapperFields.min_duration.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._positionIntialized = true;
    this._refObject.position = refObject.position;
    this._wrapperFields.min_duration.copyRefObject(refObject.min_duration);
    this._max_velocityIntialized = true;
    this._refObject.max_velocity = refObject.max_velocity;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    DurationWrapper.freeStruct(refObject.min_duration);
  }
  static destoryRawROS(msg) {
    SingleJointPosition_GoalWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'control_msgs', subFolder: 'action', interfaceName: 'SingleJointPosition_Goal'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get position() {
    if (!this._positionIntialized) {
      return undefined;
    }
    return this._refObject.position;
  }
  set position(value) {
    this._positionIntialized = true;
    this._refObject.position = value;
  }
  get min_duration() {
    return this._wrapperFields.min_duration;
  }
  set min_duration(value) {
    if (value instanceof DurationWrapper) {
      this._wrapperFields.min_duration.copy(value);
    } else {
      this._wrapperFields.min_duration.copy(new DurationWrapper(value));
    }
  }
  get max_velocity() {
    if (!this._max_velocityIntialized) {
      return undefined;
    }
    return this._refObject.max_velocity;
  }
  set max_velocity(value) {
    this._max_velocityIntialized = true;
    this._refObject.max_velocity = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SingleJointPosition_GoalRefStruct(refObject.toObject());
    this._positionIntialized = true;
    this._wrapperFields.min_duration.copyRefObject(this._refObject.min_duration);
    this._max_velocityIntialized = true;
  }
  copy(other) {
    this._refObject = new SingleJointPosition_GoalRefStruct(other._refObject.toObject());
    this._positionIntialized = true;
    this._wrapperFields.min_duration.copy(other._wrapperFields.min_duration);
    this._max_velocityIntialized = true;
  }
  static get classType() {
    return SingleJointPosition_GoalWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"position","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"min_duration","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Duration","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"max_velocity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"control_msgs","type":"SingleJointPosition_Goal","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SingleJointPosition_Goal"};
  }
  hasMember(name) {
    let memberNames = ["position","min_duration","max_velocity"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SingleJointPosition_GoalArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SingleJointPosition_GoalWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SingleJointPosition_GoalWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SingleJointPosition_GoalArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SingleJointPosition_GoalArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SingleJointPosition_GoalRefArray(size);
    this._refObject = new SingleJointPosition_GoalRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SingleJointPosition_GoalWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SingleJointPosition_GoalArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SingleJointPosition_GoalArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SingleJointPosition_GoalWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SingleJointPosition_GoalWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SingleJointPosition_GoalArrayWrapper;
  }
}
module.exports = SingleJointPosition_GoalWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "position",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "min_duration",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Duration",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "max_velocity",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "control_msgs",
    "type": "SingleJointPosition_Goal",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SingleJointPosition_Goal"
}
*/
