// rclnodejs/generated/control_msgs/control_msgs__srv__QueryTrajectoryState_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TimeWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Time.js');
const QueryTrajectoryState_RequestRefStruct = StructType({
  time: TimeWrapper.refObjectType,
});
const QueryTrajectoryState_RequestRefArray = ArrayType(QueryTrajectoryState_RequestRefStruct);
const QueryTrajectoryState_RequestRefStructArray = StructType({
  data: QueryTrajectoryState_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class QueryTrajectoryState_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new QueryTrajectoryState_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.time =  new TimeWrapper(other._wrapperFields.time);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new QueryTrajectoryState_RequestRefStruct();
    this._wrapperFields.time = new TimeWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new QueryTrajectoryState_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new QueryTrajectoryState_RequestArrayWrapper;
  }
  static get ArrayType() {
    return QueryTrajectoryState_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return QueryTrajectoryState_RequestRefStructArray
  }
  static get refObjectType() {
    return QueryTrajectoryState_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.time.freeze(own, checkConsistency);
    this._refObject.time = this._wrapperFields.time.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.time.copyRefObject(refObject.time);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    TimeWrapper.freeStruct(refObject.time);
  }
  static destoryRawROS(msg) {
    QueryTrajectoryState_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'control_msgs', subFolder: 'srv', interfaceName: 'QueryTrajectoryState_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get time() {
    return this._wrapperFields.time;
  }
  set time(value) {
    if (value instanceof TimeWrapper) {
      this._wrapperFields.time.copy(value);
    } else {
      this._wrapperFields.time.copy(new TimeWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new QueryTrajectoryState_RequestRefStruct(refObject.toObject());
    this._wrapperFields.time.copyRefObject(this._refObject.time);
  }
  copy(other) {
    this._refObject = new QueryTrajectoryState_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.time.copy(other._wrapperFields.time);
  }
  static get classType() {
    return QueryTrajectoryState_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"time","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"control_msgs","type":"QueryTrajectoryState_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"QueryTrajectoryState_Request"};
  }
  hasMember(name) {
    let memberNames = ["time"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class QueryTrajectoryState_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof QueryTrajectoryState_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new QueryTrajectoryState_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to QueryTrajectoryState_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to QueryTrajectoryState_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new QueryTrajectoryState_RequestRefArray(size);
    this._refObject = new QueryTrajectoryState_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new QueryTrajectoryState_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof QueryTrajectoryState_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "QueryTrajectoryState_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      QueryTrajectoryState_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return QueryTrajectoryState_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return QueryTrajectoryState_RequestArrayWrapper;
  }
}
module.exports = QueryTrajectoryState_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "time",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Time",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "control_msgs",
    "type": "QueryTrajectoryState_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "QueryTrajectoryState_Request"
}
*/
