// rclnodejs/generated/control_msgs/control_msgs__msg__InterfaceValue.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const Float64Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float64.js');
const InterfaceValueRefStruct = StructType({
  interface_names: StringWrapper.refObjectArrayType,
  values: Float64Wrapper.refObjectArrayType,
});
const InterfaceValueRefArray = ArrayType(InterfaceValueRefStruct);
const InterfaceValueRefStructArray = StructType({
  data: InterfaceValueRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class InterfaceValueWrapper {
  constructor(other) {
    this._wrapperFields = {};
    this._interface_namesArray = [];
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new InterfaceValueRefStruct(other._refObject.toObject());
      this._wrapperFields.interface_names = StringWrapper.createArray();
      this._wrapperFields.interface_names.copy(other._wrapperFields.interface_names);
      this.interface_names = other.interface_names;
      this._wrapperFields.values = Float64Wrapper.createArray();
      this._wrapperFields.values.copy(other._wrapperFields.values);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new InterfaceValueRefStruct();
    this._wrapperFields.interface_names = StringWrapper.createArray();
    this._wrapperFields.values = Float64Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new InterfaceValueWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new InterfaceValueArrayWrapper;
  }
  static get ArrayType() {
    return InterfaceValueArrayWrapper;
  }
  static get refObjectArrayType() {
    return InterfaceValueRefStructArray
  }
  static get refObjectType() {
    return InterfaceValueRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    if (!own) {
      this._wrapperFields.interface_names.fill(this._interface_namesArray);
      this._wrapperFields.interface_names.freeze(own, checkConsistency);
      this._refObject.interface_names = this._wrapperFields.interface_names.refObject;
    } else {
      this._wrapperFields.interface_names.fill([]);
      this._wrapperFields.interface_names.freeze(own, checkConsistency);
      this._refObject.interface_names = this._wrapperFields.interface_names.refObject;
    }
    this._wrapperFields.values.freeze(own, checkConsistency);
    this._refObject.values = this._wrapperFields.values.refObject;
    if (own) {
      this._wrapperFields.values.fill(Float64Array.from([]));
      this._wrapperFields.values.freeze(own, checkConsistency);
      this._refObject.values = this._wrapperFields.values.refObject;
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    refObject.interface_names.data.length = refObject.interface_names.size;
    for (let index = 0; index < refObject.interface_names.size; index++) {
      this._interface_namesArray[index] = refObject.interface_names.data[index].data;
    }
    this._wrapperFields.values.copyRefObject(refObject.values);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.interface_names.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.interface_names);
      if (StringWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.interface_names, StringWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.values.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.values);
      if (Float64Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.values, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    InterfaceValueWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'control_msgs', subFolder: 'msg', interfaceName: 'InterfaceValue'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get interface_names() {
    return this._interface_namesArray;
  }
  set interface_names(value) {
    this._interface_namesArray = value;
  }
  get values() {
    return this._wrapperFields['values'].data;
  }
  set values(value) {
    this._wrapperFields['values'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new InterfaceValueRefStruct(refObject.toObject());
    refObject.interface_names.data.length = refObject.interface_names.size;
    for (let index = 0; index < refObject.interface_names.size; index++) {
      this._interface_namesArray[index] = refObject.interface_names.data[index].data;
    }
    this._wrapperFields.values.copyRefObject(this._refObject.values);
  }
  copy(other) {
    this._refObject = new InterfaceValueRefStruct(other._refObject.toObject());
    this._interface_namesArray = other._interface_namesArray.slice();
    this._wrapperFields.values.copy(other._wrapperFields.values);
  }
  static get classType() {
    return InterfaceValueWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"interface_names","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"values","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"control_msgs","type":"InterfaceValue","stringUpperBound":null,"isPrimitiveType":false},"msgName":"InterfaceValue"};
  }
  hasMember(name) {
    let memberNames = ["interface_names","values"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class InterfaceValueArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof InterfaceValueWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new InterfaceValueWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to InterfaceValueArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to InterfaceValueArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new InterfaceValueRefArray(size);
    this._refObject = new InterfaceValueRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new InterfaceValueWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof InterfaceValueArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "InterfaceValueArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      InterfaceValueWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return InterfaceValueWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return InterfaceValueArrayWrapper;
  }
}
module.exports = InterfaceValueWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "interface_names",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "values",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "control_msgs",
    "type": "InterfaceValue",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "InterfaceValue"
}
*/
