// rclnodejs/generated/control_msgs/control_msgs__action__SingleJointPosition.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
class control_msgs__action__SingleJointPosition {
  static get Goal() {
    return require('./control_msgs__action__SingleJointPosition_Goal.js');
  }
  static get Result() {
    return require('./control_msgs__action__SingleJointPosition_Result.js');
  }
  static get Feedback() {
    return require('./control_msgs__action__SingleJointPosition_Feedback.js');
  }
  static get impl() {
    return {
      get SendGoalService() {
        return require('./control_msgs__action__SingleJointPosition_SendGoal.js');
      },
      get GetResultService() {
        return require('./control_msgs__action__SingleJointPosition_GetResult.js');
      },
      get FeedbackMessage() {
        return require('./control_msgs__action__SingleJointPosition_FeedbackMessage.js');
      },
      get CancelGoal() {
        return require('../action_msgs/action_msgs__srv__CancelGoal.js');
      },
      get GoalStatusArray() {
        return require('../action_msgs/action_msgs__msg__GoalStatusArray.js');
      }
    };
  }
  static type() {
    return {pkgName: 'control_msgs', subFolder: 'action', interfaceName: 'SingleJointPosition'};
  }
}
module.exports = control_msgs__action__SingleJointPosition;
