// rclnodejs/generated/control_msgs/control_msgs__action__FollowJointTrajectory_Result.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const FollowJointTrajectory_ResultRefStruct = StructType({
  error_code: primitiveTypes.int32,
  error_string: primitiveTypes.string,
});
const FollowJointTrajectory_ResultRefArray = ArrayType(FollowJointTrajectory_ResultRefStruct);
const FollowJointTrajectory_ResultRefStructArray = StructType({
  data: FollowJointTrajectory_ResultRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class FollowJointTrajectory_ResultWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new FollowJointTrajectory_ResultRefStruct(other._refObject.toObject());
      this._error_codeIntialized = true;
      this._error_stringIntialized = true;
      this._wrapperFields.error_string =  new StringWrapper(other._wrapperFields.error_string);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new FollowJointTrajectory_ResultRefStruct();
    this._error_codeIntialized = false;
    this._error_stringIntialized = false;
    this._wrapperFields.error_string = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new FollowJointTrajectory_ResultWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new FollowJointTrajectory_ResultArrayWrapper;
  }
  static get ArrayType() {
    return FollowJointTrajectory_ResultArrayWrapper;
  }
  static get refObjectArrayType() {
    return FollowJointTrajectory_ResultRefStructArray
  }
  static get refObjectType() {
    return FollowJointTrajectory_ResultRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._error_codeIntialized) {
        throw new TypeError('Invalid argument: error_code in FollowJointTrajectory_Result');
      }
      if (!this._error_stringIntialized) {
        throw new TypeError('Invalid argument: error_string in FollowJointTrajectory_Result');
      }
    }
    if (own) {
      this._wrapperFields.error_string.freeze(own, checkConsistency);
    }
    this._refObject.error_string = this._wrapperFields.error_string.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._error_codeIntialized = true;
    this._refObject.error_code = refObject.error_code;
    this._error_stringIntialized = true;
    this._wrapperFields.error_string.data = refObject.error_string.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.error_string);
  }
  static destoryRawROS(msg) {
    FollowJointTrajectory_ResultWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'control_msgs', subFolder: 'action', interfaceName: 'FollowJointTrajectory_Result'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get error_code() {
    if (!this._error_codeIntialized) {
      return undefined;
    }
    return this._refObject.error_code;
  }
  set error_code(value) {
    this._error_codeIntialized = true;
    this._refObject.error_code = value;
  }
  get error_string() {
    if (!this._error_stringIntialized) {
      return undefined;
    }
    return this._wrapperFields.error_string.data;
  }
  set error_string(value) {
    this._error_stringIntialized = true;
    this._wrapperFields.error_string.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new FollowJointTrajectory_ResultRefStruct(refObject.toObject());
    this._error_codeIntialized = true;
    this._error_stringIntialized = true;
    this._wrapperFields.error_string.copyRefObject(this._refObject.error_string);
  }
  copy(other) {
    this._refObject = new FollowJointTrajectory_ResultRefStruct(other._refObject.toObject());
    this._error_codeIntialized = true;
    this._error_stringIntialized = true;
    this._wrapperFields.error_string.copy(other._wrapperFields.error_string);
  }
  static get classType() {
    return FollowJointTrajectory_ResultWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"int32","name":"SUCCESSFUL","value":0},{"type":"int32","name":"INVALID_GOAL","value":-1},{"type":"int32","name":"INVALID_JOINTS","value":-2},{"type":"int32","name":"OLD_HEADER_TIMESTAMP","value":-3},{"type":"int32","name":"PATH_TOLERANCE_VIOLATED","value":-4},{"type":"int32","name":"GOAL_TOLERANCE_VIOLATED","value":-5}],"fields":[{"name":"error_code","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"error_string","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"control_msgs","type":"FollowJointTrajectory_Result","stringUpperBound":null,"isPrimitiveType":false},"msgName":"FollowJointTrajectory_Result"};
  }
  hasMember(name) {
    let memberNames = ["error_code","error_string"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class FollowJointTrajectory_ResultArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof FollowJointTrajectory_ResultWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new FollowJointTrajectory_ResultWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FollowJointTrajectory_ResultArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FollowJointTrajectory_ResultArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new FollowJointTrajectory_ResultRefArray(size);
    this._refObject = new FollowJointTrajectory_ResultRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new FollowJointTrajectory_ResultWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof FollowJointTrajectory_ResultArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "FollowJointTrajectory_ResultArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      FollowJointTrajectory_ResultWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return FollowJointTrajectory_ResultWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return FollowJointTrajectory_ResultArrayWrapper;
  }
}
// Define constants (6 in total)
Object.defineProperty(FollowJointTrajectory_ResultWrapper, "SUCCESSFUL", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FollowJointTrajectory_ResultWrapper, "INVALID_GOAL", {value: -1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FollowJointTrajectory_ResultWrapper, "INVALID_JOINTS", {value: -2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FollowJointTrajectory_ResultWrapper, "OLD_HEADER_TIMESTAMP", {value: -3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FollowJointTrajectory_ResultWrapper, "PATH_TOLERANCE_VIOLATED", {value: -4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(FollowJointTrajectory_ResultWrapper, "GOAL_TOLERANCE_VIOLATED", {value: -5, writable: false, enumerable: true, configurable: true});
module.exports = FollowJointTrajectory_ResultWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "int32",
      "name": "SUCCESSFUL",
      "value": 0
    },
    {
      "type": "int32",
      "name": "INVALID_GOAL",
      "value": -1
    },
    {
      "type": "int32",
      "name": "INVALID_JOINTS",
      "value": -2
    },
    {
      "type": "int32",
      "name": "OLD_HEADER_TIMESTAMP",
      "value": -3
    },
    {
      "type": "int32",
      "name": "PATH_TOLERANCE_VIOLATED",
      "value": -4
    },
    {
      "type": "int32",
      "name": "GOAL_TOLERANCE_VIOLATED",
      "value": -5
    }
  ],
  "fields": [
    {
      "name": "error_code",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "error_string",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "control_msgs",
    "type": "FollowJointTrajectory_Result",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "FollowJointTrajectory_Result"
}
*/
