// rclnodejs/generated/control_msgs/control_msgs__msg__JointTrajectoryControllerState.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const JointTrajectoryPointWrapper = require('../../generated/trajectory_msgs/trajectory_msgs__msg__JointTrajectoryPoint.js');
const JointTrajectoryControllerStateRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  joint_names: StringWrapper.refObjectArrayType,
  desired: JointTrajectoryPointWrapper.refObjectType,
  actual: JointTrajectoryPointWrapper.refObjectType,
  error: JointTrajectoryPointWrapper.refObjectType,
});
const JointTrajectoryControllerStateRefArray = ArrayType(JointTrajectoryControllerStateRefStruct);
const JointTrajectoryControllerStateRefStructArray = StructType({
  data: JointTrajectoryControllerStateRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class JointTrajectoryControllerStateWrapper {
  constructor(other) {
    this._wrapperFields = {};
    this._joint_namesArray = [];
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new JointTrajectoryControllerStateRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.joint_names = StringWrapper.createArray();
      this._wrapperFields.joint_names.copy(other._wrapperFields.joint_names);
      this.joint_names = other.joint_names;
      this._wrapperFields.desired =  new JointTrajectoryPointWrapper(other._wrapperFields.desired);
      this._wrapperFields.actual =  new JointTrajectoryPointWrapper(other._wrapperFields.actual);
      this._wrapperFields.error =  new JointTrajectoryPointWrapper(other._wrapperFields.error);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new JointTrajectoryControllerStateRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.joint_names = StringWrapper.createArray();
    this._wrapperFields.desired = new JointTrajectoryPointWrapper();
    this._wrapperFields.actual = new JointTrajectoryPointWrapper();
    this._wrapperFields.error = new JointTrajectoryPointWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new JointTrajectoryControllerStateWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new JointTrajectoryControllerStateArrayWrapper;
  }
  static get ArrayType() {
    return JointTrajectoryControllerStateArrayWrapper;
  }
  static get refObjectArrayType() {
    return JointTrajectoryControllerStateRefStructArray
  }
  static get refObjectType() {
    return JointTrajectoryControllerStateRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (!own) {
      this._wrapperFields.joint_names.fill(this._joint_namesArray);
      this._wrapperFields.joint_names.freeze(own, checkConsistency);
      this._refObject.joint_names = this._wrapperFields.joint_names.refObject;
    } else {
      this._wrapperFields.joint_names.fill([]);
      this._wrapperFields.joint_names.freeze(own, checkConsistency);
      this._refObject.joint_names = this._wrapperFields.joint_names.refObject;
    }
    this._wrapperFields.desired.freeze(own, checkConsistency);
    this._refObject.desired = this._wrapperFields.desired.refObject;
    this._wrapperFields.actual.freeze(own, checkConsistency);
    this._refObject.actual = this._wrapperFields.actual.refObject;
    this._wrapperFields.error.freeze(own, checkConsistency);
    this._refObject.error = this._wrapperFields.error.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    refObject.joint_names.data.length = refObject.joint_names.size;
    for (let index = 0; index < refObject.joint_names.size; index++) {
      this._joint_namesArray[index] = refObject.joint_names.data[index].data;
    }
    this._wrapperFields.desired.copyRefObject(refObject.desired);
    this._wrapperFields.actual.copyRefObject(refObject.actual);
    this._wrapperFields.error.copyRefObject(refObject.error);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.joint_names.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.joint_names);
      if (StringWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.joint_names, StringWrapper.refObjectArrayType, 'data');
      }
    }
    JointTrajectoryPointWrapper.freeStruct(refObject.desired);
    JointTrajectoryPointWrapper.freeStruct(refObject.actual);
    JointTrajectoryPointWrapper.freeStruct(refObject.error);
  }
  static destoryRawROS(msg) {
    JointTrajectoryControllerStateWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'control_msgs', subFolder: 'msg', interfaceName: 'JointTrajectoryControllerState'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get joint_names() {
    return this._joint_namesArray;
  }
  set joint_names(value) {
    this._joint_namesArray = value;
  }
  get desired() {
    return this._wrapperFields.desired;
  }
  set desired(value) {
    if (value instanceof JointTrajectoryPointWrapper) {
      this._wrapperFields.desired.copy(value);
    } else {
      this._wrapperFields.desired.copy(new JointTrajectoryPointWrapper(value));
    }
  }
  get actual() {
    return this._wrapperFields.actual;
  }
  set actual(value) {
    if (value instanceof JointTrajectoryPointWrapper) {
      this._wrapperFields.actual.copy(value);
    } else {
      this._wrapperFields.actual.copy(new JointTrajectoryPointWrapper(value));
    }
  }
  get error() {
    return this._wrapperFields.error;
  }
  set error(value) {
    if (value instanceof JointTrajectoryPointWrapper) {
      this._wrapperFields.error.copy(value);
    } else {
      this._wrapperFields.error.copy(new JointTrajectoryPointWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new JointTrajectoryControllerStateRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    refObject.joint_names.data.length = refObject.joint_names.size;
    for (let index = 0; index < refObject.joint_names.size; index++) {
      this._joint_namesArray[index] = refObject.joint_names.data[index].data;
    }
    this._wrapperFields.desired.copyRefObject(this._refObject.desired);
    this._wrapperFields.actual.copyRefObject(this._refObject.actual);
    this._wrapperFields.error.copyRefObject(this._refObject.error);
  }
  copy(other) {
    this._refObject = new JointTrajectoryControllerStateRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._joint_namesArray = other._joint_namesArray.slice();
    this._wrapperFields.desired.copy(other._wrapperFields.desired);
    this._wrapperFields.actual.copy(other._wrapperFields.actual);
    this._wrapperFields.error.copy(other._wrapperFields.error);
  }
  static get classType() {
    return JointTrajectoryControllerStateWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"joint_names","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"desired","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"trajectory_msgs","type":"JointTrajectoryPoint","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"actual","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"trajectory_msgs","type":"JointTrajectoryPoint","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"error","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"trajectory_msgs","type":"JointTrajectoryPoint","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"control_msgs","type":"JointTrajectoryControllerState","stringUpperBound":null,"isPrimitiveType":false},"msgName":"JointTrajectoryControllerState"};
  }
  hasMember(name) {
    let memberNames = ["header","joint_names","desired","actual","error"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class JointTrajectoryControllerStateArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof JointTrajectoryControllerStateWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new JointTrajectoryControllerStateWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to JointTrajectoryControllerStateArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to JointTrajectoryControllerStateArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new JointTrajectoryControllerStateRefArray(size);
    this._refObject = new JointTrajectoryControllerStateRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new JointTrajectoryControllerStateWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof JointTrajectoryControllerStateArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "JointTrajectoryControllerStateArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      JointTrajectoryControllerStateWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return JointTrajectoryControllerStateWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return JointTrajectoryControllerStateArrayWrapper;
  }
}
module.exports = JointTrajectoryControllerStateWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "joint_names",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "desired",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "trajectory_msgs",
        "type": "JointTrajectoryPoint",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "actual",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "trajectory_msgs",
        "type": "JointTrajectoryPoint",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "error",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "trajectory_msgs",
        "type": "JointTrajectoryPoint",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "control_msgs",
    "type": "JointTrajectoryControllerState",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "JointTrajectoryControllerState"
}
*/
