// rclnodejs/generated/control_msgs/control_msgs__action__PointHead_Goal.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PointStampedWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__PointStamped.js');
const Vector3Wrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Vector3.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const DurationWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Duration.js');
const PointHead_GoalRefStruct = StructType({
  target: PointStampedWrapper.refObjectType,
  pointing_axis: Vector3Wrapper.refObjectType,
  pointing_frame: primitiveTypes.string,
  min_duration: DurationWrapper.refObjectType,
  max_velocity: primitiveTypes.float64,
});
const PointHead_GoalRefArray = ArrayType(PointHead_GoalRefStruct);
const PointHead_GoalRefStructArray = StructType({
  data: PointHead_GoalRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class PointHead_GoalWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new PointHead_GoalRefStruct(other._refObject.toObject());
      this._wrapperFields.target =  new PointStampedWrapper(other._wrapperFields.target);
      this._wrapperFields.pointing_axis =  new Vector3Wrapper(other._wrapperFields.pointing_axis);
      this._pointing_frameIntialized = true;
      this._wrapperFields.pointing_frame =  new StringWrapper(other._wrapperFields.pointing_frame);
      this._wrapperFields.min_duration =  new DurationWrapper(other._wrapperFields.min_duration);
      this._max_velocityIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new PointHead_GoalRefStruct();
    this._wrapperFields.target = new PointStampedWrapper();
    this._wrapperFields.pointing_axis = new Vector3Wrapper();
    this._pointing_frameIntialized = false;
    this._wrapperFields.pointing_frame = new StringWrapper();
    this._wrapperFields.min_duration = new DurationWrapper();
    this._max_velocityIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new PointHead_GoalWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new PointHead_GoalArrayWrapper;
  }
  static get ArrayType() {
    return PointHead_GoalArrayWrapper;
  }
  static get refObjectArrayType() {
    return PointHead_GoalRefStructArray
  }
  static get refObjectType() {
    return PointHead_GoalRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._pointing_frameIntialized) {
        throw new TypeError('Invalid argument: pointing_frame in PointHead_Goal');
      }
      if (!this._max_velocityIntialized) {
        throw new TypeError('Invalid argument: max_velocity in PointHead_Goal');
      }
    }
    this._wrapperFields.target.freeze(own, checkConsistency);
    this._refObject.target = this._wrapperFields.target.refObject;
    this._wrapperFields.pointing_axis.freeze(own, checkConsistency);
    this._refObject.pointing_axis = this._wrapperFields.pointing_axis.refObject;
    if (own) {
      this._wrapperFields.pointing_frame.freeze(own, checkConsistency);
    }
    this._refObject.pointing_frame = this._wrapperFields.pointing_frame.refObject;
    this._wrapperFields.min_duration.freeze(own, checkConsistency);
    this._refObject.min_duration = this._wrapperFields.min_duration.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.target.copyRefObject(refObject.target);
    this._wrapperFields.pointing_axis.copyRefObject(refObject.pointing_axis);
    this._pointing_frameIntialized = true;
    this._wrapperFields.pointing_frame.data = refObject.pointing_frame.data;
    this._wrapperFields.min_duration.copyRefObject(refObject.min_duration);
    this._max_velocityIntialized = true;
    this._refObject.max_velocity = refObject.max_velocity;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    PointStampedWrapper.freeStruct(refObject.target);
    Vector3Wrapper.freeStruct(refObject.pointing_axis);
    StringWrapper.freeStruct(refObject.pointing_frame);
    DurationWrapper.freeStruct(refObject.min_duration);
  }
  static destoryRawROS(msg) {
    PointHead_GoalWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'control_msgs', subFolder: 'action', interfaceName: 'PointHead_Goal'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get target() {
    return this._wrapperFields.target;
  }
  set target(value) {
    if (value instanceof PointStampedWrapper) {
      this._wrapperFields.target.copy(value);
    } else {
      this._wrapperFields.target.copy(new PointStampedWrapper(value));
    }
  }
  get pointing_axis() {
    return this._wrapperFields.pointing_axis;
  }
  set pointing_axis(value) {
    if (value instanceof Vector3Wrapper) {
      this._wrapperFields.pointing_axis.copy(value);
    } else {
      this._wrapperFields.pointing_axis.copy(new Vector3Wrapper(value));
    }
  }
  get pointing_frame() {
    if (!this._pointing_frameIntialized) {
      return undefined;
    }
    return this._wrapperFields.pointing_frame.data;
  }
  set pointing_frame(value) {
    this._pointing_frameIntialized = true;
    this._wrapperFields.pointing_frame.data = value;
  }
  get min_duration() {
    return this._wrapperFields.min_duration;
  }
  set min_duration(value) {
    if (value instanceof DurationWrapper) {
      this._wrapperFields.min_duration.copy(value);
    } else {
      this._wrapperFields.min_duration.copy(new DurationWrapper(value));
    }
  }
  get max_velocity() {
    if (!this._max_velocityIntialized) {
      return undefined;
    }
    return this._refObject.max_velocity;
  }
  set max_velocity(value) {
    this._max_velocityIntialized = true;
    this._refObject.max_velocity = value;
  }
  copyRefObject(refObject) {
    this._refObject = new PointHead_GoalRefStruct(refObject.toObject());
    this._wrapperFields.target.copyRefObject(this._refObject.target);
    this._wrapperFields.pointing_axis.copyRefObject(this._refObject.pointing_axis);
    this._pointing_frameIntialized = true;
    this._wrapperFields.pointing_frame.copyRefObject(this._refObject.pointing_frame);
    this._wrapperFields.min_duration.copyRefObject(this._refObject.min_duration);
    this._max_velocityIntialized = true;
  }
  copy(other) {
    this._refObject = new PointHead_GoalRefStruct(other._refObject.toObject());
    this._wrapperFields.target.copy(other._wrapperFields.target);
    this._wrapperFields.pointing_axis.copy(other._wrapperFields.pointing_axis);
    this._pointing_frameIntialized = true;
    this._wrapperFields.pointing_frame.copy(other._wrapperFields.pointing_frame);
    this._wrapperFields.min_duration.copy(other._wrapperFields.min_duration);
    this._max_velocityIntialized = true;
  }
  static get classType() {
    return PointHead_GoalWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"target","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"PointStamped","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"pointing_axis","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Vector3","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"pointing_frame","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"min_duration","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Duration","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"max_velocity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"control_msgs","type":"PointHead_Goal","stringUpperBound":null,"isPrimitiveType":false},"msgName":"PointHead_Goal"};
  }
  hasMember(name) {
    let memberNames = ["target","pointing_axis","pointing_frame","min_duration","max_velocity"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class PointHead_GoalArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof PointHead_GoalWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new PointHead_GoalWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PointHead_GoalArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PointHead_GoalArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new PointHead_GoalRefArray(size);
    this._refObject = new PointHead_GoalRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new PointHead_GoalWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof PointHead_GoalArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "PointHead_GoalArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      PointHead_GoalWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return PointHead_GoalWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return PointHead_GoalArrayWrapper;
  }
}
module.exports = PointHead_GoalWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "target",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "PointStamped",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "pointing_axis",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Vector3",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "pointing_frame",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "min_duration",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Duration",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "max_velocity",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "control_msgs",
    "type": "PointHead_Goal",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "PointHead_Goal"
}
*/
