// rclnodejs/generated/control_msgs/control_msgs__action__GripperCommand_Feedback.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const GripperCommand_FeedbackRefStruct = StructType({
  position: primitiveTypes.float64,
  effort: primitiveTypes.float64,
  stalled: primitiveTypes.bool,
  reached_goal: primitiveTypes.bool,
});
const GripperCommand_FeedbackRefArray = ArrayType(GripperCommand_FeedbackRefStruct);
const GripperCommand_FeedbackRefStructArray = StructType({
  data: GripperCommand_FeedbackRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GripperCommand_FeedbackWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GripperCommand_FeedbackRefStruct(other._refObject.toObject());
      this._positionIntialized = true;
      this._effortIntialized = true;
      this._stalledIntialized = true;
      this._reached_goalIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GripperCommand_FeedbackRefStruct();
    this._positionIntialized = false;
    this._effortIntialized = false;
    this._stalledIntialized = false;
    this._reached_goalIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new GripperCommand_FeedbackWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GripperCommand_FeedbackArrayWrapper;
  }
  static get ArrayType() {
    return GripperCommand_FeedbackArrayWrapper;
  }
  static get refObjectArrayType() {
    return GripperCommand_FeedbackRefStructArray
  }
  static get refObjectType() {
    return GripperCommand_FeedbackRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._positionIntialized) {
        throw new TypeError('Invalid argument: position in GripperCommand_Feedback');
      }
      if (!this._effortIntialized) {
        throw new TypeError('Invalid argument: effort in GripperCommand_Feedback');
      }
      if (!this._stalledIntialized) {
        throw new TypeError('Invalid argument: stalled in GripperCommand_Feedback');
      }
      if (!this._reached_goalIntialized) {
        throw new TypeError('Invalid argument: reached_goal in GripperCommand_Feedback');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._positionIntialized = true;
    this._refObject.position = refObject.position;
    this._effortIntialized = true;
    this._refObject.effort = refObject.effort;
    this._stalledIntialized = true;
    this._refObject.stalled = refObject.stalled;
    this._reached_goalIntialized = true;
    this._refObject.reached_goal = refObject.reached_goal;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    GripperCommand_FeedbackWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'control_msgs', subFolder: 'action', interfaceName: 'GripperCommand_Feedback'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get position() {
    if (!this._positionIntialized) {
      return undefined;
    }
    return this._refObject.position;
  }
  set position(value) {
    this._positionIntialized = true;
    this._refObject.position = value;
  }
  get effort() {
    if (!this._effortIntialized) {
      return undefined;
    }
    return this._refObject.effort;
  }
  set effort(value) {
    this._effortIntialized = true;
    this._refObject.effort = value;
  }
  get stalled() {
    if (!this._stalledIntialized) {
      return undefined;
    }
    return this._refObject.stalled;
  }
  set stalled(value) {
    this._stalledIntialized = true;
    this._refObject.stalled = value;
  }
  get reached_goal() {
    if (!this._reached_goalIntialized) {
      return undefined;
    }
    return this._refObject.reached_goal;
  }
  set reached_goal(value) {
    this._reached_goalIntialized = true;
    this._refObject.reached_goal = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GripperCommand_FeedbackRefStruct(refObject.toObject());
    this._positionIntialized = true;
    this._effortIntialized = true;
    this._stalledIntialized = true;
    this._reached_goalIntialized = true;
  }
  copy(other) {
    this._refObject = new GripperCommand_FeedbackRefStruct(other._refObject.toObject());
    this._positionIntialized = true;
    this._effortIntialized = true;
    this._stalledIntialized = true;
    this._reached_goalIntialized = true;
  }
  static get classType() {
    return GripperCommand_FeedbackWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"position","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"effort","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"stalled","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reached_goal","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"control_msgs","type":"GripperCommand_Feedback","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GripperCommand_Feedback"};
  }
  hasMember(name) {
    let memberNames = ["position","effort","stalled","reached_goal"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GripperCommand_FeedbackArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GripperCommand_FeedbackWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GripperCommand_FeedbackWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GripperCommand_FeedbackArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GripperCommand_FeedbackArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GripperCommand_FeedbackRefArray(size);
    this._refObject = new GripperCommand_FeedbackRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GripperCommand_FeedbackWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GripperCommand_FeedbackArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GripperCommand_FeedbackArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GripperCommand_FeedbackWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GripperCommand_FeedbackWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GripperCommand_FeedbackArrayWrapper;
  }
}
module.exports = GripperCommand_FeedbackWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "position",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "effort",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "stalled",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reached_goal",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "control_msgs",
    "type": "GripperCommand_Feedback",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GripperCommand_Feedback"
}
*/
