// rclnodejs/generated/control_msgs/control_msgs__action__JointTrajectory_Goal.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const JointTrajectoryWrapper = require('../../generated/trajectory_msgs/trajectory_msgs__msg__JointTrajectory.js');
const JointTrajectory_GoalRefStruct = StructType({
  trajectory: JointTrajectoryWrapper.refObjectType,
});
const JointTrajectory_GoalRefArray = ArrayType(JointTrajectory_GoalRefStruct);
const JointTrajectory_GoalRefStructArray = StructType({
  data: JointTrajectory_GoalRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class JointTrajectory_GoalWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new JointTrajectory_GoalRefStruct(other._refObject.toObject());
      this._wrapperFields.trajectory =  new JointTrajectoryWrapper(other._wrapperFields.trajectory);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new JointTrajectory_GoalRefStruct();
    this._wrapperFields.trajectory = new JointTrajectoryWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new JointTrajectory_GoalWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new JointTrajectory_GoalArrayWrapper;
  }
  static get ArrayType() {
    return JointTrajectory_GoalArrayWrapper;
  }
  static get refObjectArrayType() {
    return JointTrajectory_GoalRefStructArray
  }
  static get refObjectType() {
    return JointTrajectory_GoalRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.trajectory.freeze(own, checkConsistency);
    this._refObject.trajectory = this._wrapperFields.trajectory.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.trajectory.copyRefObject(refObject.trajectory);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    JointTrajectoryWrapper.freeStruct(refObject.trajectory);
  }
  static destoryRawROS(msg) {
    JointTrajectory_GoalWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'control_msgs', subFolder: 'action', interfaceName: 'JointTrajectory_Goal'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get trajectory() {
    return this._wrapperFields.trajectory;
  }
  set trajectory(value) {
    if (value instanceof JointTrajectoryWrapper) {
      this._wrapperFields.trajectory.copy(value);
    } else {
      this._wrapperFields.trajectory.copy(new JointTrajectoryWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new JointTrajectory_GoalRefStruct(refObject.toObject());
    this._wrapperFields.trajectory.copyRefObject(this._refObject.trajectory);
  }
  copy(other) {
    this._refObject = new JointTrajectory_GoalRefStruct(other._refObject.toObject());
    this._wrapperFields.trajectory.copy(other._wrapperFields.trajectory);
  }
  static get classType() {
    return JointTrajectory_GoalWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"trajectory","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"trajectory_msgs","type":"JointTrajectory","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"control_msgs","type":"JointTrajectory_Goal","stringUpperBound":null,"isPrimitiveType":false},"msgName":"JointTrajectory_Goal"};
  }
  hasMember(name) {
    let memberNames = ["trajectory"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class JointTrajectory_GoalArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof JointTrajectory_GoalWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new JointTrajectory_GoalWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to JointTrajectory_GoalArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to JointTrajectory_GoalArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new JointTrajectory_GoalRefArray(size);
    this._refObject = new JointTrajectory_GoalRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new JointTrajectory_GoalWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof JointTrajectory_GoalArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "JointTrajectory_GoalArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      JointTrajectory_GoalWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return JointTrajectory_GoalWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return JointTrajectory_GoalArrayWrapper;
  }
}
module.exports = JointTrajectory_GoalWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "trajectory",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "trajectory_msgs",
        "type": "JointTrajectory",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "control_msgs",
    "type": "JointTrajectory_Goal",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "JointTrajectory_Goal"
}
*/
