// rclnodejs/generated/nav2_msgs/nav2_msgs__action__NavigateToPose_Goal.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PoseStampedWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__PoseStamped.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const NavigateToPose_GoalRefStruct = StructType({
  pose: PoseStampedWrapper.refObjectType,
  behavior_tree: primitiveTypes.string,
});
const NavigateToPose_GoalRefArray = ArrayType(NavigateToPose_GoalRefStruct);
const NavigateToPose_GoalRefStructArray = StructType({
  data: NavigateToPose_GoalRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class NavigateToPose_GoalWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new NavigateToPose_GoalRefStruct(other._refObject.toObject());
      this._wrapperFields.pose =  new PoseStampedWrapper(other._wrapperFields.pose);
      this._behavior_treeIntialized = true;
      this._wrapperFields.behavior_tree =  new StringWrapper(other._wrapperFields.behavior_tree);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new NavigateToPose_GoalRefStruct();
    this._wrapperFields.pose = new PoseStampedWrapper();
    this._behavior_treeIntialized = false;
    this._wrapperFields.behavior_tree = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new NavigateToPose_GoalWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new NavigateToPose_GoalArrayWrapper;
  }
  static get ArrayType() {
    return NavigateToPose_GoalArrayWrapper;
  }
  static get refObjectArrayType() {
    return NavigateToPose_GoalRefStructArray
  }
  static get refObjectType() {
    return NavigateToPose_GoalRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._behavior_treeIntialized) {
        throw new TypeError('Invalid argument: behavior_tree in NavigateToPose_Goal');
      }
    }
    this._wrapperFields.pose.freeze(own, checkConsistency);
    this._refObject.pose = this._wrapperFields.pose.refObject;
    if (own) {
      this._wrapperFields.behavior_tree.freeze(own, checkConsistency);
    }
    this._refObject.behavior_tree = this._wrapperFields.behavior_tree.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.pose.copyRefObject(refObject.pose);
    this._behavior_treeIntialized = true;
    this._wrapperFields.behavior_tree.data = refObject.behavior_tree.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    PoseStampedWrapper.freeStruct(refObject.pose);
    StringWrapper.freeStruct(refObject.behavior_tree);
  }
  static destoryRawROS(msg) {
    NavigateToPose_GoalWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'action', interfaceName: 'NavigateToPose_Goal'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get pose() {
    return this._wrapperFields.pose;
  }
  set pose(value) {
    if (value instanceof PoseStampedWrapper) {
      this._wrapperFields.pose.copy(value);
    } else {
      this._wrapperFields.pose.copy(new PoseStampedWrapper(value));
    }
  }
  get behavior_tree() {
    if (!this._behavior_treeIntialized) {
      return undefined;
    }
    return this._wrapperFields.behavior_tree.data;
  }
  set behavior_tree(value) {
    this._behavior_treeIntialized = true;
    this._wrapperFields.behavior_tree.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new NavigateToPose_GoalRefStruct(refObject.toObject());
    this._wrapperFields.pose.copyRefObject(this._refObject.pose);
    this._behavior_treeIntialized = true;
    this._wrapperFields.behavior_tree.copyRefObject(this._refObject.behavior_tree);
  }
  copy(other) {
    this._refObject = new NavigateToPose_GoalRefStruct(other._refObject.toObject());
    this._wrapperFields.pose.copy(other._wrapperFields.pose);
    this._behavior_treeIntialized = true;
    this._wrapperFields.behavior_tree.copy(other._wrapperFields.behavior_tree);
  }
  static get classType() {
    return NavigateToPose_GoalWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"pose","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"PoseStamped","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"behavior_tree","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"NavigateToPose_Goal","stringUpperBound":null,"isPrimitiveType":false},"msgName":"NavigateToPose_Goal"};
  }
  hasMember(name) {
    let memberNames = ["pose","behavior_tree"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class NavigateToPose_GoalArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof NavigateToPose_GoalWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new NavigateToPose_GoalWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to NavigateToPose_GoalArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to NavigateToPose_GoalArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new NavigateToPose_GoalRefArray(size);
    this._refObject = new NavigateToPose_GoalRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new NavigateToPose_GoalWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof NavigateToPose_GoalArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "NavigateToPose_GoalArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      NavigateToPose_GoalWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return NavigateToPose_GoalWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return NavigateToPose_GoalArrayWrapper;
  }
}
module.exports = NavigateToPose_GoalWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "pose",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "PoseStamped",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "behavior_tree",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "nav2_msgs",
    "type": "NavigateToPose_Goal",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "NavigateToPose_Goal"
}
*/
