// rclnodejs/generated/nav2_msgs/nav2_msgs__action__FollowPath_Goal.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PathWrapper = require('../../generated/nav_msgs/nav_msgs__msg__Path.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const FollowPath_GoalRefStruct = StructType({
  path: PathWrapper.refObjectType,
  controller_id: primitiveTypes.string,
});
const FollowPath_GoalRefArray = ArrayType(FollowPath_GoalRefStruct);
const FollowPath_GoalRefStructArray = StructType({
  data: FollowPath_GoalRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class FollowPath_GoalWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new FollowPath_GoalRefStruct(other._refObject.toObject());
      this._wrapperFields.path =  new PathWrapper(other._wrapperFields.path);
      this._controller_idIntialized = true;
      this._wrapperFields.controller_id =  new StringWrapper(other._wrapperFields.controller_id);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new FollowPath_GoalRefStruct();
    this._wrapperFields.path = new PathWrapper();
    this._controller_idIntialized = false;
    this._wrapperFields.controller_id = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new FollowPath_GoalWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new FollowPath_GoalArrayWrapper;
  }
  static get ArrayType() {
    return FollowPath_GoalArrayWrapper;
  }
  static get refObjectArrayType() {
    return FollowPath_GoalRefStructArray
  }
  static get refObjectType() {
    return FollowPath_GoalRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._controller_idIntialized) {
        throw new TypeError('Invalid argument: controller_id in FollowPath_Goal');
      }
    }
    this._wrapperFields.path.freeze(own, checkConsistency);
    this._refObject.path = this._wrapperFields.path.refObject;
    if (own) {
      this._wrapperFields.controller_id.freeze(own, checkConsistency);
    }
    this._refObject.controller_id = this._wrapperFields.controller_id.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.path.copyRefObject(refObject.path);
    this._controller_idIntialized = true;
    this._wrapperFields.controller_id.data = refObject.controller_id.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    PathWrapper.freeStruct(refObject.path);
    StringWrapper.freeStruct(refObject.controller_id);
  }
  static destoryRawROS(msg) {
    FollowPath_GoalWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'action', interfaceName: 'FollowPath_Goal'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get path() {
    return this._wrapperFields.path;
  }
  set path(value) {
    if (value instanceof PathWrapper) {
      this._wrapperFields.path.copy(value);
    } else {
      this._wrapperFields.path.copy(new PathWrapper(value));
    }
  }
  get controller_id() {
    if (!this._controller_idIntialized) {
      return undefined;
    }
    return this._wrapperFields.controller_id.data;
  }
  set controller_id(value) {
    this._controller_idIntialized = true;
    this._wrapperFields.controller_id.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new FollowPath_GoalRefStruct(refObject.toObject());
    this._wrapperFields.path.copyRefObject(this._refObject.path);
    this._controller_idIntialized = true;
    this._wrapperFields.controller_id.copyRefObject(this._refObject.controller_id);
  }
  copy(other) {
    this._refObject = new FollowPath_GoalRefStruct(other._refObject.toObject());
    this._wrapperFields.path.copy(other._wrapperFields.path);
    this._controller_idIntialized = true;
    this._wrapperFields.controller_id.copy(other._wrapperFields.controller_id);
  }
  static get classType() {
    return FollowPath_GoalWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"path","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"nav_msgs","type":"Path","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"controller_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"FollowPath_Goal","stringUpperBound":null,"isPrimitiveType":false},"msgName":"FollowPath_Goal"};
  }
  hasMember(name) {
    let memberNames = ["path","controller_id"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class FollowPath_GoalArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof FollowPath_GoalWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new FollowPath_GoalWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FollowPath_GoalArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FollowPath_GoalArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new FollowPath_GoalRefArray(size);
    this._refObject = new FollowPath_GoalRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new FollowPath_GoalWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof FollowPath_GoalArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "FollowPath_GoalArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      FollowPath_GoalWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return FollowPath_GoalWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return FollowPath_GoalArrayWrapper;
  }
}
module.exports = FollowPath_GoalWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "path",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "nav_msgs",
        "type": "Path",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "controller_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "nav2_msgs",
    "type": "FollowPath_Goal",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "FollowPath_Goal"
}
*/
