// rclnodejs/generated/nav2_msgs/nav2_msgs__action__FollowWaypoints_Goal.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PoseStampedWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__PoseStamped.js');
const FollowWaypoints_GoalRefStruct = StructType({
  poses: PoseStampedWrapper.refObjectArrayType,
});
const FollowWaypoints_GoalRefArray = ArrayType(FollowWaypoints_GoalRefStruct);
const FollowWaypoints_GoalRefStructArray = StructType({
  data: FollowWaypoints_GoalRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class FollowWaypoints_GoalWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new FollowWaypoints_GoalRefStruct(other._refObject.toObject());
      this._wrapperFields.poses = PoseStampedWrapper.createArray();
      this._wrapperFields.poses.copy(other._wrapperFields.poses);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new FollowWaypoints_GoalRefStruct();
    this._wrapperFields.poses = PoseStampedWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new FollowWaypoints_GoalWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new FollowWaypoints_GoalArrayWrapper;
  }
  static get ArrayType() {
    return FollowWaypoints_GoalArrayWrapper;
  }
  static get refObjectArrayType() {
    return FollowWaypoints_GoalRefStructArray
  }
  static get refObjectType() {
    return FollowWaypoints_GoalRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.poses.freeze(own, checkConsistency);
    this._refObject.poses = this._wrapperFields.poses.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.poses.copyRefObject(refObject.poses);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.poses.size != 0) {
      PoseStampedWrapper.ArrayType.freeArray(refObject.poses);
      if (PoseStampedWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.poses, PoseStampedWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    FollowWaypoints_GoalWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'action', interfaceName: 'FollowWaypoints_Goal'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get poses() {
    return this._wrapperFields.poses;
  }
  set poses(value) {
    this._wrapperFields.poses.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new FollowWaypoints_GoalRefStruct(refObject.toObject());
    this._wrapperFields.poses.copyRefObject(this._refObject.poses);
  }
  copy(other) {
    this._refObject = new FollowWaypoints_GoalRefStruct(other._refObject.toObject());
    this._wrapperFields.poses.copy(other._wrapperFields.poses);
  }
  static get classType() {
    return FollowWaypoints_GoalWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"poses","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"geometry_msgs","type":"PoseStamped","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"FollowWaypoints_Goal","stringUpperBound":null,"isPrimitiveType":false},"msgName":"FollowWaypoints_Goal"};
  }
  hasMember(name) {
    let memberNames = ["poses"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class FollowWaypoints_GoalArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof FollowWaypoints_GoalWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new FollowWaypoints_GoalWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FollowWaypoints_GoalArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FollowWaypoints_GoalArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new FollowWaypoints_GoalRefArray(size);
    this._refObject = new FollowWaypoints_GoalRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new FollowWaypoints_GoalWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof FollowWaypoints_GoalArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "FollowWaypoints_GoalArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      FollowWaypoints_GoalWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return FollowWaypoints_GoalWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return FollowWaypoints_GoalArrayWrapper;
  }
}
module.exports = FollowWaypoints_GoalWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "poses",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "geometry_msgs",
        "type": "PoseStamped",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "nav2_msgs",
    "type": "FollowWaypoints_Goal",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "FollowWaypoints_Goal"
}
*/
