// rclnodejs/generated/nav2_msgs/nav2_msgs__action__BackUp_Goal.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PointWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Point.js');
const BackUp_GoalRefStruct = StructType({
  target: PointWrapper.refObjectType,
  speed: primitiveTypes.float32,
});
const BackUp_GoalRefArray = ArrayType(BackUp_GoalRefStruct);
const BackUp_GoalRefStructArray = StructType({
  data: BackUp_GoalRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class BackUp_GoalWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new BackUp_GoalRefStruct(other._refObject.toObject());
      this._wrapperFields.target =  new PointWrapper(other._wrapperFields.target);
      this._speedIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new BackUp_GoalRefStruct();
    this._wrapperFields.target = new PointWrapper();
    this._speedIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new BackUp_GoalWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new BackUp_GoalArrayWrapper;
  }
  static get ArrayType() {
    return BackUp_GoalArrayWrapper;
  }
  static get refObjectArrayType() {
    return BackUp_GoalRefStructArray
  }
  static get refObjectType() {
    return BackUp_GoalRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._speedIntialized) {
        throw new TypeError('Invalid argument: speed in BackUp_Goal');
      }
    }
    this._wrapperFields.target.freeze(own, checkConsistency);
    this._refObject.target = this._wrapperFields.target.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.target.copyRefObject(refObject.target);
    this._speedIntialized = true;
    this._refObject.speed = refObject.speed;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    PointWrapper.freeStruct(refObject.target);
  }
  static destoryRawROS(msg) {
    BackUp_GoalWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'action', interfaceName: 'BackUp_Goal'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get target() {
    return this._wrapperFields.target;
  }
  set target(value) {
    if (value instanceof PointWrapper) {
      this._wrapperFields.target.copy(value);
    } else {
      this._wrapperFields.target.copy(new PointWrapper(value));
    }
  }
  get speed() {
    if (!this._speedIntialized) {
      return undefined;
    }
    return this._refObject.speed;
  }
  set speed(value) {
    this._speedIntialized = true;
    this._refObject.speed = value;
  }
  copyRefObject(refObject) {
    this._refObject = new BackUp_GoalRefStruct(refObject.toObject());
    this._wrapperFields.target.copyRefObject(this._refObject.target);
    this._speedIntialized = true;
  }
  copy(other) {
    this._refObject = new BackUp_GoalRefStruct(other._refObject.toObject());
    this._wrapperFields.target.copy(other._wrapperFields.target);
    this._speedIntialized = true;
  }
  static get classType() {
    return BackUp_GoalWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"target","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Point","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"speed","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"BackUp_Goal","stringUpperBound":null,"isPrimitiveType":false},"msgName":"BackUp_Goal"};
  }
  hasMember(name) {
    let memberNames = ["target","speed"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class BackUp_GoalArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof BackUp_GoalWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new BackUp_GoalWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to BackUp_GoalArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to BackUp_GoalArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new BackUp_GoalRefArray(size);
    this._refObject = new BackUp_GoalRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new BackUp_GoalWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof BackUp_GoalArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "BackUp_GoalArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      BackUp_GoalWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return BackUp_GoalWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return BackUp_GoalArrayWrapper;
  }
}
module.exports = BackUp_GoalWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "target",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Point",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "speed",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "nav2_msgs",
    "type": "BackUp_Goal",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "BackUp_Goal"
}
*/
