// rclnodejs/generated/nav2_msgs/nav2_msgs__msg__VoxelGrid.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const UInt32Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt32.js');
const Point32Wrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Point32.js');
const Vector3Wrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Vector3.js');
const VoxelGridRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  data: UInt32Wrapper.refObjectArrayType,
  origin: Point32Wrapper.refObjectType,
  resolutions: Vector3Wrapper.refObjectType,
  size_x: primitiveTypes.uint32,
  size_y: primitiveTypes.uint32,
  size_z: primitiveTypes.uint32,
});
const VoxelGridRefArray = ArrayType(VoxelGridRefStruct);
const VoxelGridRefStructArray = StructType({
  data: VoxelGridRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class VoxelGridWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new VoxelGridRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.data = UInt32Wrapper.createArray();
      this._wrapperFields.data.copy(other._wrapperFields.data);
      this._wrapperFields.origin =  new Point32Wrapper(other._wrapperFields.origin);
      this._wrapperFields.resolutions =  new Vector3Wrapper(other._wrapperFields.resolutions);
      this._size_xIntialized = true;
      this._size_yIntialized = true;
      this._size_zIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new VoxelGridRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.data = UInt32Wrapper.createArray();
    this._wrapperFields.origin = new Point32Wrapper();
    this._wrapperFields.resolutions = new Vector3Wrapper();
    this._size_xIntialized = false;
    this._size_yIntialized = false;
    this._size_zIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new VoxelGridWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new VoxelGridArrayWrapper;
  }
  static get ArrayType() {
    return VoxelGridArrayWrapper;
  }
  static get refObjectArrayType() {
    return VoxelGridRefStructArray
  }
  static get refObjectType() {
    return VoxelGridRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._size_xIntialized) {
        throw new TypeError('Invalid argument: size_x in VoxelGrid');
      }
      if (!this._size_yIntialized) {
        throw new TypeError('Invalid argument: size_y in VoxelGrid');
      }
      if (!this._size_zIntialized) {
        throw new TypeError('Invalid argument: size_z in VoxelGrid');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.data.freeze(own, checkConsistency);
    this._refObject.data = this._wrapperFields.data.refObject;
    if (own) {
      this._wrapperFields.data.fill(Uint32Array.from([]));
      this._wrapperFields.data.freeze(own, checkConsistency);
      this._refObject.data = this._wrapperFields.data.refObject;
    }
    this._wrapperFields.origin.freeze(own, checkConsistency);
    this._refObject.origin = this._wrapperFields.origin.refObject;
    this._wrapperFields.resolutions.freeze(own, checkConsistency);
    this._refObject.resolutions = this._wrapperFields.resolutions.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.data.copyRefObject(refObject.data);
    this._wrapperFields.origin.copyRefObject(refObject.origin);
    this._wrapperFields.resolutions.copyRefObject(refObject.resolutions);
    this._size_xIntialized = true;
    this._refObject.size_x = refObject.size_x;
    this._size_yIntialized = true;
    this._refObject.size_y = refObject.size_y;
    this._size_zIntialized = true;
    this._refObject.size_z = refObject.size_z;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.data.size != 0) {
      UInt32Wrapper.ArrayType.freeArray(refObject.data);
      if (UInt32Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.data, UInt32Wrapper.refObjectArrayType, 'data');
      }
    }
    Point32Wrapper.freeStruct(refObject.origin);
    Vector3Wrapper.freeStruct(refObject.resolutions);
  }
  static destoryRawROS(msg) {
    VoxelGridWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'msg', interfaceName: 'VoxelGrid'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get data() {
    return this._wrapperFields['data'].data;
  }
  set data(value) {
    this._wrapperFields['data'].fill(value);
  }
  get origin() {
    return this._wrapperFields.origin;
  }
  set origin(value) {
    if (value instanceof Point32Wrapper) {
      this._wrapperFields.origin.copy(value);
    } else {
      this._wrapperFields.origin.copy(new Point32Wrapper(value));
    }
  }
  get resolutions() {
    return this._wrapperFields.resolutions;
  }
  set resolutions(value) {
    if (value instanceof Vector3Wrapper) {
      this._wrapperFields.resolutions.copy(value);
    } else {
      this._wrapperFields.resolutions.copy(new Vector3Wrapper(value));
    }
  }
  get size_x() {
    if (!this._size_xIntialized) {
      return undefined;
    }
    return this._refObject.size_x;
  }
  set size_x(value) {
    this._size_xIntialized = true;
    this._refObject.size_x = value;
  }
  get size_y() {
    if (!this._size_yIntialized) {
      return undefined;
    }
    return this._refObject.size_y;
  }
  set size_y(value) {
    this._size_yIntialized = true;
    this._refObject.size_y = value;
  }
  get size_z() {
    if (!this._size_zIntialized) {
      return undefined;
    }
    return this._refObject.size_z;
  }
  set size_z(value) {
    this._size_zIntialized = true;
    this._refObject.size_z = value;
  }
  copyRefObject(refObject) {
    this._refObject = new VoxelGridRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.data.copyRefObject(this._refObject.data);
    this._wrapperFields.origin.copyRefObject(this._refObject.origin);
    this._wrapperFields.resolutions.copyRefObject(this._refObject.resolutions);
    this._size_xIntialized = true;
    this._size_yIntialized = true;
    this._size_zIntialized = true;
  }
  copy(other) {
    this._refObject = new VoxelGridRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.data.copy(other._wrapperFields.data);
    this._wrapperFields.origin.copy(other._wrapperFields.origin);
    this._wrapperFields.resolutions.copy(other._wrapperFields.resolutions);
    this._size_xIntialized = true;
    this._size_yIntialized = true;
    this._size_zIntialized = true;
  }
  static get classType() {
    return VoxelGridWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"data","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"origin","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Point32","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"resolutions","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Vector3","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"size_x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"size_y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"size_z","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"VoxelGrid","stringUpperBound":null,"isPrimitiveType":false},"msgName":"VoxelGrid"};
  }
  hasMember(name) {
    let memberNames = ["header","data","origin","resolutions","size_x","size_y","size_z"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class VoxelGridArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof VoxelGridWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new VoxelGridWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VoxelGridArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VoxelGridArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new VoxelGridRefArray(size);
    this._refObject = new VoxelGridRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new VoxelGridWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof VoxelGridArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "VoxelGridArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      VoxelGridWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return VoxelGridWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return VoxelGridArrayWrapper;
  }
}
module.exports = VoxelGridWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "data",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "origin",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Point32",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "resolutions",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Vector3",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "size_x",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "size_y",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "size_z",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "nav2_msgs",
    "type": "VoxelGrid",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "VoxelGrid"
}
*/
