// rclnodejs/generated/nav2_msgs/nav2_msgs__srv__GetCostmap_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const CostmapMetaDataWrapper = require('../../generated/nav2_msgs/nav2_msgs__msg__CostmapMetaData.js');
const GetCostmap_RequestRefStruct = StructType({
  specs: CostmapMetaDataWrapper.refObjectType,
});
const GetCostmap_RequestRefArray = ArrayType(GetCostmap_RequestRefStruct);
const GetCostmap_RequestRefStructArray = StructType({
  data: GetCostmap_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GetCostmap_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetCostmap_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.specs =  new CostmapMetaDataWrapper(other._wrapperFields.specs);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GetCostmap_RequestRefStruct();
    this._wrapperFields.specs = new CostmapMetaDataWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GetCostmap_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetCostmap_RequestArrayWrapper;
  }
  static get ArrayType() {
    return GetCostmap_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetCostmap_RequestRefStructArray
  }
  static get refObjectType() {
    return GetCostmap_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.specs.freeze(own, checkConsistency);
    this._refObject.specs = this._wrapperFields.specs.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.specs.copyRefObject(refObject.specs);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    CostmapMetaDataWrapper.freeStruct(refObject.specs);
  }
  static destoryRawROS(msg) {
    GetCostmap_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'srv', interfaceName: 'GetCostmap_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get specs() {
    return this._wrapperFields.specs;
  }
  set specs(value) {
    if (value instanceof CostmapMetaDataWrapper) {
      this._wrapperFields.specs.copy(value);
    } else {
      this._wrapperFields.specs.copy(new CostmapMetaDataWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new GetCostmap_RequestRefStruct(refObject.toObject());
    this._wrapperFields.specs.copyRefObject(this._refObject.specs);
  }
  copy(other) {
    this._refObject = new GetCostmap_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.specs.copy(other._wrapperFields.specs);
  }
  static get classType() {
    return GetCostmap_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"specs","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"nav2_msgs","type":"CostmapMetaData","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"GetCostmap_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetCostmap_Request"};
  }
  hasMember(name) {
    let memberNames = ["specs"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GetCostmap_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetCostmap_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetCostmap_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetCostmap_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetCostmap_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GetCostmap_RequestRefArray(size);
    this._refObject = new GetCostmap_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetCostmap_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetCostmap_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetCostmap_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetCostmap_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetCostmap_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetCostmap_RequestArrayWrapper;
  }
}
module.exports = GetCostmap_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "specs",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "nav2_msgs",
        "type": "CostmapMetaData",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "nav2_msgs",
    "type": "GetCostmap_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GetCostmap_Request"
}
*/
