// rclnodejs/generated/nav2_msgs/nav2_msgs__msg__BehaviorTreeLog.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TimeWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Time.js');
const BehaviorTreeStatusChangeWrapper = require('../../generated/nav2_msgs/nav2_msgs__msg__BehaviorTreeStatusChange.js');
const BehaviorTreeLogRefStruct = StructType({
  timestamp: TimeWrapper.refObjectType,
  event_log: BehaviorTreeStatusChangeWrapper.refObjectArrayType,
});
const BehaviorTreeLogRefArray = ArrayType(BehaviorTreeLogRefStruct);
const BehaviorTreeLogRefStructArray = StructType({
  data: BehaviorTreeLogRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class BehaviorTreeLogWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new BehaviorTreeLogRefStruct(other._refObject.toObject());
      this._wrapperFields.timestamp =  new TimeWrapper(other._wrapperFields.timestamp);
      this._wrapperFields.event_log = BehaviorTreeStatusChangeWrapper.createArray();
      this._wrapperFields.event_log.copy(other._wrapperFields.event_log);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new BehaviorTreeLogRefStruct();
    this._wrapperFields.timestamp = new TimeWrapper();
    this._wrapperFields.event_log = BehaviorTreeStatusChangeWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new BehaviorTreeLogWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new BehaviorTreeLogArrayWrapper;
  }
  static get ArrayType() {
    return BehaviorTreeLogArrayWrapper;
  }
  static get refObjectArrayType() {
    return BehaviorTreeLogRefStructArray
  }
  static get refObjectType() {
    return BehaviorTreeLogRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.timestamp.freeze(own, checkConsistency);
    this._refObject.timestamp = this._wrapperFields.timestamp.refObject;
    this._wrapperFields.event_log.freeze(own, checkConsistency);
    this._refObject.event_log = this._wrapperFields.event_log.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.timestamp.copyRefObject(refObject.timestamp);
    this._wrapperFields.event_log.copyRefObject(refObject.event_log);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    TimeWrapper.freeStruct(refObject.timestamp);
    if (refObject.event_log.size != 0) {
      BehaviorTreeStatusChangeWrapper.ArrayType.freeArray(refObject.event_log);
      if (BehaviorTreeStatusChangeWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.event_log, BehaviorTreeStatusChangeWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    BehaviorTreeLogWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'msg', interfaceName: 'BehaviorTreeLog'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    return this._wrapperFields.timestamp;
  }
  set timestamp(value) {
    if (value instanceof TimeWrapper) {
      this._wrapperFields.timestamp.copy(value);
    } else {
      this._wrapperFields.timestamp.copy(new TimeWrapper(value));
    }
  }
  get event_log() {
    return this._wrapperFields.event_log;
  }
  set event_log(value) {
    this._wrapperFields.event_log.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new BehaviorTreeLogRefStruct(refObject.toObject());
    this._wrapperFields.timestamp.copyRefObject(this._refObject.timestamp);
    this._wrapperFields.event_log.copyRefObject(this._refObject.event_log);
  }
  copy(other) {
    this._refObject = new BehaviorTreeLogRefStruct(other._refObject.toObject());
    this._wrapperFields.timestamp.copy(other._wrapperFields.timestamp);
    this._wrapperFields.event_log.copy(other._wrapperFields.event_log);
  }
  static get classType() {
    return BehaviorTreeLogWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"event_log","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"nav2_msgs","type":"BehaviorTreeStatusChange","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"BehaviorTreeLog","stringUpperBound":null,"isPrimitiveType":false},"msgName":"BehaviorTreeLog"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","event_log"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class BehaviorTreeLogArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof BehaviorTreeLogWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new BehaviorTreeLogWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to BehaviorTreeLogArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to BehaviorTreeLogArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new BehaviorTreeLogRefArray(size);
    this._refObject = new BehaviorTreeLogRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new BehaviorTreeLogWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof BehaviorTreeLogArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "BehaviorTreeLogArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      BehaviorTreeLogWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return BehaviorTreeLogWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return BehaviorTreeLogArrayWrapper;
  }
}
module.exports = BehaviorTreeLogWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Time",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "event_log",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "nav2_msgs",
        "type": "BehaviorTreeStatusChange",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "nav2_msgs",
    "type": "BehaviorTreeLog",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "BehaviorTreeLog"
}
*/
