// rclnodejs/generated/nav2_msgs/nav2_msgs__srv__GetCostmap_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const CostmapWrapper = require('../../generated/nav2_msgs/nav2_msgs__msg__Costmap.js');
const GetCostmap_ResponseRefStruct = StructType({
  map: CostmapWrapper.refObjectType,
});
const GetCostmap_ResponseRefArray = ArrayType(GetCostmap_ResponseRefStruct);
const GetCostmap_ResponseRefStructArray = StructType({
  data: GetCostmap_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GetCostmap_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetCostmap_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.map =  new CostmapWrapper(other._wrapperFields.map);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GetCostmap_ResponseRefStruct();
    this._wrapperFields.map = new CostmapWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GetCostmap_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetCostmap_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return GetCostmap_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetCostmap_ResponseRefStructArray
  }
  static get refObjectType() {
    return GetCostmap_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.map.freeze(own, checkConsistency);
    this._refObject.map = this._wrapperFields.map.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.map.copyRefObject(refObject.map);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    CostmapWrapper.freeStruct(refObject.map);
  }
  static destoryRawROS(msg) {
    GetCostmap_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'srv', interfaceName: 'GetCostmap_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get map() {
    return this._wrapperFields.map;
  }
  set map(value) {
    if (value instanceof CostmapWrapper) {
      this._wrapperFields.map.copy(value);
    } else {
      this._wrapperFields.map.copy(new CostmapWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new GetCostmap_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.map.copyRefObject(this._refObject.map);
  }
  copy(other) {
    this._refObject = new GetCostmap_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.map.copy(other._wrapperFields.map);
  }
  static get classType() {
    return GetCostmap_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"map","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"nav2_msgs","type":"Costmap","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"GetCostmap_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetCostmap_Response"};
  }
  hasMember(name) {
    let memberNames = ["map"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GetCostmap_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetCostmap_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetCostmap_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetCostmap_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetCostmap_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GetCostmap_ResponseRefArray(size);
    this._refObject = new GetCostmap_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetCostmap_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetCostmap_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetCostmap_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetCostmap_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetCostmap_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetCostmap_ResponseArrayWrapper;
  }
}
module.exports = GetCostmap_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "map",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "nav2_msgs",
        "type": "Costmap",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "nav2_msgs",
    "type": "GetCostmap_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GetCostmap_Response"
}
*/
