// rclnodejs/generated/nav2_msgs/nav2_msgs__srv__LoadMap_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const LoadMap_RequestRefStruct = StructType({
  map_url: primitiveTypes.string,
});
const LoadMap_RequestRefArray = ArrayType(LoadMap_RequestRefStruct);
const LoadMap_RequestRefStructArray = StructType({
  data: LoadMap_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class LoadMap_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new LoadMap_RequestRefStruct(other._refObject.toObject());
      this._map_urlIntialized = true;
      this._wrapperFields.map_url =  new StringWrapper(other._wrapperFields.map_url);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new LoadMap_RequestRefStruct();
    this._map_urlIntialized = false;
    this._wrapperFields.map_url = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new LoadMap_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LoadMap_RequestArrayWrapper;
  }
  static get ArrayType() {
    return LoadMap_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return LoadMap_RequestRefStructArray
  }
  static get refObjectType() {
    return LoadMap_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._map_urlIntialized) {
        throw new TypeError('Invalid argument: map_url in LoadMap_Request');
      }
    }
    if (own) {
      this._wrapperFields.map_url.freeze(own, checkConsistency);
    }
    this._refObject.map_url = this._wrapperFields.map_url.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._map_urlIntialized = true;
    this._wrapperFields.map_url.data = refObject.map_url.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.map_url);
  }
  static destoryRawROS(msg) {
    LoadMap_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'srv', interfaceName: 'LoadMap_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get map_url() {
    if (!this._map_urlIntialized) {
      return undefined;
    }
    return this._wrapperFields.map_url.data;
  }
  set map_url(value) {
    this._map_urlIntialized = true;
    this._wrapperFields.map_url.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new LoadMap_RequestRefStruct(refObject.toObject());
    this._map_urlIntialized = true;
    this._wrapperFields.map_url.copyRefObject(this._refObject.map_url);
  }
  copy(other) {
    this._refObject = new LoadMap_RequestRefStruct(other._refObject.toObject());
    this._map_urlIntialized = true;
    this._wrapperFields.map_url.copy(other._wrapperFields.map_url);
  }
  static get classType() {
    return LoadMap_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"map_url","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"LoadMap_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"LoadMap_Request"};
  }
  hasMember(name) {
    let memberNames = ["map_url"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class LoadMap_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LoadMap_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LoadMap_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LoadMap_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LoadMap_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new LoadMap_RequestRefArray(size);
    this._refObject = new LoadMap_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LoadMap_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LoadMap_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LoadMap_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LoadMap_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LoadMap_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LoadMap_RequestArrayWrapper;
  }
}
module.exports = LoadMap_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "map_url",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "nav2_msgs",
    "type": "LoadMap_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "LoadMap_Request"
}
*/
