// rclnodejs/generated/nav2_msgs/nav2_msgs__msg__CostmapMetaData.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TimeWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Time.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const PoseWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Pose.js');
const CostmapMetaDataRefStruct = StructType({
  map_load_time: TimeWrapper.refObjectType,
  update_time: TimeWrapper.refObjectType,
  layer: primitiveTypes.string,
  resolution: primitiveTypes.float32,
  size_x: primitiveTypes.uint32,
  size_y: primitiveTypes.uint32,
  origin: PoseWrapper.refObjectType,
});
const CostmapMetaDataRefArray = ArrayType(CostmapMetaDataRefStruct);
const CostmapMetaDataRefStructArray = StructType({
  data: CostmapMetaDataRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class CostmapMetaDataWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new CostmapMetaDataRefStruct(other._refObject.toObject());
      this._wrapperFields.map_load_time =  new TimeWrapper(other._wrapperFields.map_load_time);
      this._wrapperFields.update_time =  new TimeWrapper(other._wrapperFields.update_time);
      this._layerIntialized = true;
      this._wrapperFields.layer =  new StringWrapper(other._wrapperFields.layer);
      this._resolutionIntialized = true;
      this._size_xIntialized = true;
      this._size_yIntialized = true;
      this._wrapperFields.origin =  new PoseWrapper(other._wrapperFields.origin);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new CostmapMetaDataRefStruct();
    this._wrapperFields.map_load_time = new TimeWrapper();
    this._wrapperFields.update_time = new TimeWrapper();
    this._layerIntialized = false;
    this._wrapperFields.layer = new StringWrapper();
    this._resolutionIntialized = false;
    this._size_xIntialized = false;
    this._size_yIntialized = false;
    this._wrapperFields.origin = new PoseWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new CostmapMetaDataWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new CostmapMetaDataArrayWrapper;
  }
  static get ArrayType() {
    return CostmapMetaDataArrayWrapper;
  }
  static get refObjectArrayType() {
    return CostmapMetaDataRefStructArray
  }
  static get refObjectType() {
    return CostmapMetaDataRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._layerIntialized) {
        throw new TypeError('Invalid argument: layer in CostmapMetaData');
      }
      if (!this._resolutionIntialized) {
        throw new TypeError('Invalid argument: resolution in CostmapMetaData');
      }
      if (!this._size_xIntialized) {
        throw new TypeError('Invalid argument: size_x in CostmapMetaData');
      }
      if (!this._size_yIntialized) {
        throw new TypeError('Invalid argument: size_y in CostmapMetaData');
      }
    }
    this._wrapperFields.map_load_time.freeze(own, checkConsistency);
    this._refObject.map_load_time = this._wrapperFields.map_load_time.refObject;
    this._wrapperFields.update_time.freeze(own, checkConsistency);
    this._refObject.update_time = this._wrapperFields.update_time.refObject;
    if (own) {
      this._wrapperFields.layer.freeze(own, checkConsistency);
    }
    this._refObject.layer = this._wrapperFields.layer.refObject;
    this._wrapperFields.origin.freeze(own, checkConsistency);
    this._refObject.origin = this._wrapperFields.origin.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.map_load_time.copyRefObject(refObject.map_load_time);
    this._wrapperFields.update_time.copyRefObject(refObject.update_time);
    this._layerIntialized = true;
    this._wrapperFields.layer.data = refObject.layer.data;
    this._resolutionIntialized = true;
    this._refObject.resolution = refObject.resolution;
    this._size_xIntialized = true;
    this._refObject.size_x = refObject.size_x;
    this._size_yIntialized = true;
    this._refObject.size_y = refObject.size_y;
    this._wrapperFields.origin.copyRefObject(refObject.origin);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    TimeWrapper.freeStruct(refObject.map_load_time);
    TimeWrapper.freeStruct(refObject.update_time);
    StringWrapper.freeStruct(refObject.layer);
    PoseWrapper.freeStruct(refObject.origin);
  }
  static destoryRawROS(msg) {
    CostmapMetaDataWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'msg', interfaceName: 'CostmapMetaData'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get map_load_time() {
    return this._wrapperFields.map_load_time;
  }
  set map_load_time(value) {
    if (value instanceof TimeWrapper) {
      this._wrapperFields.map_load_time.copy(value);
    } else {
      this._wrapperFields.map_load_time.copy(new TimeWrapper(value));
    }
  }
  get update_time() {
    return this._wrapperFields.update_time;
  }
  set update_time(value) {
    if (value instanceof TimeWrapper) {
      this._wrapperFields.update_time.copy(value);
    } else {
      this._wrapperFields.update_time.copy(new TimeWrapper(value));
    }
  }
  get layer() {
    if (!this._layerIntialized) {
      return undefined;
    }
    return this._wrapperFields.layer.data;
  }
  set layer(value) {
    this._layerIntialized = true;
    this._wrapperFields.layer.data = value;
  }
  get resolution() {
    if (!this._resolutionIntialized) {
      return undefined;
    }
    return this._refObject.resolution;
  }
  set resolution(value) {
    this._resolutionIntialized = true;
    this._refObject.resolution = value;
  }
  get size_x() {
    if (!this._size_xIntialized) {
      return undefined;
    }
    return this._refObject.size_x;
  }
  set size_x(value) {
    this._size_xIntialized = true;
    this._refObject.size_x = value;
  }
  get size_y() {
    if (!this._size_yIntialized) {
      return undefined;
    }
    return this._refObject.size_y;
  }
  set size_y(value) {
    this._size_yIntialized = true;
    this._refObject.size_y = value;
  }
  get origin() {
    return this._wrapperFields.origin;
  }
  set origin(value) {
    if (value instanceof PoseWrapper) {
      this._wrapperFields.origin.copy(value);
    } else {
      this._wrapperFields.origin.copy(new PoseWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new CostmapMetaDataRefStruct(refObject.toObject());
    this._wrapperFields.map_load_time.copyRefObject(this._refObject.map_load_time);
    this._wrapperFields.update_time.copyRefObject(this._refObject.update_time);
    this._layerIntialized = true;
    this._wrapperFields.layer.copyRefObject(this._refObject.layer);
    this._resolutionIntialized = true;
    this._size_xIntialized = true;
    this._size_yIntialized = true;
    this._wrapperFields.origin.copyRefObject(this._refObject.origin);
  }
  copy(other) {
    this._refObject = new CostmapMetaDataRefStruct(other._refObject.toObject());
    this._wrapperFields.map_load_time.copy(other._wrapperFields.map_load_time);
    this._wrapperFields.update_time.copy(other._wrapperFields.update_time);
    this._layerIntialized = true;
    this._wrapperFields.layer.copy(other._wrapperFields.layer);
    this._resolutionIntialized = true;
    this._size_xIntialized = true;
    this._size_yIntialized = true;
    this._wrapperFields.origin.copy(other._wrapperFields.origin);
  }
  static get classType() {
    return CostmapMetaDataWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"map_load_time","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"update_time","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"layer","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"resolution","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"size_x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"size_y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"origin","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Pose","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"CostmapMetaData","stringUpperBound":null,"isPrimitiveType":false},"msgName":"CostmapMetaData"};
  }
  hasMember(name) {
    let memberNames = ["map_load_time","update_time","layer","resolution","size_x","size_y","origin"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class CostmapMetaDataArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof CostmapMetaDataWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new CostmapMetaDataWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CostmapMetaDataArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CostmapMetaDataArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new CostmapMetaDataRefArray(size);
    this._refObject = new CostmapMetaDataRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new CostmapMetaDataWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof CostmapMetaDataArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "CostmapMetaDataArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      CostmapMetaDataWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return CostmapMetaDataWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return CostmapMetaDataArrayWrapper;
  }
}
module.exports = CostmapMetaDataWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "map_load_time",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Time",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "update_time",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Time",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "layer",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "resolution",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "size_x",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "size_y",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "origin",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Pose",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "nav2_msgs",
    "type": "CostmapMetaData",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "CostmapMetaData"
}
*/
