// rclnodejs/generated/nav2_msgs/nav2_msgs__action__FollowPath_Feedback.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const FollowPath_FeedbackRefStruct = StructType({
  distance_to_goal: primitiveTypes.float32,
  speed: primitiveTypes.float32,
});
const FollowPath_FeedbackRefArray = ArrayType(FollowPath_FeedbackRefStruct);
const FollowPath_FeedbackRefStructArray = StructType({
  data: FollowPath_FeedbackRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class FollowPath_FeedbackWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new FollowPath_FeedbackRefStruct(other._refObject.toObject());
      this._distance_to_goalIntialized = true;
      this._speedIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new FollowPath_FeedbackRefStruct();
    this._distance_to_goalIntialized = false;
    this._speedIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new FollowPath_FeedbackWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new FollowPath_FeedbackArrayWrapper;
  }
  static get ArrayType() {
    return FollowPath_FeedbackArrayWrapper;
  }
  static get refObjectArrayType() {
    return FollowPath_FeedbackRefStructArray
  }
  static get refObjectType() {
    return FollowPath_FeedbackRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._distance_to_goalIntialized) {
        throw new TypeError('Invalid argument: distance_to_goal in FollowPath_Feedback');
      }
      if (!this._speedIntialized) {
        throw new TypeError('Invalid argument: speed in FollowPath_Feedback');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._distance_to_goalIntialized = true;
    this._refObject.distance_to_goal = refObject.distance_to_goal;
    this._speedIntialized = true;
    this._refObject.speed = refObject.speed;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    FollowPath_FeedbackWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'action', interfaceName: 'FollowPath_Feedback'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get distance_to_goal() {
    if (!this._distance_to_goalIntialized) {
      return undefined;
    }
    return this._refObject.distance_to_goal;
  }
  set distance_to_goal(value) {
    this._distance_to_goalIntialized = true;
    this._refObject.distance_to_goal = value;
  }
  get speed() {
    if (!this._speedIntialized) {
      return undefined;
    }
    return this._refObject.speed;
  }
  set speed(value) {
    this._speedIntialized = true;
    this._refObject.speed = value;
  }
  copyRefObject(refObject) {
    this._refObject = new FollowPath_FeedbackRefStruct(refObject.toObject());
    this._distance_to_goalIntialized = true;
    this._speedIntialized = true;
  }
  copy(other) {
    this._refObject = new FollowPath_FeedbackRefStruct(other._refObject.toObject());
    this._distance_to_goalIntialized = true;
    this._speedIntialized = true;
  }
  static get classType() {
    return FollowPath_FeedbackWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"distance_to_goal","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"speed","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"FollowPath_Feedback","stringUpperBound":null,"isPrimitiveType":false},"msgName":"FollowPath_Feedback"};
  }
  hasMember(name) {
    let memberNames = ["distance_to_goal","speed"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class FollowPath_FeedbackArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof FollowPath_FeedbackWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new FollowPath_FeedbackWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FollowPath_FeedbackArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FollowPath_FeedbackArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new FollowPath_FeedbackRefArray(size);
    this._refObject = new FollowPath_FeedbackRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new FollowPath_FeedbackWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof FollowPath_FeedbackArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "FollowPath_FeedbackArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      FollowPath_FeedbackWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return FollowPath_FeedbackWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return FollowPath_FeedbackArrayWrapper;
  }
}
module.exports = FollowPath_FeedbackWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "distance_to_goal",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "speed",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "nav2_msgs",
    "type": "FollowPath_Feedback",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "FollowPath_Feedback"
}
*/
