// rclnodejs/generated/nav2_msgs/nav2_msgs__action__Spin_Goal.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Spin_GoalRefStruct = StructType({
  target_yaw: primitiveTypes.float32,
});
const Spin_GoalRefArray = ArrayType(Spin_GoalRefStruct);
const Spin_GoalRefStructArray = StructType({
  data: Spin_GoalRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class Spin_GoalWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new Spin_GoalRefStruct(other._refObject.toObject());
      this._target_yawIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new Spin_GoalRefStruct();
    this._target_yawIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new Spin_GoalWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new Spin_GoalArrayWrapper;
  }
  static get ArrayType() {
    return Spin_GoalArrayWrapper;
  }
  static get refObjectArrayType() {
    return Spin_GoalRefStructArray
  }
  static get refObjectType() {
    return Spin_GoalRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._target_yawIntialized) {
        throw new TypeError('Invalid argument: target_yaw in Spin_Goal');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._target_yawIntialized = true;
    this._refObject.target_yaw = refObject.target_yaw;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    Spin_GoalWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'action', interfaceName: 'Spin_Goal'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get target_yaw() {
    if (!this._target_yawIntialized) {
      return undefined;
    }
    return this._refObject.target_yaw;
  }
  set target_yaw(value) {
    this._target_yawIntialized = true;
    this._refObject.target_yaw = value;
  }
  copyRefObject(refObject) {
    this._refObject = new Spin_GoalRefStruct(refObject.toObject());
    this._target_yawIntialized = true;
  }
  copy(other) {
    this._refObject = new Spin_GoalRefStruct(other._refObject.toObject());
    this._target_yawIntialized = true;
  }
  static get classType() {
    return Spin_GoalWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"target_yaw","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"Spin_Goal","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Spin_Goal"};
  }
  hasMember(name) {
    let memberNames = ["target_yaw"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class Spin_GoalArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof Spin_GoalWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new Spin_GoalWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Spin_GoalArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Spin_GoalArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new Spin_GoalRefArray(size);
    this._refObject = new Spin_GoalRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new Spin_GoalWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof Spin_GoalArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "Spin_GoalArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      Spin_GoalWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return Spin_GoalWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return Spin_GoalArrayWrapper;
  }
}
module.exports = Spin_GoalWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "target_yaw",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "nav2_msgs",
    "type": "Spin_Goal",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Spin_Goal"
}
*/
