// rclnodejs/generated/nav2_msgs/nav2_msgs__action__DummyRecovery_Goal.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const DummyRecovery_GoalRefStruct = StructType({
  command: StringWrapper.refObjectType,
});
const DummyRecovery_GoalRefArray = ArrayType(DummyRecovery_GoalRefStruct);
const DummyRecovery_GoalRefStructArray = StructType({
  data: DummyRecovery_GoalRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class DummyRecovery_GoalWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new DummyRecovery_GoalRefStruct(other._refObject.toObject());
      this._wrapperFields.command =  new StringWrapper(other._wrapperFields.command);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new DummyRecovery_GoalRefStruct();
    this._wrapperFields.command = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new DummyRecovery_GoalWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new DummyRecovery_GoalArrayWrapper;
  }
  static get ArrayType() {
    return DummyRecovery_GoalArrayWrapper;
  }
  static get refObjectArrayType() {
    return DummyRecovery_GoalRefStructArray
  }
  static get refObjectType() {
    return DummyRecovery_GoalRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.command.freeze(own, checkConsistency);
    this._refObject.command = this._wrapperFields.command.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.command.copyRefObject(refObject.command);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.command);
  }
  static destoryRawROS(msg) {
    DummyRecovery_GoalWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'action', interfaceName: 'DummyRecovery_Goal'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get command() {
    return this._wrapperFields.command;
  }
  set command(value) {
    if (value instanceof StringWrapper) {
      this._wrapperFields.command.copy(value);
    } else {
      this._wrapperFields.command.copy(new StringWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new DummyRecovery_GoalRefStruct(refObject.toObject());
    this._wrapperFields.command.copyRefObject(this._refObject.command);
  }
  copy(other) {
    this._refObject = new DummyRecovery_GoalRefStruct(other._refObject.toObject());
    this._wrapperFields.command.copy(other._wrapperFields.command);
  }
  static get classType() {
    return DummyRecovery_GoalWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"command","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"String","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"DummyRecovery_Goal","stringUpperBound":null,"isPrimitiveType":false},"msgName":"DummyRecovery_Goal"};
  }
  hasMember(name) {
    let memberNames = ["command"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class DummyRecovery_GoalArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof DummyRecovery_GoalWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new DummyRecovery_GoalWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DummyRecovery_GoalArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DummyRecovery_GoalArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new DummyRecovery_GoalRefArray(size);
    this._refObject = new DummyRecovery_GoalRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new DummyRecovery_GoalWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof DummyRecovery_GoalArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "DummyRecovery_GoalArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      DummyRecovery_GoalWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return DummyRecovery_GoalWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return DummyRecovery_GoalArrayWrapper;
  }
}
module.exports = DummyRecovery_GoalWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "command",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "String",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "nav2_msgs",
    "type": "DummyRecovery_Goal",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "DummyRecovery_Goal"
}
*/
