// rclnodejs/generated/nav2_msgs/nav2_msgs__msg__BehaviorTreeStatusChange.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TimeWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Time.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const BehaviorTreeStatusChangeRefStruct = StructType({
  timestamp: TimeWrapper.refObjectType,
  node_name: primitiveTypes.string,
  previous_status: primitiveTypes.string,
  current_status: primitiveTypes.string,
});
const BehaviorTreeStatusChangeRefArray = ArrayType(BehaviorTreeStatusChangeRefStruct);
const BehaviorTreeStatusChangeRefStructArray = StructType({
  data: BehaviorTreeStatusChangeRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class BehaviorTreeStatusChangeWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new BehaviorTreeStatusChangeRefStruct(other._refObject.toObject());
      this._wrapperFields.timestamp =  new TimeWrapper(other._wrapperFields.timestamp);
      this._node_nameIntialized = true;
      this._wrapperFields.node_name =  new StringWrapper(other._wrapperFields.node_name);
      this._previous_statusIntialized = true;
      this._wrapperFields.previous_status =  new StringWrapper(other._wrapperFields.previous_status);
      this._current_statusIntialized = true;
      this._wrapperFields.current_status =  new StringWrapper(other._wrapperFields.current_status);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new BehaviorTreeStatusChangeRefStruct();
    this._wrapperFields.timestamp = new TimeWrapper();
    this._node_nameIntialized = false;
    this._wrapperFields.node_name = new StringWrapper();
    this._previous_statusIntialized = false;
    this._wrapperFields.previous_status = new StringWrapper();
    this._current_statusIntialized = false;
    this._wrapperFields.current_status = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new BehaviorTreeStatusChangeWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new BehaviorTreeStatusChangeArrayWrapper;
  }
  static get ArrayType() {
    return BehaviorTreeStatusChangeArrayWrapper;
  }
  static get refObjectArrayType() {
    return BehaviorTreeStatusChangeRefStructArray
  }
  static get refObjectType() {
    return BehaviorTreeStatusChangeRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._node_nameIntialized) {
        throw new TypeError('Invalid argument: node_name in BehaviorTreeStatusChange');
      }
      if (!this._previous_statusIntialized) {
        throw new TypeError('Invalid argument: previous_status in BehaviorTreeStatusChange');
      }
      if (!this._current_statusIntialized) {
        throw new TypeError('Invalid argument: current_status in BehaviorTreeStatusChange');
      }
    }
    this._wrapperFields.timestamp.freeze(own, checkConsistency);
    this._refObject.timestamp = this._wrapperFields.timestamp.refObject;
    if (own) {
      this._wrapperFields.node_name.freeze(own, checkConsistency);
    }
    this._refObject.node_name = this._wrapperFields.node_name.refObject;
    if (own) {
      this._wrapperFields.previous_status.freeze(own, checkConsistency);
    }
    this._refObject.previous_status = this._wrapperFields.previous_status.refObject;
    if (own) {
      this._wrapperFields.current_status.freeze(own, checkConsistency);
    }
    this._refObject.current_status = this._wrapperFields.current_status.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.timestamp.copyRefObject(refObject.timestamp);
    this._node_nameIntialized = true;
    this._wrapperFields.node_name.data = refObject.node_name.data;
    this._previous_statusIntialized = true;
    this._wrapperFields.previous_status.data = refObject.previous_status.data;
    this._current_statusIntialized = true;
    this._wrapperFields.current_status.data = refObject.current_status.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    TimeWrapper.freeStruct(refObject.timestamp);
    StringWrapper.freeStruct(refObject.node_name);
    StringWrapper.freeStruct(refObject.previous_status);
    StringWrapper.freeStruct(refObject.current_status);
  }
  static destoryRawROS(msg) {
    BehaviorTreeStatusChangeWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'msg', interfaceName: 'BehaviorTreeStatusChange'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    return this._wrapperFields.timestamp;
  }
  set timestamp(value) {
    if (value instanceof TimeWrapper) {
      this._wrapperFields.timestamp.copy(value);
    } else {
      this._wrapperFields.timestamp.copy(new TimeWrapper(value));
    }
  }
  get node_name() {
    if (!this._node_nameIntialized) {
      return undefined;
    }
    return this._wrapperFields.node_name.data;
  }
  set node_name(value) {
    this._node_nameIntialized = true;
    this._wrapperFields.node_name.data = value;
  }
  get previous_status() {
    if (!this._previous_statusIntialized) {
      return undefined;
    }
    return this._wrapperFields.previous_status.data;
  }
  set previous_status(value) {
    this._previous_statusIntialized = true;
    this._wrapperFields.previous_status.data = value;
  }
  get current_status() {
    if (!this._current_statusIntialized) {
      return undefined;
    }
    return this._wrapperFields.current_status.data;
  }
  set current_status(value) {
    this._current_statusIntialized = true;
    this._wrapperFields.current_status.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new BehaviorTreeStatusChangeRefStruct(refObject.toObject());
    this._wrapperFields.timestamp.copyRefObject(this._refObject.timestamp);
    this._node_nameIntialized = true;
    this._wrapperFields.node_name.copyRefObject(this._refObject.node_name);
    this._previous_statusIntialized = true;
    this._wrapperFields.previous_status.copyRefObject(this._refObject.previous_status);
    this._current_statusIntialized = true;
    this._wrapperFields.current_status.copyRefObject(this._refObject.current_status);
  }
  copy(other) {
    this._refObject = new BehaviorTreeStatusChangeRefStruct(other._refObject.toObject());
    this._wrapperFields.timestamp.copy(other._wrapperFields.timestamp);
    this._node_nameIntialized = true;
    this._wrapperFields.node_name.copy(other._wrapperFields.node_name);
    this._previous_statusIntialized = true;
    this._wrapperFields.previous_status.copy(other._wrapperFields.previous_status);
    this._current_statusIntialized = true;
    this._wrapperFields.current_status.copy(other._wrapperFields.current_status);
  }
  static get classType() {
    return BehaviorTreeStatusChangeWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"node_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"previous_status","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"current_status","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"BehaviorTreeStatusChange","stringUpperBound":null,"isPrimitiveType":false},"msgName":"BehaviorTreeStatusChange"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","node_name","previous_status","current_status"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class BehaviorTreeStatusChangeArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof BehaviorTreeStatusChangeWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new BehaviorTreeStatusChangeWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to BehaviorTreeStatusChangeArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to BehaviorTreeStatusChangeArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new BehaviorTreeStatusChangeRefArray(size);
    this._refObject = new BehaviorTreeStatusChangeRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new BehaviorTreeStatusChangeWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof BehaviorTreeStatusChangeArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "BehaviorTreeStatusChangeArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      BehaviorTreeStatusChangeWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return BehaviorTreeStatusChangeWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return BehaviorTreeStatusChangeArrayWrapper;
  }
}
module.exports = BehaviorTreeStatusChangeWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Time",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "node_name",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "previous_status",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "current_status",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "nav2_msgs",
    "type": "BehaviorTreeStatusChange",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "BehaviorTreeStatusChange"
}
*/
