// rclnodejs/generated/nav2_msgs/nav2_msgs__srv__ClearCostmapAroundRobot_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const ClearCostmapAroundRobot_RequestRefStruct = StructType({
  window_size_x: primitiveTypes.float32,
  window_size_y: primitiveTypes.float32,
});
const ClearCostmapAroundRobot_RequestRefArray = ArrayType(ClearCostmapAroundRobot_RequestRefStruct);
const ClearCostmapAroundRobot_RequestRefStructArray = StructType({
  data: ClearCostmapAroundRobot_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ClearCostmapAroundRobot_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ClearCostmapAroundRobot_RequestRefStruct(other._refObject.toObject());
      this._window_size_xIntialized = true;
      this._window_size_yIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ClearCostmapAroundRobot_RequestRefStruct();
    this._window_size_xIntialized = false;
    this._window_size_yIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new ClearCostmapAroundRobot_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ClearCostmapAroundRobot_RequestArrayWrapper;
  }
  static get ArrayType() {
    return ClearCostmapAroundRobot_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return ClearCostmapAroundRobot_RequestRefStructArray
  }
  static get refObjectType() {
    return ClearCostmapAroundRobot_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._window_size_xIntialized) {
        throw new TypeError('Invalid argument: window_size_x in ClearCostmapAroundRobot_Request');
      }
      if (!this._window_size_yIntialized) {
        throw new TypeError('Invalid argument: window_size_y in ClearCostmapAroundRobot_Request');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._window_size_xIntialized = true;
    this._refObject.window_size_x = refObject.window_size_x;
    this._window_size_yIntialized = true;
    this._refObject.window_size_y = refObject.window_size_y;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    ClearCostmapAroundRobot_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'srv', interfaceName: 'ClearCostmapAroundRobot_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get window_size_x() {
    if (!this._window_size_xIntialized) {
      return undefined;
    }
    return this._refObject.window_size_x;
  }
  set window_size_x(value) {
    this._window_size_xIntialized = true;
    this._refObject.window_size_x = value;
  }
  get window_size_y() {
    if (!this._window_size_yIntialized) {
      return undefined;
    }
    return this._refObject.window_size_y;
  }
  set window_size_y(value) {
    this._window_size_yIntialized = true;
    this._refObject.window_size_y = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ClearCostmapAroundRobot_RequestRefStruct(refObject.toObject());
    this._window_size_xIntialized = true;
    this._window_size_yIntialized = true;
  }
  copy(other) {
    this._refObject = new ClearCostmapAroundRobot_RequestRefStruct(other._refObject.toObject());
    this._window_size_xIntialized = true;
    this._window_size_yIntialized = true;
  }
  static get classType() {
    return ClearCostmapAroundRobot_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"window_size_x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"window_size_y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"ClearCostmapAroundRobot_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ClearCostmapAroundRobot_Request"};
  }
  hasMember(name) {
    let memberNames = ["window_size_x","window_size_y"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ClearCostmapAroundRobot_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ClearCostmapAroundRobot_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ClearCostmapAroundRobot_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ClearCostmapAroundRobot_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ClearCostmapAroundRobot_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ClearCostmapAroundRobot_RequestRefArray(size);
    this._refObject = new ClearCostmapAroundRobot_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ClearCostmapAroundRobot_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ClearCostmapAroundRobot_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ClearCostmapAroundRobot_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ClearCostmapAroundRobot_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ClearCostmapAroundRobot_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ClearCostmapAroundRobot_RequestArrayWrapper;
  }
}
module.exports = ClearCostmapAroundRobot_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "window_size_x",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "window_size_y",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "nav2_msgs",
    "type": "ClearCostmapAroundRobot_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ClearCostmapAroundRobot_Request"
}
*/
